
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

    <title>Changelog &#8212; lcdb-wf 0.1 documentation</title>
    <link rel="stylesheet" type="text/css" href="_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="_static/alabaster.css" />
    <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
    <script src="_static/jquery.js"></script>
    <script src="_static/underscore.js"></script>
    <script src="_static/doctools.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="For Developers" href="developers.html" />
    <link rel="prev" title="FAQs" href="faqs.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  <div class="document">
    
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="toc.html">lcdb-wf</a></h1>



<p class="blurb">Customizable workflows for high-throughput sequencing analysis</p>






<h3>Navigation</h3>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="index.html">Introduction</a></li>
<li class="toctree-l1"><a class="reference internal" href="getting-started.html">Getting started</a></li>
<li class="toctree-l1"><a class="reference internal" href="guide.html">Guide to file hierarchy</a></li>
<li class="toctree-l1"><a class="reference internal" href="tests.html">Testing the installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="workflows.html">Overview of workflows</a></li>
<li class="toctree-l1"><a class="reference internal" href="config.html">Configuration details</a></li>
<li class="toctree-l1"><a class="reference internal" href="references.html">References workflow</a></li>
<li class="toctree-l1"><a class="reference internal" href="rnaseq.html">RNA-seq workflow</a></li>
<li class="toctree-l1"><a class="reference internal" href="downstream-rnaseq.html">RNA-Seq downstream analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="chipseq.html">ChIP-seq workflow</a></li>
<li class="toctree-l1"><a class="reference internal" href="integrative.html">Integrative workflows</a></li>
<li class="toctree-l1"><a class="reference internal" href="faqs.html">FAQs</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Changelog</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#v1-8">v1.8</a></li>
<li class="toctree-l2"><a class="reference internal" href="#v1-7">v1.7</a></li>
<li class="toctree-l2"><a class="reference internal" href="#v1-6">v1.6</a></li>
<li class="toctree-l2"><a class="reference internal" href="#v1-5-3">v1.5.3</a></li>
<li class="toctree-l2"><a class="reference internal" href="#v1-5-2">v1.5.2</a></li>
<li class="toctree-l2"><a class="reference internal" href="#v1-5-1">v1.5.1</a></li>
<li class="toctree-l2"><a class="reference internal" href="#v1-5-sept-2019">v1.5 (Sept 2019)</a></li>
<li class="toctree-l2"><a class="reference internal" href="#v1-4-2-jul-2019">v1.4.2 (Jul 2019)</a></li>
<li class="toctree-l2"><a class="reference internal" href="#v1-4-1-jul-2019">v1.4.1 (Jul 2019)</a></li>
<li class="toctree-l2"><a class="reference internal" href="#v1-4-may-2019">v1.4 (May 2019)</a></li>
<li class="toctree-l2"><a class="reference internal" href="#v1-3-may-2019">v1.3 (May 2019)</a></li>
<li class="toctree-l2"><a class="reference internal" href="#v1-2-mar-2019">v1.2 (Mar 2019)</a></li>
<li class="toctree-l2"><a class="reference internal" href="#v1-1-aug-2018">v1.1 (Aug 2018)</a></li>
<li class="toctree-l2"><a class="reference internal" href="#v1-0-1-jun-2018">v1.0.1 (Jun 2018)</a></li>
<li class="toctree-l2"><a class="reference internal" href="#v1-0-may-2018">v1.0 (May 2018)</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="developers.html">For Developers</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="toc.html">Documentation overview</a><ul>
      <li>Previous: <a href="faqs.html" title="previous chapter">FAQs</a></li>
      <li>Next: <a href="developers.html" title="next chapter">For Developers</a></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" />
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <section id="changelog">
<h1>Changelog<a class="headerlink" href="#changelog" title="Permalink to this headline">¶</a></h1>
<section id="v1-8">
<h2>v1.8<a class="headerlink" href="#v1-8" title="Permalink to this headline">¶</a></h2>
<section id="general">
<h3>General<a class="headerlink" href="#general" title="Permalink to this headline">¶</a></h3>
<ul>
<li><p>Complete shift to using pinned <code class="docutils literal notranslate"><span class="pre">env.yaml</span></code> files to specify conda
environments, and using <code class="docutils literal notranslate"><span class="pre">mamba</span></code> for building environments (consistent with
recent versions of Snakemake). This is now reflected in documentation and
the updated-and-improved <code class="docutils literal notranslate"><span class="pre">deploy.py</span></code>.</p></li>
<li><p>Reorganization/cleanup of the <code class="docutils literal notranslate"><span class="pre">include</span></code> directory</p></li>
<li><p>Added conda troubleshooting notes to the documentation (see
<a class="reference internal" href="conda.html#conda-troubleshooting"><span class="std std-ref">Troubleshooting environments</span></a>).</p></li>
<li><p>The <code class="docutils literal notranslate"><span class="pre">lib.helpers.preflight</span></code> function no requires the first column of the
sampletable to be named <cite>samplename</cite> when checking configs.</p></li>
<li><p>Improvements to the deployment script <code class="docutils literal notranslate"><span class="pre">deploy.py</span></code>:</p>
<blockquote>
<div><ul class="simple">
<li><p>now requires Python &gt;3.6</p></li>
<li><p>proper logs (so you can easily see how long it takes to build an env)</p></li>
<li><p>supports downloading and running the script directly, which will clone
a temporary copy and deploy from there</p></li>
<li><p>using Control-C to stop the deployment will also stop mamba/conda</p></li>
<li><p>colored output</p></li>
<li><p>mamba is used by default, but <code class="docutils literal notranslate"><span class="pre">--conda-frontend</span></code> will use conda instead</p></li>
</ul>
</div></blockquote>
</li>
<li><p>fastq-dump log is sent to file rather than printed to stdout</p></li>
<li><p>Threads: cutadapt single-end now uses specified threads (it was using
1 thread by default); use 6 threads for fastqc</p></li>
<li><p>Added new preflight checks for RNA-seq and ChIP-seq specific configs.</p></li>
<li><p>Added a <code class="docutils literal notranslate"><span class="pre">run_complex_test.sh</span></code> driver script for testing the workflows on
full-scale publicly available data</p></li>
</ul>
</section>
<section id="rna-seq">
<h3>RNA-seq<a class="headerlink" href="#rna-seq" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p><strong>Configuration change:</strong> The <code class="docutils literal notranslate"><span class="pre">stranded:</span></code> field is now required for RNA-seq.
This is used to choose the correct parameters for various rules, and avoids
one of the main reasons to edit the Snakefile. See <a class="reference internal" href="config-yaml.html#cfg-stranded"><span class="std std-ref">stranded field</span></a> for
more details on its use.</p></li>
<li><p>added <code class="docutils literal notranslate"><span class="pre">stranded:</span></code> field to all configs used in testing</p></li>
<li><p>The <code class="docutils literal notranslate"><span class="pre">strand_check</span></code> rule now runs MultiQC for a convenient way of evaluating
strandedness of a library.</p></li>
<li><p>Kallisto is now supported in both the RNA-seq Snakefile, references
Snakefile, included reference configs, and downstream <code class="docutils literal notranslate"><span class="pre">rnaseq.Rmd</span></code></p></li>
</ul>
</section>
<section id="references">
<h3>References<a class="headerlink" href="#references" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p>When checking URLs in reference configs, don’t use <code class="docutils literal notranslate"><span class="pre">curl</span></code> to check
<code class="docutils literal notranslate"><span class="pre">file://</span></code> URIs.</p></li>
<li><p>There is a new feature for reference configs that allows chaining
post-processing functions together, see <a class="reference internal" href="references-config.html#advanced-postprocessing"><span class="std std-ref">More advanced postprocessing</span></a>. This
means that it is possible, for example, to add ERCC spike-ins (which need
post-processing) onto references that themselves need post-processing.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">lib/postprocess/ercc.py</span></code> has new helper functions for adding ERCC
spike-ins to fasta files and GTF files.</p></li>
<li><p>added <code class="docutils literal notranslate"><span class="pre">'kallisto'</span></code> to included reference configs</p></li>
</ul>
</section>
<section id="chip-seq">
<h3>ChIP-seq<a class="headerlink" href="#chip-seq" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p>symlinks rule is now local</p></li>
<li><p>added collectinsertsizes pattern to support PE ChIP-seq experiments</p></li>
<li><p>merging bigwigs log no longer goes to stdout</p></li>
</ul>
</section>
</section>
<section id="v1-7">
<h2>v1.7<a class="headerlink" href="#v1-7" title="Permalink to this headline">¶</a></h2>
<section id="setup">
<h3>Setup<a class="headerlink" href="#setup" title="Permalink to this headline">¶</a></h3>
<p>Use mamba for installation of environments, consistent with Snakemake recommendations</p>
</section>
<section id="testing">
<h3>Testing<a class="headerlink" href="#testing" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p>We now recommend using <a class="reference external" href="https://github.com/mamba-org/mamba">mamba</a> to
create conda environments. This is dramatically faster and solves some
dependency issues. Our automated tests now use this.</p></li>
<li><p>We have moved from requirements.txt files to env.yaml files. We also now
encourage the use of the strictly-pinned environments for a more stable
experience to hopefully avoid transient issues in the packaging ecosystem.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">tbb=2020.2</span></code> as a dependency to fix a recent packaging issue with conda-forge.</p></li>
<li><p>many documentation improvements</p></li>
<li><p>symlinks rule is only set to localrule when it exists (it does not exist when
running an analysis exclusively from SRA)</p></li>
</ul>
</section>
<section id="id1">
<h3>References<a class="headerlink" href="#id1" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p>updated URLs for those that have changes (e.g., Sanger -&gt; EBI; using https
instead of ftp for UCSC-hosted genomes)</p></li>
<li><p>new <code class="docutils literal notranslate"><span class="pre">gff2gtf</span></code> post-process tool for when an annotation is only available as
GFF. <em>S. pombe</em> needs this, for example, and the
<cite>Schizosaccharomyces_pombe.yaml`</cite> reference config has been updated
accordingly.</p></li>
<li><p>The references workflow no longer reads the config file in its directory.
This fixes some subtle overwriting issues when providing config files or
items from the command line during as is used during certain test runs. If
running the references workflow alone, it must be called with
<code class="docutils literal notranslate"><span class="pre">--configfile</span></code></p></li>
</ul>
</section>
<section id="id2">
<h3>RNA-seq<a class="headerlink" href="#id2" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p>featureCounts now uses BAM files with duplicates marked. Previously if you
wanted to run featureCounts in a mode where it excluded duplicates you would
need to reconfigure rules.</p></li>
<li><p>improved comments in RNA-seq downstream RMarkdown files</p></li>
</ul>
</section>
<section id="id3">
<h3>Testing<a class="headerlink" href="#id3" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p>new test that checks all URLs identified in config files to ensure that the
included reference files remain valid</p></li>
<li><p>there is now a separate <code class="docutils literal notranslate"><span class="pre">run_downstream_test</span></code> script`</p></li>
<li><p>simplified the CircleCI DAG to optimize testing resources</p></li>
</ul>
</section>
</section>
<section id="v1-6">
<h2>v1.6<a class="headerlink" href="#v1-6" title="Permalink to this headline">¶</a></h2>
<section id="id4">
<h3>References<a class="headerlink" href="#id4" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p>overhaul the way transcriptome fastas are created. Instead of requiring
separate download, they are now created out of the provided GTF and fasta
files. The reference config section now uses keys <code class="docutils literal notranslate"><span class="pre">genome:</span></code>,
<code class="docutils literal notranslate"><span class="pre">transcriptome:</span></code>, and <code class="docutils literal notranslate"><span class="pre">annotation:</span></code> rather than the <code class="docutils literal notranslate"><span class="pre">fasta:</span></code> and
<code class="docutils literal notranslate"><span class="pre">gtf:</span></code> keys.</p></li>
<li><p><strong>backwards-incompatible change:</strong> reference config files have been updated
to reflect the changes in the references workflow</p></li>
<li><p>Update PhiX genome fasta to use NCBI rather than Illumina iGenomes</p></li>
</ul>
</section>
<section id="chip-seq-workflow">
<h3>ChIP-seq workflow<a class="headerlink" href="#chip-seq-workflow" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p>ChIP-seq workflow now properly supports paired-end reads</p></li>
<li><p>A ChIP-seq workflow can now be run when the <code class="docutils literal notranslate"><span class="pre">chipseq:</span></code> and/or
<code class="docutils literal notranslate"><span class="pre">peak_calling:</span></code> sections are omitted.</p></li>
<li><p>added a missing bowtie2 config entry in <code class="docutils literal notranslate"><span class="pre">clusterconfig.yaml</span></code> which could
result in out-of-memory errors when submitting to a cluster using that file</p></li>
</ul>
</section>
<section id="rna-seq-workflow">
<h3>RNA-seq workflow<a class="headerlink" href="#rna-seq-workflow" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p>if colData is a tibble this no longer causes issues for importing counts</p></li>
<li><p>dupRadar removed from RNA-seq workflow. We ended up never using it, and it
depends on R which we’ve since removed from the main environment.</p></li>
<li><p>new <code class="docutils literal notranslate"><span class="pre">strand_test</span></code> rule, which can be run explicitly with <code class="docutils literal notranslate"><span class="pre">snakemake</span> <span class="pre">-j2</span>
<span class="pre">strand_check</span></code>. This generates <code class="docutils literal notranslate"><span class="pre">strandedness.tsv</span></code> in the current directory,
which is the summarize output of RSeQC’s <code class="docutils literal notranslate"><span class="pre">infer_experiment.py</span></code> across all
samples.</p></li>
<li><p>implement STAR two-pass alignment. Default is still single-pass.</p></li>
<li><p>Clean up hard-coded STAR indexing Log.out file</p></li>
<li><p>Include <code class="docutils literal notranslate"><span class="pre">ashr</span></code> and <code class="docutils literal notranslate"><span class="pre">ihw</span></code> Bioconductor packages in the R requirements, for
use with recent versions of DESeq2.</p></li>
</ul>
</section>
<section id="rna-seq-downstream">
<h3>RNA-seq downstream<a class="headerlink" href="#rna-seq-downstream" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p>Functional enrichment and gene patterns are now separate child documents.
This makes it easier to turn them on/off by only needing to adjust the chunk
options of the child chunk</p></li>
<li><p>Created a new documentation method for rnaseq.Rmd. Now there is a separate,
dedicated documentation page with sections that exactly correspond to each
named chunk in the Rmd, as well as a tool for ensuring that chunks and docs
stay synchronized. See <a class="reference internal" href="rnaseq-rmd.html#rnaseqrmd"><span class="std std-ref">Primary analysis script</span></a> for the new docs.</p></li>
<li><p>New <code class="docutils literal notranslate"><span class="pre">counts.df</span></code> and <code class="docutils literal notranslate"><span class="pre">counts.plot</span></code> functions to make it much easier to
make custom dotplots of top counts by melting and joining the counts table
with the metadata in colData.</p></li>
<li><p>DEGpatterns cluster IDs are now added as additional columns in the output
TSVs for each contrast</p></li>
<li><p>Many functions in the rnaseq.Rmd now expect a list of <a class="reference internal" href="rnaseq-rmd.html#term-dds"><span class="xref std std-term">dds</span></a> objects.
See <a class="reference internal" href="rnaseq-rmd.html#dds-list"><span class="std std-ref">dds_list</span></a> for more info on this.</p></li>
<li><p>Created a new R package, <code class="docutils literal notranslate"><span class="pre">lcdbwf</span></code> stored in <code class="file docutils literal notranslate"><span class="pre">lib/lcdbwf</span></code>. This can be
edited in place, and it is loaded from disk within <code class="docutils literal notranslate"><span class="pre">rnaseq.Rmd</span></code>.</p></li>
<li><p>Modified some output keys to support recent versions of Snakemake, for which
<code class="docutils literal notranslate"><span class="pre">count</span></code> is a reserved keyword</p></li>
</ul>
</section>
<section id="id5">
<h3>General<a class="headerlink" href="#id5" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p>Conda environments are now split into R and non-R. See <a class="reference internal" href="conda.html#conda-envs"><span class="std std-ref">conda and conda envs in lcdb-wf</span></a> for
details. Updated <code class="docutils literal notranslate"><span class="pre">deploy.py</span></code> accordingly</p></li>
<li><p>symlinks rules are now set to be localrules</p></li>
<li><p>updated workflows to work on recent Snakemake versions</p></li>
<li><p>split environments into non-R and R. This, along with a loose pinning of
versions (<code class="docutils literal notranslate"><span class="pre">&gt;=</span></code>), dramatically speeds up environment creation.</p></li>
<li><p>updates to support latest Snakemake versions</p></li>
<li><dl class="simple">
<dt>improvements to testing:</dt><dd><ul>
<li><p>environment YAML files, rendered HTML, and docs are stored as artifacts on CircleCI</p></li>
<li><p>consolidations of some RNA-seq tests to reduce total time</p></li>
<li><p>additional comments in the test config yaml to help new users understand the system</p></li>
</ul>
</dd>
</dl>
</li>
<li><p>new “preflight check” function is run to hopefully catch errors before running workflows</p></li>
<li><p>updates to support recent Picard versions</p></li>
<li><p>added wildcard constraints to help Snakemake solve DAG</p></li>
</ul>
</section>
</section>
<section id="v1-5-3">
<h2>v1.5.3<a class="headerlink" href="#v1-5-3" title="Permalink to this headline">¶</a></h2>
<section id="id6">
<h3>General<a class="headerlink" href="#id6" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p>default 12-hr wall time in WRAPPER_SLURM</p></li>
<li><p>update .gitignore (<a class="reference external" href="https://github.com/lcdb/lcdb-wf/issues/223">#223</a>)</p></li>
<li><p>remove the FastQC status checks section from the MultiQC report (which shows
up in recent MultiQC versions) (<a class="reference external" href="https://github.com/lcdb/lcdb-wf/issues/246">#246</a></p></li>
</ul>
</section>
<section id="bugs">
<h3>Bugs<a class="headerlink" href="#bugs" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p>add bed12 conversion for all species with default reference configs</p></li>
<li><p>presence of an orig_filename_R2 in sampletable is sufficient to consider the
experiment PE</p></li>
<li><p>ensure DEGpattern output only contains unique genes</p></li>
<li><p>bring back featurecounts in multiqc report</p></li>
<li><p>“attach” chunk in rnaseq.Rmd was not properly set to depend on the “results” chunk</p></li>
</ul>
</section>
<section id="id9">
<h3>RNA-seq<a class="headerlink" href="#id9" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p>dds objects can now be created from a full featureCounts input file and
a subsetted colData table, if subset.counts=TRUE</p></li>
<li><p>improve the dependencies between rnaseq.Rmd chunks so that cache=TRUE behaves
as expected: (<a class="reference external" href="https://github.com/lcdb/lcdb-wf/issues/232">#232</a>)</p></li>
<li><p>add plots for rnaseq.Rmd size factors (<a class="reference external" href="https://github.com/lcdb/lcdb-wf/issues/222">#222</a>)</p></li>
<li><p>run rseqc instead of CollectRnaSeqMetrics (the multiqc output is nicer for
it, and it’s pretty much doing the same thing) (<a class="reference external" href="https://github.com/lcdb/lcdb-wf/issues/218">#218</a>)</p></li>
<li><p>when converting Ensembl to symbol, if there is no symbol then fall back to
the Ensembl ID to avoid NA (<a class="reference external" href="https://github.com/lcdb/lcdb-wf/issues/246">#246</a>)</p></li>
<li><p>in rnaseq.Rmd, all caches will be invalidated if the sampletable or the
featurecounts table have changed.</p></li>
</ul>
</section>
<section id="tests">
<h3>Tests<a class="headerlink" href="#tests" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p>using continuumio/miniconda3 container; finally got en_US.utf8 locale
installed and working correctly in that container so that multiqc works.</p></li>
</ul>
</section>
</section>
<section id="v1-5-2">
<h2>v1.5.2<a class="headerlink" href="#v1-5-2" title="Permalink to this headline">¶</a></h2>
<section id="bug-fixes">
<h3>Bug fixes<a class="headerlink" href="#bug-fixes" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p>When some samples were substrings of other samples (e.g., <cite>WT_1_1</cite> and
<cite>WT_1_10</cite>), DESeqDataSetFromCombinedFeatureCounts was assigning the wrong
names. This has now been fixed in <cite>helpers.Rmd</cite>.</p></li>
</ul>
</section>
</section>
<section id="v1-5-1">
<h2>v1.5.1<a class="headerlink" href="#v1-5-1" title="Permalink to this headline">¶</a></h2>
<section id="id14">
<h3>Bug fixes<a class="headerlink" href="#id14" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p>DESeqDataSetFromCombinedFeatureCounts (added in v1.5) was incorrectly
assigning labels to samples when the order of the sampletable did not match
the order of the samples in the featureCounts table columns. This has been
fixed.</p></li>
</ul>
</section>
<section id="id15">
<h3>General<a class="headerlink" href="#id15" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p><cite>deploy.py</cite> deployment script now only pays attention to files checked in to
version control and optionally can create a conda environment in the target
directory.</p></li>
<li><p>tests now work out of a newly-deployed instance to better reflect real-world
usage</p></li>
</ul>
</section>
<section id="chip-seq-and-rna-seq">
<h3>ChIP-seq and RNA-seq<a class="headerlink" href="#chip-seq-and-rna-seq" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p>reorder cutadapt commands to avoid a MultQC parsing bug in the cutadapt
module (see <a class="reference external" href="https://github.com/ewels/MultiQC/issues/949">https://github.com/ewels/MultiQC/issues/949</a>)</p></li>
</ul>
</section>
<section id="id16">
<h3>RNA-seq<a class="headerlink" href="#id16" title="Permalink to this headline">¶</a></h3>
<p>The majority of these changes affect <code class="docutils literal notranslate"><span class="pre">rnaseq.Rmd</span></code>:</p>
<ul class="simple">
<li><p>modifications to MultiQC config to get back featureCounts output</p></li>
<li><p><cite>plotMA.label</cite> function (in <code class="docutils literal notranslate"><span class="pre">helpers.Rmd</span></code>) now defaults to FDR &lt; 0.1
(instead of 0.01), and additionally supports labeling using different columns
of the results object (e.g., “symbol”).</p></li>
<li><p>remove some now-redundant featureCounts code</p></li>
<li><p>add a comment showing where to collapse replicates</p></li>
<li><p>convert colData’s first column to rownames</p></li>
<li><p>implement lower limit for DEGpatterns clustering (default is 0, but can
easily set to higher if you’re getting issues)</p></li>
<li><p>expose arbitrary additional function arguments to <code class="docutils literal notranslate"><span class="pre">top.plots</span></code>. This allows
different <cite>intgroup</cite> arguments to be passed to the <cite>my.counts</cite> function,
enabling different ways of plotting the gene dotplots.</p></li>
</ul>
</section>
</section>
<section id="v1-5-sept-2019">
<h2>v1.5 (Sept 2019)<a class="headerlink" href="#v1-5-sept-2019" title="Permalink to this headline">¶</a></h2>
<p>Major change: <strong>it is no longer possible to mix single-end and paired-end
samples within the same run of the workflow.</strong> See <a class="reference external" href="https://github.com/lcdb/lcdb-wf/pull/208">#208</a> and the corresponding issue
description at <a class="reference external" href="https://github.com/lcdb/lcdb-wf/issues/175">#175</a>.</p>
<p>This version also has many improvements to the <code class="docutils literal notranslate"><span class="pre">rnaseq.Rmd</span></code> file for RNA-seq,
as described below.</p>
<section id="id19">
<h3>RNA-seq<a class="headerlink" href="#id19" title="Permalink to this headline">¶</a></h3>
<p>Many changes and improvements to <code class="docutils literal notranslate"><span class="pre">rnaseq.Rmd</span></code>, including:</p>
<ul class="simple">
<li><p>Differential analysis summaries now include labeled MA plots (<a class="reference external" href="https://github.com/lcdb/lcdb-wf/pull/192/files">#192</a>)</p></li>
<li><p>PCA plots now use plotly for improved insepction of samples (<a class="reference external" href="https://github.com/lcdb/lcdb-wf/pull/192/files">#192</a></p></li>
<li><p>don’t use knitrBootstrap any more (<a class="reference external" href="https://github.com/lcdb/lcdb-wf/pull/192/files">#192</a></p></li>
<li><p>heatmaps use heatmaply package for better interaction (<a class="reference external" href="https://github.com/lcdb/lcdb-wf/pull/192/files">#192</a></p></li>
<li><p>allow <code class="docutils literal notranslate"><span class="pre">sel.list</span></code> to be used for UpSet plots and fix some typos <a class="reference external" href="https://github.com/lcdb/lcdb-wf/pull/205">#205</a></p></li>
<li><p>workaround for degPatterns for corner cases where there are few clusters because of the <code class="docutils literal notranslate"><span class="pre">minc</span></code> parameter (<a class="reference external" href="https://github.com/lcdb/lcdb-wf/pull/205">#205</a>)</p></li>
<li><p>alpha and lfc.thresh are now pulled out into a separate chunk (<a class="reference external" href="https://github.com/lcdb/lcdb-wf/pull/206">#206</a>)</p></li>
<li><p>Support AnnotationHub http proxy handling in new version of AnnotationHub (<a class="reference external" href="https://github.com/lcdb/lcdb-wf/pull/207">#207</a>).</p></li>
</ul>
<p>As well as the following changes to other parts of the RNA-seq workflow, such as:</p>
<ul class="simple">
<li><p>better bigWig file nomenclature (<a class="reference external" href="https://github.com/lcdb/lcdb-wf/pull/194/files">#194</a>), uses “pos” and “neg”.</p></li>
<li><p>featureCounts only runs once on all BAMs rather than individual samples (<a class="reference external" href="https://github.com/lcdb/lcdb-wf/pull/195">#195</a>)</p></li>
<li><p>support <cite>rseqc infer_experiment</cite>, which replaces running featureCounts in multiple stranded modes (<a class="reference external" href="https://github.com/lcdb/lcdb-wf/pull/199">#199</a>, <a class="reference external" href="https://github.com/lcdb/lcdb-wf/pull/203">#203</a>)</p></li>
<li><p>use <code class="docutils literal notranslate"><span class="pre">--validateMappings</span></code> for salmon (<a class="reference external" href="https://github.com/lcdb/lcdb-wf/pull/203">#203</a>)</p></li>
</ul>
</section>
<section id="id33">
<h3>References<a class="headerlink" href="#id33" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p>fix typo in <em>S. pombe</em> name</p></li>
</ul>
</section>
<section id="all-workflows">
<h3>All workflows<a class="headerlink" href="#all-workflows" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p>Documentation now recommends creating an environment for each directory using the <cite>-p</cite> argument (<a class="reference external" href="https://github.com/lcdb/lcdb-wf/pull/195">#195</a>)</p></li>
</ul>
</section>
</section>
<section id="v1-4-2-jul-2019">
<h2>v1.4.2 (Jul 2019)<a class="headerlink" href="#v1-4-2-jul-2019" title="Permalink to this headline">¶</a></h2>
<section id="bugfixes">
<h3>Bugfixes<a class="headerlink" href="#bugfixes" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p>Don’t require ChIP-seq configs to have at least one block for each supported
peak-caller</p></li>
</ul>
</section>
</section>
<section id="v1-4-1-jul-2019">
<h2>v1.4.1 (Jul 2019)<a class="headerlink" href="#v1-4-1-jul-2019" title="Permalink to this headline">¶</a></h2>
<section id="id35">
<h3>RNA-seq<a class="headerlink" href="#id35" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p>KEGG results were not being added to the <code class="docutils literal notranslate"><span class="pre">all.enrich</span></code> list in <code class="docutils literal notranslate"><span class="pre">rnaseq.Rmd</span></code></p></li>
<li><p>symlinking bigWigs is now a local rule</p></li>
<li><p>default cutadapt options have changed to reflect current recommendations from
the author, and the cutadapt rule is now explicity using arguments rather
than requiring a separate <code class="docutils literal notranslate"><span class="pre">adapters.fa</span></code> file.</p></li>
<li><p>featureCounts now auto-detects whether it should be run with the <code class="docutils literal notranslate"><span class="pre">-p</span></code>
argument in paired-end mode (previously it was up to the user to make sure
this was added). The rule does have an override if this behavior is not wanted.</p></li>
</ul>
</section>
<section id="id36">
<h3>References<a class="headerlink" href="#id36" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p>The reference config for <em>Drosophila</em> is now fixed. Previously it depended on
<cite>chrom_convert</cite>. That script was a fly-specific script in lcdblib, but
lcdblib is no longer a dependency since v1.3. This fix uses the
<cite>convert_fastq_chroms</cite> and <cite>convert_gtf_chroms</cite> used in reference configs for
other species.</p></li>
</ul>
</section>
</section>
<section id="v1-4-may-2019">
<h2>v1.4 (May 2019)<a class="headerlink" href="#v1-4-may-2019" title="Permalink to this headline">¶</a></h2>
<section id="id37">
<h3>RNA-seq<a class="headerlink" href="#id37" title="Permalink to this headline">¶</a></h3>
<p>Much-improved <code class="docutils literal notranslate"><span class="pre">rnaseq.Rmd</span></code>:</p>
<ul class="simple">
<li><p>tabbed PCA plot</p></li>
<li><p>improved DEGpatterns chunk</p></li>
<li><p>dramatically improved functional enrichment section, with tabbed clusterprofiler plots and exported data in two flavors (combined and split)</p></li>
<li><p>improved upset plots, with exported files showing sets of genes</p></li>
<li><p>improved comments to highlight where to make changes</p></li>
<li><dl class="simple">
<dt>add new helper functions to <code class="docutils literal notranslate"><span class="pre">helpers.R</span></code>:</dt><dd><ul>
<li><p><code class="docutils literal notranslate"><span class="pre">fromList.with.names</span></code>, for getting UpSet plot output</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">rownames.first.col</span></code>, to make tidier dataframes</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">nested.lapply</span></code>, for convenient 2-level nested list apply</p></li>
<li><p>clusterprofiler helper functions</p></li>
</ul>
</dd>
</dl>
</li>
</ul>
</section>
</section>
<section id="v1-3-may-2019">
<h2>v1.3 (May 2019)<a class="headerlink" href="#v1-3-may-2019" title="Permalink to this headline">¶</a></h2>
<section id="id38">
<h3>Bugfixes<a class="headerlink" href="#id38" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p>Fix broken paired-end support for RNA-seq. Previously, when using data from
elsewhere on disk and using the symlink rules, R2 would be symlinked to the
same file as R1.</p></li>
<li><p>Support for Snakemake 5.4.0 which changes behavior of the <code class="docutils literal notranslate"><span class="pre">expand()</span></code>
function.</p></li>
</ul>
</section>
<section id="infrastructure">
<h3>Infrastructure<a class="headerlink" href="#infrastructure" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p>new deploy script to copy over only the files necessary for an analysis,
avoiding the clutter of testing infrastructure.</p></li>
<li><p>lcdblib, an external package, is no longer a dependency. In the interest of
better transparency and to make the code here easier to follow, the relevant
code from lcdblib was copied over to the <code class="docutils literal notranslate"><span class="pre">lib</span></code> directory in this
repository.</p></li>
</ul>
</section>
<section id="id39">
<h3>ChIP-seq and RNA-seq<a class="headerlink" href="#id39" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p>Bowtie2, HISAT2, and rRNA rules no longer use wrappers. This makes it easier
to track down what parameters are being used in each rule.</p></li>
<li><p>RSeQC is now available in Python 3 so wrappers have been removed.</p></li>
<li><p>NextGenMap support removed</p></li>
</ul>
</section>
</section>
<section id="v1-2-mar-2019">
<h2>v1.2 (Mar 2019)<a class="headerlink" href="#v1-2-mar-2019" title="Permalink to this headline">¶</a></h2>
<section id="id40">
<h3>RNA-seq<a class="headerlink" href="#id40" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p>First-class paired-end support, including mixing PE and SE samples in the
same sampletable</p></li>
<li><p>Support for STAR aligner</p></li>
</ul>
</section>
<section id="id41">
<h3>References<a class="headerlink" href="#id41" title="Permalink to this headline">¶</a></h3>
<ul>
<li><p>FASTA files are always symlinked into the directories of indexes that were
created from it</p></li>
<li><p>Reference configs:</p>
<blockquote>
<div><ul class="simple">
<li><p>updated existing</p></li>
<li><p>added more species</p></li>
<li><p>new post-process for fasta or gtf: you can now use
NICHD-BSPC/chrom-name-mappings to convert chromosome names between UCSC
and Ensembl (see reference configs for examples of use)</p></li>
</ul>
</div></blockquote>
</li>
</ul>
</section>
<section id="id42">
<h3>ChIP-seq and RNA-seq<a class="headerlink" href="#id42" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p>Updates to dependencies and MultiQC config</p></li>
</ul>
</section>
<section id="id43">
<h3>Infrastructure<a class="headerlink" href="#id43" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p>Updated requirements in <code class="docutils literal notranslate"><span class="pre">requirements.txt</span></code> and in wrappers</p></li>
<li><p>Changed all <code class="docutils literal notranslate"><span class="pre">pd.read_table()</span></code> to <code class="docutils literal notranslate"><span class="pre">pd.read_csv(sep=&quot;\t&quot;)</span></code> to prevent warnings</p></li>
<li><p>Changed all <code class="docutils literal notranslate"><span class="pre">yaml.load()</span></code> to <code class="docutils literal notranslate"><span class="pre">yaml.load(Loader=yaml.FullLoader)</span></code> to
prevent warnings</p></li>
<li><p>Using DeprecationWarning rather than UserWarning in the deprecation handler
so there’s less spam in the logs</p></li>
<li><p>Improved tests:</p>
<ul>
<li><p>using data from pybedtools repo because modENCODE seems to be down</p></li>
<li><p>append rather than prepend base conda to PATH on circleci</p></li>
<li><p>separate isolated tests for STAR, ngm, and SRA</p></li>
<li><p>updated conda</p></li>
</ul>
</li>
<li><p>Docs additions:</p>
<ul>
<li><p>TMPDIR handling</p></li>
<li><p>clusterconfig</p></li>
<li><p>WRAPPER_SLURM</p></li>
<li><p>docs for developers</p></li>
<li><p>symlinking fastqs</p></li>
<li><p>using SRA sampletables</p></li>
<li><p>paired-end data</p></li>
</ul>
</li>
</ul>
</section>
<section id="colocalization">
<h3>Colocalization<a class="headerlink" href="#colocalization" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p>From colocalization, removed the GAT “fractions” heatmap due to unresolved
pandas index errors</p></li>
</ul>
</section>
</section>
<section id="v1-1-aug-2018">
<h2>v1.1 (Aug 2018)<a class="headerlink" href="#v1-1-aug-2018" title="Permalink to this headline">¶</a></h2>
<section id="id44">
<h3>Infrastructure<a class="headerlink" href="#id44" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p>The default settings in Snakefiles are for real-world use, rather than for
testing. This reduces the amount of editing necessary before running actual
data. See <a class="reference internal" href="tests.html#test-settings"><span class="std std-ref">A note about test settings</span></a> for the extra step to take when testing
locally.</p></li>
<li><p>new <code class="docutils literal notranslate"><span class="pre">run_test.sh</span></code> script in each workflow directory to automatically run
the preprocessor when running test data</p></li>
<li><p>added extensive comments to Snakefiles with <code class="docutils literal notranslate"><span class="pre">NOTE:</span></code> string to make it
obvious where and how to make changes.</p></li>
<li><p>Documentation overhaul to bring everything up to v1.1. This includes Sphinx
autodocs on the <code class="docutils literal notranslate"><span class="pre">lib</span></code> module.</p></li>
<li><p>pytest test suite is run on the <code class="docutils literal notranslate"><span class="pre">lib</span></code> module</p></li>
</ul>
</section>
<section id="id45">
<h3>References<a class="headerlink" href="#id45" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p>new <cite>metadata</cite> section in references config, which can be used to store
additional information like mappable bases and genome size.</p></li>
<li><p>References can now be included from other YAML files into the main config
file. This dramatically simplifies individual configfiles, and allows
multiple workflows to use identical references without having to do
error-prone and hard-to-maintain copy/pastes between workflow configs. See
<a class="reference internal" href="references-config.html#references-config"><span class="std std-ref">References config</span></a> for details.</p></li>
<li><p>New GTF conversion, <code class="docutils literal notranslate"><span class="pre">mappings</span></code>. This is intended to replace the
<code class="docutils literal notranslate"><span class="pre">annotation_hub</span></code> conversion, which was problematic because 1) a particular
annotation hub accession is not guaranteed to be found in new versions of
AnnotationHub, resulting in lack of reproducibility, and 2) it was difficult
to synchronize the results with a particular GTF annotation. The
<code class="docutils literal notranslate"><span class="pre">annotation_hub</span></code> conversion is still supported, but if it’s used then
a DeprecationWarning will be emitted, recommending <code class="docutils literal notranslate"><span class="pre">mappings</span></code> instead.</p></li>
</ul>
</section>
<section id="both-rna-seq-and-chip-seq">
<h3>Both RNA-seq and ChIP-seq<a class="headerlink" href="#both-rna-seq-and-chip-seq" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p><cite>fastq_screen</cite> is now configured via <code class="docutils literal notranslate"><span class="pre">config.yaml</span></code>. This reduces the need
to edit the Snakefile and coordinate between the config and the fastq_screen
rule. Now everything is done within the config file.</p></li>
<li><p><cite>fastq_screen</cite> wrapper now handles additional output files created when using
the <code class="docutils literal notranslate"><span class="pre">--tag</span></code> and <code class="docutils literal notranslate"><span class="pre">--filter</span></code> arguments to <code class="docutils literal notranslate"><span class="pre">fastq_screen</span></code>.</p></li>
<li><p>In the config file, <code class="docutils literal notranslate"><span class="pre">assembly</span></code> has been changed to the more-descriptive
<code class="docutils literal notranslate"><span class="pre">organism</span></code>. The change is backwards compatible, but a DeprecationWarning is
raised if <code class="docutils literal notranslate"><span class="pre">assembly:</span></code> is still used, and changed to <code class="docutils literal notranslate"><span class="pre">organism</span></code> (though
only in memory, not on disk).</p></li>
<li><p>Patterns no longer use <code class="docutils literal notranslate"><span class="pre">{sample_dir}</span></code>, <code class="docutils literal notranslate"><span class="pre">{agg_dir}</span></code>, etc placeholders that
need to be configured in the config YAML. Instead, these directories are
hard-coded directly into the patterns. This simplifies the config files,
simplifies the patterns, and removes one layer of disconnect between the
filenames and how they are determined.</p></li>
<li><p>removed 4C workflow since it used 4c-ker</p></li>
</ul>
</section>
<section id="id46">
<h3>ChIP-seq<a class="headerlink" href="#id46" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p>macs2 and sicer can accept mappable genome size overrides</p></li>
</ul>
</section>
<section id="id47">
<h3>RNA-seq<a class="headerlink" href="#id47" title="Permalink to this headline">¶</a></h3>
<ul>
<li><p>RNA-seq downstream:</p>
<blockquote>
<div><ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">downstream/help_docs.Rmd</span></code> can be included for first-time users to
describe the sections of the RNA-seq analysis</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">rnaseq.Rmd</span></code> now uses the same <code class="docutils literal notranslate"><span class="pre">NOTE:</span></code> syntax as the Snakefiles for
indicating where/what to change</p></li>
<li><p>Easy swapping of which strand to use from the three featureCounts runs
performed by the workflow</p></li>
<li><p>Be explicit about using DESeq2::lfcShrink as is now the default in recent
DESeq2 versions</p></li>
<li><p>improved the mechanism for keeping together results objects, dds objects, and
labels (list of lists, rather than individual list object; refactored
functions to use this new structure</p></li>
</ul>
</div></blockquote>
</li>
</ul>
</section>
</section>
<section id="v1-0-1-jun-2018">
<h2>v1.0.1 (Jun 2018)<a class="headerlink" href="#v1-0-1-jun-2018" title="Permalink to this headline">¶</a></h2>
<p>Bugfixes, last release before references changes.</p>
<section id="id48">
<h3>Infrastructure<a class="headerlink" href="#id48" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p>Transition to CircleCI for testing</p></li>
<li><p>Use production settings by default; see <a class="reference internal" href="tests.html#test-settings"><span class="std std-ref">A note about test settings</span></a> for
more.</p></li>
<li><p>lots o’ docs</p></li>
<li><p>new <code class="docutils literal notranslate"><span class="pre">include/references_configs</span></code> to help organize references. These are
currently not used by the workflows directly.</p></li>
<li><p>bugfix: use additional options when uncompressing downloaded reference files
(<code class="docutils literal notranslate"><span class="pre">--no-same-owner</span></code> for <code class="docutils literal notranslate"><span class="pre">tar</span></code>, <code class="docutils literal notranslate"><span class="pre">-f</span></code> for <code class="docutils literal notranslate"><span class="pre">gunzip</span></code>)</p></li>
<li><p>additional dependencies in the top-level environment to support the
additional features in rnaseq.Rmd and track hubs.</p></li>
<li><p>colocalization workflow, external workflow, figures workflow to demonstrate
vertical integration</p></li>
</ul>
</section>
<section id="id49">
<h3>RNA-seq<a class="headerlink" href="#id49" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p>remove kallisto indexing, use salmon</p></li>
<li><p>improvements to how chipseq sampletables are parsed (with more informative
error messages)</p></li>
<li><p>run preseq for RNA-seq library complexity QC</p></li>
<li><p>support for merging bigwigs</p></li>
<li><p>featureCounts is now run in all three strandedness modes, and results
incorporated into MultiQC as separate modules.</p></li>
<li><p>RNA-seq now symlinks “pos” and “neg” bigWigs, which describe how reads map to
the <em>reference</em>, to “sense” and “antisense” bigWigs, which describe the
<em>originating RNA</em>. This makes it easy to swap strands depending on protocol.</p></li>
<li><p>new <code class="docutils literal notranslate"><span class="pre">downstream/helpers.Rmd</span></code> which factors out a lot of the work previously
done in <code class="docutils literal notranslate"><span class="pre">rnaseq.Rmd</span></code> into separate functions.</p></li>
<li><p>track hub building respects new sense/antisense bigwig symlinks</p></li>
</ul>
</section>
<section id="downstream-rnaseq-rmd">
<h3><code class="docutils literal notranslate"><span class="pre">downstream/rnaseq.Rmd</span></code><a class="headerlink" href="#downstream-rnaseq-rmd" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p>AnnotationHub uses cache dir that will not clobber default home directory cache</p></li>
<li><p>use varianceStabilizingTransform instead of rlog</p></li>
<li><p>print a size factors table</p></li>
<li><p>use multiple cores for computationally expensive DESeq2 operations</p></li>
<li><p>using separate lists for results, dds objects, and nice labels for automated
plots for each contrast</p></li>
<li><p>UpSet plots for comparing gene lists across contrasts</p></li>
<li><p>DEGpattern plots for showing clusters of expression patterns (from the
DEGreport package)</p></li>
<li><p>attach normalized counts per sample and per factor (parsed from the model
used for the contrast) as well as TPM estimates to the results tables</p></li>
<li><p>trim the labels in GO enrichment plots when too long</p></li>
</ul>
</section>
<section id="id50">
<h3>ChIP-seq<a class="headerlink" href="#id50" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p>sicer for chipseq domain calling</p></li>
<li><p>pin snakemake &lt;4.5.0 so that subworkflows behave correctly</p></li>
<li><p>chipseq peak-calling rules (and therefore wrappers) now expect a chromsizes
file as input</p></li>
<li><p>bigbed files for narrowPeak and broadPeak files are created correctly
depending on their format</p></li>
<li><p>run multiBigWigSummary and plotCorrelation from deepTools for ChIP-seq QC</p></li>
<li><p>ChIP-seq track hub generation script</p></li>
</ul>
</section>
<section id="id51">
<h3>Both RNA-seq and ChIP-seq<a class="headerlink" href="#id51" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p>update deeptools calls to reflect &gt;v3.0 syntax</p></li>
<li><p>support for SRA run tables so it’s trivial to re-run experiments
in SRA</p></li>
<li><p>multiple FastQC runs are shown separately in MultiQC output</p></li>
</ul>
</section>
</section>
<section id="v1-0-may-2018">
<h2>v1.0 (May 2018)<a class="headerlink" href="#v1-0-may-2018" title="Permalink to this headline">¶</a></h2>
<p>First official full release.</p>
</section>
</section>


          </div>
          
        </div>
      </div>
    <div class="clearer"></div>
  </div>
    <div class="footer">
      &copy;2017, Ryan Dale, Justin Fear.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 4.0.2</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.12</a>
      
      |
      <a href="_sources/changelog.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>