<!DOCTYPE html>

<html lang="en" data-content_root="./">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>conda and conda envs in lcdb-wf &#8212; lcdb-wf 1.9 documentation</title>
    <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=fa44fd50" />
    <link rel="stylesheet" type="text/css" href="_static/basic.css?v=b08954a9" />
    <link rel="stylesheet" type="text/css" href="_static/alabaster.css?v=9735eac1" />
    <script src="_static/documentation_options.js?v=754f84b7"></script>
    <script src="_static/doctools.js?v=9bcbadda"></script>
    <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Testing the installation" href="tests.html" />
    <link rel="prev" title="Integrative workflows" href="integrative.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  

  
  

  </head><body>
  <div class="document">
    
      <div class="sphinxsidebar" role="navigation" aria-label="Main">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="toc.html">lcdb-wf</a></h1>



<p class="blurb">Customizable workflows for high-throughput sequencing analysis</p>






<h3>Navigation</h3>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="index.html">Introduction</a></li>
<li class="toctree-l1"><a class="reference internal" href="getting-started.html">Getting started</a></li>
<li class="toctree-l1"><a class="reference internal" href="guide.html">Guide to file hierarchy</a></li>
<li class="toctree-l1"><a class="reference internal" href="workflows.html">Overview of workflows</a></li>
<li class="toctree-l1"><a class="reference internal" href="config.html">Configuration</a></li>
<li class="toctree-l1"><a class="reference internal" href="references.html">References workflow</a></li>
<li class="toctree-l1"><a class="reference internal" href="rnaseq.html">RNA-seq workflow</a></li>
<li class="toctree-l1"><a class="reference internal" href="downstream-rnaseq.html">RNA-Seq downstream analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="chipseq.html">ChIP-seq workflow</a></li>
<li class="toctree-l1"><a class="reference internal" href="integrative.html">Integrative workflows</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">conda and conda envs in <cite>lcdb-wf</cite></a><ul>
<li class="toctree-l2"><a class="reference internal" href="#conda-basics">Conda basics</a></li>
<li class="toctree-l2"><a class="reference internal" href="#building-the-environments">Building the environments</a></li>
<li class="toctree-l2"><a class="reference internal" href="#troubleshooting-environments">Troubleshooting environments</a></li>
<li class="toctree-l2"><a class="reference internal" href="#conda-envs-in-lcdb-wf">Conda envs in lcdb-wf</a></li>
<li class="toctree-l2"><a class="reference internal" href="#design-decisions">Design decisions</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="tests.html">Testing the installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="faqs.html">FAQs</a></li>
<li class="toctree-l1"><a class="reference internal" href="changelog.html">Changelog</a></li>
<li class="toctree-l1"><a class="reference internal" href="developers.html">For Developers</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="toc.html">Documentation overview</a><ul>
      <li>Previous: <a href="integrative.html" title="previous chapter">Integrative workflows</a></li>
      <li>Next: <a href="tests.html" title="next chapter">Testing the installation</a></li>
  </ul></li>
</ul>
</div>
<search id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" autocomplete="off" autocorrect="off" autocapitalize="off" spellcheck="false"/>
      <input type="submit" value="Go" />
    </form>
    </div>
</search>
<script>document.getElementById('searchbox').style.display = "block"</script>
        </div>
      </div>
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <section id="conda-and-conda-envs-in-lcdb-wf">
<span id="conda-envs"></span><h1>conda and conda envs in <cite>lcdb-wf</cite><a class="headerlink" href="#conda-and-conda-envs-in-lcdb-wf" title="Link to this heading">¶</a></h1>
<section id="conda-basics">
<h2>Conda basics<a class="headerlink" href="#conda-basics" title="Link to this heading">¶</a></h2>
<p>If you’re not familiar with <code class="docutils literal notranslate"><span class="pre">conda</span></code>, it is a way of keeping software isolated
on a computer in an “environment” (basically a directory with the executables
for all the software you want to use). When you “activate” the environment, it
places that location at the beginning of your <code class="docutils literal notranslate"><span class="pre">$PATH</span></code> variable, so that any
executables there are found first. It does not affect any existing installation
of any software on your machine and does not need root privileges.</p>
<p>If you don’t already have conda installed and the Bioconda channel set up, see
the <a class="reference external" href="https://bioconda.github.io">Bioconda docs</a> for details.</p>
<p>You’ll also probably want <a class="reference external" href="https://github.com/mamba-org/mamba">mamba</a>. Mamba
is a drop-in replacement for conda that is faster and more robust. In fact, it
is now the default conda front-end for Snakemake. If you don’t already have
mamba, you can install it into your base conda environment with:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>conda<span class="w"> </span>install<span class="w"> </span>-n<span class="w"> </span>base<span class="w"> </span>-c<span class="w"> </span>conda-forge<span class="w"> </span>mamba
</pre></div>
</div>
<p>It’s recommended that you install mamba into the base env (just like conda
itself is) so that it behaves like conda. It does <em>not</em> need to be installed
into each individual environment.</p>
</section>
<section id="building-the-environments">
<h2>Building the environments<a class="headerlink" href="#building-the-environments" title="Link to this heading">¶</a></h2>
<p><strong>It is recommended that you create a separate environment directory for
each project</strong>, rather than a single environment for all projects. That way you
can update packages in each project independently of any others, and yet the
environment will always be close at hand. This is an especially good practice
in shared space as others can easily find and activate the environment specific
to the project.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>We recommend using mamba rather than conda for the speed increase and
ability to more correctly solve environments. See the <a class="reference external" href="https://snakemake.readthedocs.io/en/stable/getting_started/installation.html#installation-via-conda">snakemake docs</a>
for more info.</p>
</div>
<p>If you use the <code class="docutils literal notranslate"><span class="pre">--build-envs</span></code> argument when deploying lcdb-wf to a project
directory (see <a class="reference internal" href="getting-started.html#setup-proj"><span class="std std-ref">Setting up a project</span></a>), two conda environments will be built in the
directories: <code class="docutils literal notranslate"><span class="pre">env</span></code>, which has all of the non-R requirements, and <code class="docutils literal notranslate"><span class="pre">env-r</span></code>
which has the R packages used in particular for downstream RNA-seq analysis.
These environments will use the fully-pinned environments in <code class="docutils literal notranslate"><span class="pre">env.yml</span></code> and
<code class="docutils literal notranslate"><span class="pre">env-r.yml</span></code>. If you’ve already deployed but didn’t use the <code class="docutils literal notranslate"><span class="pre">--build-envs</span></code>
argument, then then the equivalent command to run in the deployed directory is:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>mamba<span class="w"> </span>env<span class="w"> </span>create<span class="w"> </span>-p<span class="w"> </span>./env<span class="w"> </span>--file<span class="w"> </span>env.yml
mamba<span class="w"> </span>env<span class="w"> </span>create<span class="w"> </span>-p<span class="w"> </span>./env-r<span class="w"> </span>--file<span class="w"> </span>env-r.yml
</pre></div>
</div>
</section>
<section id="troubleshooting-environments">
<span id="conda-troubleshooting"></span><h2>Troubleshooting environments<a class="headerlink" href="#troubleshooting-environments" title="Link to this heading">¶</a></h2>
<p>Sometimes there is a problem with creating an environment. For example, the
exact package specified in the env yaml might not be available for some reason
(this should not happen, but in practice sometimes it does in corner cases).</p>
<p>If this happens, you can try a couple things.</p>
<p>First, some terminology with how packages are specified in the environment
yamls. Here’s an example for <code class="docutils literal notranslate"><span class="pre">libpng</span></code> version 1.6.37:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">libpng</span><span class="o">=</span><span class="mf">1.6.37</span><span class="o">=</span><span class="n">hed695b0_2</span>
<span class="o">|</span><span class="n">____</span><span class="o">|</span> <span class="o">|</span><span class="n">____</span><span class="o">|</span> <span class="o">|</span><span class="n">________</span><span class="o">|</span>
 <span class="o">|</span>       <span class="o">|</span>       <span class="o">|</span>
<span class="n">name</span>     <span class="o">|</span>       <span class="o">|</span>
       <span class="n">version</span>   <span class="o">|</span>
               <span class="n">build</span> <span class="n">string</span>
</pre></div>
</div>
<p>The package name (libpng) and version (1.6.37) are pretty standard and
self-explanatory. The <cite>build</cite> string refers to different built versions of the
<em>conda package</em>, but for the same version (1.6.37 in this case) of the package.
For example, if a conda package was built for version 1.1 of a tool, but that
package itself had an error unrelated to the tool, then a fixed build would be
made. The package version would remain the same (1.1) but the build string
would change.</p>
<p>In this example, the build string contains a hash <code class="docutils literal notranslate"><span class="pre">hed695b0</span></code> which is a hash
of all the pinned dependencies for this package at packaging time. The
<a class="reference external" href="https://conda-forge.org/docs/maintainer/pinning_deps.html">conda-forge pinning docs</a> give more detail
on what this pinning is about, but basically if that pinning changes then this
hash will change. The <code class="docutils literal notranslate"><span class="pre">_2</span></code> on the end of the build string hash indicates that
this is the third built package (build numbers start at zero) for this version
of <code class="docutils literal notranslate"><span class="pre">libpng</span></code> using the same pinning. In other words, there also likely exists
<code class="docutils literal notranslate"><span class="pre">libpng=1.6.37=hed695b0_1</span></code> and <code class="docutils literal notranslate"><span class="pre">libpng=1.6.37=hed695b0_0</span></code>. At the time of
this writing, there is also <code class="docutils literal notranslate"><span class="pre">libpng-1.6.37-h21135ba_2</span></code> (notice the different
hash) which is the same libpng version but uses different pinnings.</p>
<p>What does this mean for troubleshooting?</p>
<p>For any package that seems to be problematic, try editing the respective
environment yaml (e.g., <code class="docutils literal notranslate"><span class="pre">env.yml</span></code>) to remove the build string (so in the
example above, you would try changing it to just <code class="docutils literal notranslate"><span class="pre">libpng=1.6.37</span></code>) and try
building the environment again. If that doesn’t work, try removing the version
as well (so just <code class="docutils literal notranslate"><span class="pre">libpng</span></code>).</p>
<p>Alternatively for very problematic cases or cases where there are multiple
problematic packages, you can try creating an environment with the “loose”
pinning in <code class="docutils literal notranslate"><span class="pre">include/requirements.txt</span></code> which effectively does not require any
particular versions with the exception of a few corner cases. Keep in mind that
using that file may cause the environment to take a while to build as conda (or
mamba) solves the dependencies of all the specified packages.</p>
</section>
<section id="conda-envs-in-lcdb-wf">
<h2>Conda envs in lcdb-wf<a class="headerlink" href="#conda-envs-in-lcdb-wf" title="Link to this heading">¶</a></h2>
<p>Given all of the software used across all of <cite>lcdb-wf</cite>, the environments can
take a lot of time to build because the solver needs to figure out the entire
dependency tree and come up with a solution that works to satisfy the entire
set of specified requirements.</p>
<p>We chose to split the conda environments in two: the <strong>main</strong> environment and the <strong>R</strong>
environment (see <a class="reference internal" href="#conda-design-decisions"><span class="std std-ref">Design decisions</span></a>). These environments are
described by both “strict” and “loose” files. By default we use the “strict”
version, which pins all versions of all packages exactly. This is preferred
wherever possible. However we also provide a “loose” version that is not
specific about versions. The following table describes these files:</p>
<table class="docutils align-default">
<thead>
<tr class="row-odd"><th class="head"><p>strict version</p></th>
<th class="head"><p>loose version</p></th>
<th class="head"><p>used for</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">env.yml</span></code></p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">include/requirements.txt</span></code></p></td>
<td><p>Main Snakefiles</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">env-r.yaml</span></code></p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">include/requirements-r.txt</span></code></p></td>
<td><p>Downstream RNA-seq analysis in R</p></td>
</tr>
</tbody>
</table>
<p>When deploying new instances, use the <code class="docutils literal notranslate"><span class="pre">--build-envs</span></code> argument which will use
the strict version. Or use the following commands in a deployed directory:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>mamba<span class="w"> </span>env<span class="w"> </span>create<span class="w"> </span>-p<span class="w"> </span>./env<span class="w"> </span>--file<span class="w"> </span>env.yml
mamba<span class="w"> </span>env<span class="w"> </span>create<span class="w"> </span>-p<span class="w"> </span>./env-r<span class="w"> </span>--file<span class="w"> </span>env-r.yml
</pre></div>
</div>
<p>When getting ready to release a new lcdb-wf version, create a new environment
using the loose version to prepare the env and then when tests pass, export it
to yaml. That is:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="c1"># use loose version when preparing a new version of lcdb-wf</span>
mamba<span class="w"> </span>create<span class="w"> </span>-p<span class="w"> </span>./env<span class="w"> </span>--file<span class="w"> </span>include/requirements.txt
mamba<span class="w"> </span>create<span class="w"> </span>-p<span class="w"> </span>./env-r<span class="w"> </span>--file<span class="w"> </span>include/requirements-r.txt

<span class="c1"># then do testing....</span>

<span class="c1"># when tests pass, export the envs</span>
conda<span class="w"> </span>env<span class="w"> </span><span class="nb">export</span><span class="w"> </span>-p<span class="w"> </span>./env<span class="w"> </span>&gt;<span class="w"> </span>env.yml
conda<span class="w"> </span>env<span class="w"> </span><span class="nb">export</span><span class="w"> </span>-p<span class="w"> </span>./env-r<span class="w"> </span>&gt;<span class="w"> </span>env-r.yaml

<span class="c1"># commit, push, finalize release</span>
</pre></div>
</div>
</section>
<section id="design-decisions">
<span id="conda-design-decisions"></span><h2>Design decisions<a class="headerlink" href="#design-decisions" title="Link to this heading">¶</a></h2>
<p>We made the design decision to split the conda envs into two different
environments – one for R, one for non-R. We found that by by removing the
entire sub-DAG of R packages from the main environment we can dramatically
reduce the creation time.</p>
<p>We also made the decision to use large top-level environments rather than
smaller environments created for each rule using the <code class="docutils literal notranslate"><span class="pre">conda:</span></code> directive.
There are two reasons for this choice. First, it allows us to activate a single
environment to give us access to all the tools used. This streamlines
troubleshooting because we don’t have to dig through the <code class="docutils literal notranslate"><span class="pre">.snakemake/conda</span></code>
directory to figure out which hash corresponds to which file, but comes with
the up-front cost of creating the environment initially. Second, it simplifies
running the tests on CircleCI, allowing us to cache the env directories as
a whole to be re-used for multiple tests rather than caching the individual
.snakemake directories for each tested workflow.</p>
<p>Given that the conda and snakemake ecosystem are in flux, this may change in
the future to using small conda environments for each rule separately if it
turns out to be more beneficial to do so.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Prior to v1.7, we used requirements.txt files with loose pinning. Moving to
yaml files allows us the option of also installing pip packages if needed.
It also allows us to specify channels directly in the yaml file for
streamlined installation.</p>
<p>Using strictly-pinned yaml files that are consistently tested will
hopefully result in a more stable experience for users. For example, if you
happen to create an environment around the time of a new R/Bioconductor
release, the environment may not build correctly using a loose pinning.
Other transient issues in the packaging ecosystem can similarly cause
issues.</p>
</div>
</section>
</section>


          </div>
          
        </div>
      </div>
    <div class="clearer"></div>
  </div>
    <div class="footer">
      &#169;2017, Ryan Dale, Justin Fear.
      
      |
      Powered by <a href="https://www.sphinx-doc.org/">Sphinx 8.2.3</a>
      &amp; <a href="https://alabaster.readthedocs.io">Alabaster 1.0.0</a>
      
      |
      <a href="_sources/conda.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>