<!DOCTYPE html>

<html lang="en" data-content_root="./">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>Configuration &#8212; lcdb-wf 1.9 documentation</title>
    <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=03e43079" />
    <link rel="stylesheet" type="text/css" href="_static/basic.css?v=b08954a9" />
    <link rel="stylesheet" type="text/css" href="_static/alabaster.css?v=9735eac1" />
    <script src="_static/documentation_options.js?v=754f84b7"></script>
    <script src="_static/doctools.js?v=9bcbadda"></script>
    <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Config YAML" href="config-yaml.html" />
    <link rel="prev" title="Overview of workflows" href="workflows.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  

  
  

  </head><body>
  <div class="document">
    
      <div class="sphinxsidebar" role="navigation" aria-label="Main">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="toc.html">lcdb-wf</a></h1>



<p class="blurb">Customizable workflows for high-throughput sequencing analysis</p>






<h3>Navigation</h3>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="index.html">Introduction</a></li>
<li class="toctree-l1"><a class="reference internal" href="getting-started.html">Getting started</a></li>
<li class="toctree-l1"><a class="reference internal" href="guide.html">Guide to file hierarchy</a></li>
<li class="toctree-l1"><a class="reference internal" href="workflows.html">Overview of workflows</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Configuration</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#general-configuration">General configuration</a><ul>
<li class="toctree-l3"><a class="reference internal" href="config-yaml.html">Config YAML</a></li>
<li class="toctree-l3"><a class="reference internal" href="references-config.html">References config</a></li>
<li class="toctree-l3"><a class="reference internal" href="sampletable.html">Sample tables</a></li>
<li class="toctree-l3"><a class="reference internal" href="patterns-targets.html">Patterns and targets</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#running-on-a-cluster">Running on a cluster</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#setting-up-a-cluster-profile">Setting up a cluster profile</a></li>
<li class="toctree-l3"><a class="reference internal" href="#for-nih-s-biowulf-cluster">For NIH’s Biowulf cluster</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id1">Running on a cluster</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="references.html">References workflow</a></li>
<li class="toctree-l1"><a class="reference internal" href="rnaseq.html">RNA-seq workflow</a></li>
<li class="toctree-l1"><a class="reference internal" href="downstream-rnaseq.html">RNA-Seq downstream analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="chipseq.html">ChIP-seq workflow</a></li>
<li class="toctree-l1"><a class="reference internal" href="integrative.html">Integrative workflows</a></li>
<li class="toctree-l1"><a class="reference internal" href="conda.html">conda and conda envs in <cite>lcdb-wf</cite></a></li>
<li class="toctree-l1"><a class="reference internal" href="tests.html">Testing the installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="faqs.html">FAQs</a></li>
<li class="toctree-l1"><a class="reference internal" href="changelog.html">Changelog</a></li>
<li class="toctree-l1"><a class="reference internal" href="developers.html">For Developers</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="toc.html">Documentation overview</a><ul>
      <li>Previous: <a href="workflows.html" title="previous chapter">Overview of workflows</a></li>
      <li>Next: <a href="config-yaml.html" title="next chapter">Config YAML</a></li>
  </ul></li>
</ul>
</div>
<search id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" autocomplete="off" autocorrect="off" autocapitalize="off" spellcheck="false"/>
      <input type="submit" value="Go" />
    </form>
    </div>
</search>
<script>document.getElementById('searchbox').style.display = "block"</script>
        </div>
      </div>
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <section id="configuration">
<span id="config"></span><h1>Configuration<a class="headerlink" href="#configuration" title="Link to this heading">¶</a></h1>
<section id="general-configuration">
<h2>General configuration<a class="headerlink" href="#general-configuration" title="Link to this heading">¶</a></h2>
<p>The majority of the work in setting up a new project is in the configuration –
which samples to run, where the data files are located, which references are
needed, etc.</p>
<p><strong>The entry point for configuration</strong> is in the <code class="docutils literal notranslate"><span class="pre">config/config.yaml</span></code> file found
in each workflow directory. See <a class="reference internal" href="config-yaml.html#config-yaml"><span class="std std-ref">Config YAML</span></a> for more.</p>
<div class="toctree-wrapper compound">
<ul>
<li class="toctree-l1"><a class="reference internal" href="config-yaml.html">Config YAML</a><ul>
<li class="toctree-l2"><a class="reference internal" href="config-yaml.html#example-configs">Example configs</a></li>
<li class="toctree-l2"><a class="reference internal" href="config-yaml.html#field-descriptions">Field descriptions</a></li>
</ul>
</li>
</ul>
</div>
<p>The <strong>references section</strong> of the config file configures the genomes,
transcriptomes, and annotations to be used. See <a class="reference internal" href="references-config.html#references-config"><span class="std std-ref">References config</span></a> for more.</p>
<div class="toctree-wrapper compound">
<ul>
<li class="toctree-l1"><a class="reference internal" href="references-config.html">References config</a><ul>
<li class="toctree-l2"><a class="reference internal" href="references-config.html#specifying-a-references-directory">Specifying a references directory</a></li>
<li class="toctree-l2"><a class="reference internal" href="references-config.html#including-existing-reference-configs">Including existing reference configs</a></li>
<li class="toctree-l2"><a class="reference internal" href="references-config.html#overview">Overview</a></li>
<li class="toctree-l2"><a class="reference internal" href="references-config.html#example-references-config">Example references config</a></li>
<li class="toctree-l2"><a class="reference internal" href="references-config.html#post-processing">Post processing</a></li>
<li class="toctree-l2"><a class="reference internal" href="references-config.html#more-advanced-postprocessing">More advanced postprocessing</a></li>
<li class="toctree-l2"><a class="reference internal" href="references-config.html#locations-of-downloaded-and-post-processed-fasta-and-gtf-files">Locations of downloaded-and-post-processed FASTA and GTF files</a></li>
<li class="toctree-l2"><a class="reference internal" href="references-config.html#available-indexes-and-conversions">Available indexes and conversions</a></li>
</ul>
</li>
</ul>
</div>
<p>The <strong>sample table</strong>, lists sample IDs, filenames, and other metadata. Its path
is specified in the config file. See <a class="reference internal" href="sampletable.html#sampletable"><span class="std std-ref">Sample tables</span></a> for more.</p>
<div class="toctree-wrapper compound">
<ul>
<li class="toctree-l1"><a class="reference internal" href="sampletable.html">Sample tables</a><ul>
<li class="toctree-l2"><a class="reference internal" href="sampletable.html#rna-seq-sample-table">RNA-seq sample table</a></li>
<li class="toctree-l2"><a class="reference internal" href="sampletable.html#chip-seq-sample-table">ChIP-seq sample table</a></li>
</ul>
</li>
</ul>
</div>
<p>A <strong>patterns file</strong> only needs to be edited
if you’re doing custom work. It determines the patterns of files that will be
created by the workflow. See <a class="reference internal" href="patterns-targets.html#patterns-and-targets"><span class="std std-ref">Patterns and targets</span></a> for more.</p>
<div class="toctree-wrapper compound">
<ul>
<li class="toctree-l1"><a class="reference internal" href="patterns-targets.html">Patterns and targets</a><ul>
<li class="toctree-l2"><a class="reference internal" href="patterns-targets.html#patterns">Patterns</a></li>
<li class="toctree-l2"><a class="reference internal" href="patterns-targets.html#targets">Targets</a></li>
<li class="toctree-l2"><a class="reference internal" href="patterns-targets.html#how-patterns-and-targets-are-used-in-snakefiles">How patterns and targets are used in Snakefiles</a></li>
</ul>
</li>
</ul>
</div>
</section>
<section id="running-on-a-cluster">
<span id="cluster"></span><h2>Running on a cluster<a class="headerlink" href="#running-on-a-cluster" title="Link to this heading">¶</a></h2>
<p>The example commands in <a class="reference internal" href="getting-started.html#getting-started"><span class="std std-ref">Getting started</span></a> describe running Snakemake
locally. For larger data sets, you’ll want to run them on an HPC cluster.
Snakemake <a class="reference external" href="http://snakemake.readthedocs.io/en/latest/snakefiles/configuration.html">supports arbitrary cluster commands</a>,
making it easy to run these workflows on many different cluster environments.</p>
<p>Snakemake and these workflows are designed to decouple the code from the
configuration. Each rule has resources specified. When running with
a cluster-specific <a class="reference external" href="https://snakemake.readthedocs.io/en/stable/executing/cli.html#profiles">Snakemake profile</a>,
these resources are translated into cluster-specific commands.</p>
<section id="setting-up-a-cluster-profile">
<h3>Setting up a cluster profile<a class="headerlink" href="#setting-up-a-cluster-profile" title="Link to this heading">¶</a></h3>
<p>To run on a cluster, you need a Snakemake profile for your specific cluster
environment.</p>
<p>Typically this is set up by the cluster admins, but you can find some existing
profiles for common submission systems in the <a class="reference external" href="https://github.com/snakemake-profiles/doc">Snakemake-Profiles project</a>.</p>
<p>The profile, in turn, may require additional snakemake plugins to be
installed. These would need to go into the main environment; you can use the
<code class="docutils literal notranslate"><span class="pre">--additional-main</span></code> argument for <code class="file docutils literal notranslate"><span class="pre">deploy.py</span></code> if/when you know the
package in advance. Otherwise, install the required plugin into the main
environment.</p>
<p>Then set the <code class="docutils literal notranslate"><span class="pre">LCDBWF_SNAKEMAKE_PROFILE</span></code> and <code class="docutils literal notranslate"><span class="pre">LCDBWF_SNAKEMAKE_PROFILE_V8</span></code>
env vars to point to the &lt;8 and &gt;8 profiles respectively.</p>
</section>
<section id="for-nih-s-biowulf-cluster">
<h3>For NIH’s Biowulf cluster<a class="headerlink" href="#for-nih-s-biowulf-cluster" title="Link to this heading">¶</a></h3>
<p>On NIH’s Biowulf HPC cluster, use the <a class="reference external" href="https://github.com/NIH-HPC/snakemake_profile">Biowulf profile</a>. For recent versions
of lcdb-wf, you need the <code class="docutils literal notranslate"><span class="pre">snakemake8</span></code> branch of the profile:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>git<span class="w"> </span>clone<span class="w"> </span>https://github.com/NIH-HPC/snakemake_profile<span class="w"> </span>~/snakemake_profile_v8
<span class="nb">cd</span><span class="w"> </span>~/snakemake_profile_v8
git<span class="w"> </span>checkout<span class="w"> </span>snakemake8

<span class="c1"># Set the environment variable (add to your ~/.bashrc for persistence)</span>
<span class="nb">export</span><span class="w"> </span><span class="nv">LCDBWF_SNAKEMAKE_PROFILE_V8</span><span class="o">=</span>~/snakemake_profile_v8
</pre></div>
</div>
</section>
<section id="id1">
<h3>Running on a cluster<a class="headerlink" href="#id1" title="Link to this heading">¶</a></h3>
<p>In general, we should avoid running long-running Snakemake workflows on the
login node of a cluster. lcdb-wf comes with a wrapper script,
<code class="docutils literal notranslate"><span class="pre">include/WRAPPER_SLURM</span></code>, that runs Snakemake as a batch job on Slurm.</p>
<p>To run a workflow on a Slurm cluster, from the workflow directory (e.g.,
<code class="docutils literal notranslate"><span class="pre">workflows/rnaseq</span></code>), and with the main environment activated, run:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>sbatch<span class="w"> </span>../../include/WRAPPER_SLURM
</pre></div>
</div>
<p>The environment is inherited, and the version of Snakemake is detected so that
the appropriate profile can be used. Then Snakemake is run. Via the profile,
the main Snakemake process started by the initial WRAPPER_SLURM batch job will
automatically submit jobs to the cluster.</p>
<p>You can keep an eye on <code class="file docutils literal notranslate"><span class="pre">Snakefile.log</span></code> to watch progress (e.g. <cite>tail -f
Snakefile.log</cite>).</p>
<p>For other cluster environments, consult the Snakemake documentation on:
- <a class="reference external" href="https://snakemake.readthedocs.io/en/stable/executing/cluster.html">Cluster execution</a>
- <a class="reference external" href="https://snakemake.readthedocs.io/en/stable/executing/cloud.html">Cloud execution</a></p>
</section>
</section>
</section>


          </div>
          
        </div>
      </div>
    <div class="clearer"></div>
  </div>
    <div class="footer">
      &#169;2017, Ryan Dale, Justin Fear.
      
      |
      Powered by <a href="https://www.sphinx-doc.org/">Sphinx 8.2.3</a>
      &amp; <a href="https://alabaster.readthedocs.io">Alabaster 1.0.0</a>
      
      |
      <a href="_sources/config.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>