<!DOCTYPE html>

<html lang="en" data-content_root="./">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>RNA-Seq downstream analysis &#8212; lcdb-wf 1.9 documentation</title>
    <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=03e43079" />
    <link rel="stylesheet" type="text/css" href="_static/basic.css?v=b08954a9" />
    <link rel="stylesheet" type="text/css" href="_static/alabaster.css?v=9735eac1" />
    <script src="_static/documentation_options.js?v=754f84b7"></script>
    <script src="_static/doctools.js?v=9bcbadda"></script>
    <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Detailed documentation of RNA-Seq downstream" href="rnaseq-rmd.html" />
    <link rel="prev" title="RNA-seq workflow" href="rnaseq.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  

  
  

  </head><body>
  <div class="document">
    
      <div class="sphinxsidebar" role="navigation" aria-label="Main">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="toc.html">lcdb-wf</a></h1>



<p class="blurb">Customizable workflows for high-throughput sequencing analysis</p>






<h3>Navigation</h3>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="index.html">Introduction</a></li>
<li class="toctree-l1"><a class="reference internal" href="getting-started.html">Getting started</a></li>
<li class="toctree-l1"><a class="reference internal" href="guide.html">Guide to file hierarchy</a></li>
<li class="toctree-l1"><a class="reference internal" href="workflows.html">Overview of workflows</a></li>
<li class="toctree-l1"><a class="reference internal" href="config.html">Configuration</a></li>
<li class="toctree-l1"><a class="reference internal" href="references.html">References workflow</a></li>
<li class="toctree-l1"><a class="reference internal" href="rnaseq.html">RNA-seq workflow</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">RNA-Seq downstream analysis</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#how-to-use-this-code">How to use this code</a></li>
<li class="toctree-l2"><a class="reference internal" href="#more-details">More details</a><ul>
<li class="toctree-l3"><a class="reference internal" href="rnaseq-rmd.html">Detailed documentation of RNA-Seq downstream</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="chipseq.html">ChIP-seq workflow</a></li>
<li class="toctree-l1"><a class="reference internal" href="integrative.html">Integrative workflows</a></li>
<li class="toctree-l1"><a class="reference internal" href="conda.html">conda and conda envs in <cite>lcdb-wf</cite></a></li>
<li class="toctree-l1"><a class="reference internal" href="tests.html">Testing the installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="faqs.html">FAQs</a></li>
<li class="toctree-l1"><a class="reference internal" href="changelog.html">Changelog</a></li>
<li class="toctree-l1"><a class="reference internal" href="developers.html">For Developers</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="toc.html">Documentation overview</a><ul>
  <li><a href="rnaseq.html">RNA-seq workflow</a><ul>
      <li>Previous: <a href="rnaseq.html" title="previous chapter">RNA-seq workflow</a></li>
      <li>Next: <a href="rnaseq-rmd.html" title="next chapter">Detailed documentation of RNA-Seq downstream</a></li>
  </ul></li>
  </ul></li>
</ul>
</div>
<search id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" autocomplete="off" autocorrect="off" autocapitalize="off" spellcheck="false"/>
      <input type="submit" value="Go" />
    </form>
    </div>
</search>
<script>document.getElementById('searchbox').style.display = "block"</script>
        </div>
      </div>
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <section id="rna-seq-downstream-analysis">
<span id="downstream"></span><h1>RNA-Seq downstream analysis<a class="headerlink" href="#rna-seq-downstream-analysis" title="Link to this heading">¶</a></h1>
<p>In a typical RNA-seq analysis, it is relatively straightforward to go from raw
reads to read counts in features to importing them into R. After that however,
expression analysis gets a bit more complicated and highly depends on the
design of the experiment.</p>
<p>We attempted to strike the balance between simplicity – where as much
configuration as possible takes place via a config file – and flexibility
where the R code can be modified as needed depending on the project.</p>
<p>This file is <code class="docutils literal notranslate"><span class="pre">workflows/rnaseq/downstream/rnaseq.Rmd</span></code>. It uses a separate
conda environment that just has the R dependencies. It is rendered via
<code class="docutils literal notranslate"><span class="pre">knitr</span></code> to create an HTML file. The inputs for the rule are the featureCounts
output, the sample table, the <code class="docutils literal notranslate"><span class="pre">lib/lcdbwf</span></code> R package, and the Rmd.</p>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>This RMarkdown file is <strong>intended to be edited and customized per experiment</strong>.</p>
</div>
<section id="how-to-use-this-code">
<h2>How to use this code<a class="headerlink" href="#how-to-use-this-code" title="Link to this heading">¶</a></h2>
<ol class="arabic simple">
<li><p>Activate the <code class="docutils literal notranslate"><span class="pre">env-r</span></code> conda environment (created as part of setting up the
<cite>lcdb-wf</cite> deployment)</p></li>
<li><p>Edit the <code class="file docutils literal notranslate"><span class="pre">workflows/rnaseq/downstream/config.yaml</span></code> file. It is
heavily commented and should be self-explanatory.</p></li>
<li><p>Customize the contrasts you want to run (see below for details on this)</p></li>
<li><p>From the <code class="file docutils literal notranslate"><span class="pre">workflows/rnaseq/downstream</span></code> directory, run
<code class="docutils literal notranslate"><span class="pre">rmarkdown::render(&quot;rnaseq.Rmd&quot;)</span></code> to get <code class="file docutils literal notranslate"><span class="pre">rnaseq.html</span></code></p></li>
</ol>
<p>Here are some additional notes:</p>
<ul>
<li><p>Many of the code chunks have the <code class="docutils literal notranslate"><span class="pre">cache=TRUE</span></code> option to speed up
re-rendering and make iterative development quicker. When everything’s in
a final state, you may want to delete the <code class="docutils literal notranslate"><span class="pre">rnaseq_cache</span></code> directory and
re-run.</p></li>
<li><p>Many of the cached code chunks also specify a config argument. These config
items are taken from the <code class="file docutils literal notranslate"><span class="pre">config.yaml</span></code> file living alongside the
<code class="file docutils literal notranslate"><span class="pre">rnaseq.Rmd</span></code>. If a cached chunk specifies a config option, and the
value in the config file changes, the chunk will be re-run because its cache
is invalidated.</p></li>
<li><p>As with many analyses in R, the work is highly iterative. You may want to
consider using an interactive interpreter, either via the command line or
RStudio. To ensure that RStudio is using the same packages as the workflows,
you should set the <code class="docutils literal notranslate"><span class="pre">RSTUDIO_WHICH_R</span></code> environment variable.</p>
<p>The easiest way to do this is to activate the conda environment you’re using
for the analysis, then export the identified location of R to that variable:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="nb">source</span><span class="w"> </span>activate<span class="w"> </span>lcdb-wf
<span class="nb">export</span><span class="w"> </span><span class="nv">RSTUDIO_WHICH_R</span><span class="o">=</span><span class="k">$(</span>which<span class="w"> </span>R<span class="k">)</span>
</pre></div>
</div>
<p>On MacOS, you may additionally need the following:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>launchctl<span class="w"> </span>setenv<span class="w"> </span>RSTUDIO_WHICH_R<span class="w"> </span><span class="nv">$RSTUDIO_WHICH_R</span>
</pre></div>
</div>
<p>Then run RStudio, which should pick up the conda environment’s version of R and
which will already have packages like DESeq2 installed in the environment.</p>
</li>
</ul>
</section>
<section id="more-details">
<h2>More details<a class="headerlink" href="#more-details" title="Link to this heading">¶</a></h2>
<p>For more detailed documentation, see <a class="reference internal" href="rnaseq-rmd.html#downstream-detailed"><span class="std std-ref">Detailed documentation of RNA-Seq downstream</span></a>.</p>
<div class="toctree-wrapper compound">
<ul>
<li class="toctree-l1"><a class="reference internal" href="rnaseq-rmd.html">Detailed documentation of RNA-Seq downstream</a><ul>
<li class="toctree-l2"><a class="reference internal" href="rnaseq-rmd.html#global-options"><code class="docutils literal notranslate"><span class="pre">global_options</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="rnaseq-rmd.html#lcdbwf"><code class="docutils literal notranslate"><span class="pre">lcdbwf</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="rnaseq-rmd.html#config"><code class="docutils literal notranslate"><span class="pre">config</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="rnaseq-rmd.html#libraries"><code class="docutils literal notranslate"><span class="pre">libraries</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="rnaseq-rmd.html#coldata-setup"><code class="docutils literal notranslate"><span class="pre">coldata_setup</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="rnaseq-rmd.html#dds-initial"><code class="docutils literal notranslate"><span class="pre">dds_initial</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="rnaseq-rmd.html#print-coldata"><code class="docutils literal notranslate"><span class="pre">print_coldata</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="rnaseq-rmd.html#sample-heatmap"><code class="docutils literal notranslate"><span class="pre">sample_heatmap</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="rnaseq-rmd.html#pca"><code class="docutils literal notranslate"><span class="pre">pca</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="rnaseq-rmd.html#sizefactors"><code class="docutils literal notranslate"><span class="pre">sizefactors</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="rnaseq-rmd.html#dds-list"><code class="docutils literal notranslate"><span class="pre">dds_list</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="rnaseq-rmd.html#dds-diagnostics"><code class="docutils literal notranslate"><span class="pre">dds_diagnostics</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="rnaseq-rmd.html#results"><code class="docutils literal notranslate"><span class="pre">results_*</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="rnaseq-rmd.html#assemble-variables"><code class="docutils literal notranslate"><span class="pre">assemble_variables</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="rnaseq-rmd.html#summary"><code class="docutils literal notranslate"><span class="pre">summary</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="rnaseq-rmd.html#reportresults"><code class="docutils literal notranslate"><span class="pre">reportresults</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="rnaseq-rmd.html#upsetplots"><code class="docutils literal notranslate"><span class="pre">upsetplots</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="rnaseq-rmd.html#excel"><code class="docutils literal notranslate"><span class="pre">excel</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="rnaseq-rmd.html#write-output"><code class="docutils literal notranslate"><span class="pre">write_output</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="rnaseq-rmd.html#combined-rds"><code class="docutils literal notranslate"><span class="pre">combined_rds</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="rnaseq-rmd.html#sessioninfo"><code class="docutils literal notranslate"><span class="pre">sessioninfo</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="rnaseq-rmd.html#glossary">Glossary</a></li>
</ul>
</li>
</ul>
</div>
</section>
</section>


          </div>
          
        </div>
      </div>
    <div class="clearer"></div>
  </div>
    <div class="footer">
      &#169;2017, Ryan Dale, Justin Fear.
      
      |
      Powered by <a href="https://www.sphinx-doc.org/">Sphinx 8.2.3</a>
      &amp; <a href="https://alabaster.readthedocs.io">Alabaster 1.0.0</a>
      
      |
      <a href="_sources/downstream-rnaseq.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>