<!DOCTYPE html>

<html lang="en" data-content_root="./">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>Getting started &#8212; lcdb-wf 1.9 documentation</title>
    <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=fa44fd50" />
    <link rel="stylesheet" type="text/css" href="_static/basic.css?v=686e5160" />
    <link rel="stylesheet" type="text/css" href="_static/alabaster.css?v=9735eac1" />
    <script src="_static/documentation_options.js?v=754f84b7"></script>
    <script src="_static/doctools.js?v=9bcbadda"></script>
    <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Guide to file hierarchy" href="guide.html" />
    <link rel="prev" title="Introduction" href="index.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  

  
  

  </head><body>
  <div class="document">
    
      <div class="sphinxsidebar" role="navigation" aria-label="Main">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="toc.html">lcdb-wf</a></h1>



<p class="blurb">Customizable workflows for high-throughput sequencing analysis</p>






<h3>Navigation</h3>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="index.html">Introduction</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Getting started</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#setting-up-a-project">Setting up a project</a></li>
<li class="toctree-l2"><a class="reference internal" href="#deploying-lcdb-wf">1. Deploying lcdb-wf</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#option-1-download-and-run-the-deployment-script">Option 1: Download and run the deployment script</a></li>
<li class="toctree-l3"><a class="reference internal" href="#option-2-clone-repo-manually">Option 2: Clone repo manually</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#configure">2. Configure</a></li>
<li class="toctree-l2"><a class="reference internal" href="#run">3. Run</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#next-steps">Next steps</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="guide.html">Guide to file hierarchy</a></li>
<li class="toctree-l1"><a class="reference internal" href="workflows.html">Overview of workflows</a></li>
<li class="toctree-l1"><a class="reference internal" href="config.html">Configuration</a></li>
<li class="toctree-l1"><a class="reference internal" href="references.html">References workflow</a></li>
<li class="toctree-l1"><a class="reference internal" href="rnaseq.html">RNA-seq workflow</a></li>
<li class="toctree-l1"><a class="reference internal" href="downstream-rnaseq.html">RNA-Seq downstream analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="chipseq.html">ChIP-seq workflow</a></li>
<li class="toctree-l1"><a class="reference internal" href="integrative.html">Integrative workflows</a></li>
<li class="toctree-l1"><a class="reference internal" href="conda.html">conda and conda envs in <cite>lcdb-wf</cite></a></li>
<li class="toctree-l1"><a class="reference internal" href="tests.html">Testing the installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="faqs.html">FAQs</a></li>
<li class="toctree-l1"><a class="reference internal" href="changelog.html">Changelog</a></li>
<li class="toctree-l1"><a class="reference internal" href="developers.html">For Developers</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="toc.html">Documentation overview</a><ul>
      <li>Previous: <a href="index.html" title="previous chapter">Introduction</a></li>
      <li>Next: <a href="guide.html" title="next chapter">Guide to file hierarchy</a></li>
  </ul></li>
</ul>
</div>
<search id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" autocomplete="off" autocorrect="off" autocapitalize="off" spellcheck="false"/>
      <input type="submit" value="Go" />
    </form>
    </div>
</search>
<script>document.getElementById('searchbox').style.display = "block"</script>
        </div>
      </div>
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <section id="getting-started">
<span id="id1"></span><h1>Getting started<a class="headerlink" href="#getting-started" title="Link to this heading">¶</a></h1>
<p>The main prerequisite for <cite>lcdb-wf</cite> is <cite>conda &lt;https://docs.conda.io/en/latest/&gt;_</cite>, with the <a class="reference external" href="https://bioconda.github.io">bioconda</a>. channel set up and the <a class="reference external" href="https://github.com/mamba-org/mamba">mamba</a> drop-in replacement for conda installed.</p>
<p>If this is new to you, please see <a class="reference internal" href="conda.html#conda-envs"><span class="std std-ref">conda and conda envs in lcdb-wf</span></a>.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p><cite>lcdb-wf</cite> is tested and heavily used on Linux. It is only supported on
Linux.</p>
</div>
<section id="setting-up-a-project">
<span id="setup-proj"></span><h2>Setting up a project<a class="headerlink" href="#setting-up-a-project" title="Link to this heading">¶</a></h2>
<p>The general steps to use lcdb-wf in a new project are:</p>
<ol class="arabic simple">
<li><p><strong>Deploy:</strong> download and run <code class="docutils literal notranslate"><span class="pre">deploy.py</span></code> to copy files into a project directory</p></li>
<li><p><strong>Configure:</strong> set up samples table for experiments and edit configuration file</p></li>
<li><p><strong>Run:</strong> activate environment and run the Snakemake file either locally or on a cluster</p></li>
</ol>
</section>
<section id="deploying-lcdb-wf">
<span id="deploy"></span><h2>1. Deploying lcdb-wf<a class="headerlink" href="#deploying-lcdb-wf" title="Link to this heading">¶</a></h2>
<p>Using <cite>lcdb-wf</cite> starts with copying files to a project directory, or
“deploying”.</p>
<p>Unlike other tools you may have used, <cite>lcdb-wf</cite> is not actually installed per
se. Rather, it is “deployed” by copying over relevant files from the <cite>lcdb-wf</cite>
repository to your project directory. This includes Snakefiles, config files,
and other infrastructure required to run, and excludes files like these docs
and testing files that are not necessary for an actual project. The reason is
to use this script is so you end up with a cleaner project directory, compared
to cloning the repo directly.</p>
<p>This script also writes a file to the destination called
<code class="docutils literal notranslate"><span class="pre">.lcdb-wf-deployment.json</span></code>. It stores the timestamp and details about what
commit was used to deploy it. This tracks provenance of the code, so you can
always figure out what lcdb-wf commit your deployment originally started from.</p>
<p>There are a few ways of doing this.</p>
<section id="option-1-download-and-run-the-deployment-script">
<h3>Option 1: Download and run the deployment script<a class="headerlink" href="#option-1-download-and-run-the-deployment-script" title="Link to this heading">¶</a></h3>
<p>This is the most convenient method, although it does not allow running tests
locally.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="nv">BRANCH</span><span class="o">=</span>master<span class="w">  </span><span class="c1"># optionally change branch</span>
wget<span class="w"> </span>https://raw.githubusercontent.com/lcdb/lcdb-wf/<span class="nv">$BRANCH</span>/deploy.py
</pre></div>
</div>
<p>Run <code class="docutils literal notranslate"><span class="pre">python</span> <span class="pre">deploy.py</span> <span class="pre">-h</span></code> to see help. Be sure to use the <code class="docutils literal notranslate"><span class="pre">--staging</span></code> and
<code class="docutils literal notranslate"><span class="pre">--branch=$BRANCH</span></code> arguments when using this method, which will clone the
repository to a location of your choosing. Once you deploy you can remove the
script. For example:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>python<span class="w"> </span>deploy.py<span class="w"> </span><span class="se">\</span>
<span class="w">  </span>--dest<span class="w"> </span>analysis/project<span class="w"> </span><span class="se">\</span>
<span class="w">  </span>--staging<span class="w"> </span>/tmp/lcdb-wf-tmp<span class="w"> </span><span class="se">\</span>
<span class="w">  </span>--branch<span class="w"> </span><span class="nv">$BRANCH</span><span class="w"> </span><span class="se">\</span>
<span class="w">  </span>--flavor<span class="w"> </span>rnaseq<span class="w"> </span><span class="se">\</span>
<span class="w">  </span>--clone<span class="w"> </span><span class="se">\</span>
<span class="w">  </span>--build-envs

<span class="c1"># You can clean up the cloned copy if you want:</span>
<span class="c1"># rm -rf /tmp/lcdb-wf-tmp</span>
</pre></div>
</div>
<p>This will clone the full git repo to <code class="docutils literal notranslate"><span class="pre">/tmp/lcdb-wf-tmp</span></code>, check out the master
branch (or whatever branch <code class="docutils literal notranslate"><span class="pre">$BRANCH</span></code> is set to), copy the files required for
an RNA-seq project over to <code class="docutils literal notranslate"><span class="pre">analysis/project</span></code>, build the main conda
environment and the R environment, save the <code class="docutils literal notranslate"><span class="pre">.lcdb-wf-deployment.json</span></code> file
there, and then delete the temporary repo.</p>
</section>
<section id="option-2-clone-repo-manually">
<h3>Option 2: Clone repo manually<a class="headerlink" href="#option-2-clone-repo-manually" title="Link to this heading">¶</a></h3>
<p>Clone a repo using git and check out the branch. Use this method for running
tests):</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="nv">BRANCH</span><span class="o">=</span>master<span class="w">  </span><span class="c1"># optionally change branch</span>
git<span class="w"> </span>clone<span class="w"> </span>https://github.com/lcdb/lcdb-wf<span class="w"> </span>/tmp/lcdb-wf
<span class="nb">cd</span><span class="w"> </span>/tmp/lcdb-wf
git<span class="w"> </span>checkout<span class="w"> </span><span class="nv">$BRANCH</span>
</pre></div>
</div>
<p>Then run the deploy script, <code class="docutils literal notranslate"><span class="pre">python</span> <span class="pre">deploy.py</span> <span class="pre">-h</span></code> to see usage info. Here is
an example for RNA-seq:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>python<span class="w"> </span>deploy.py<span class="w"> </span><span class="se">\</span>
<span class="w">  </span>--dest<span class="w"> </span>analysis/project<span class="w"> </span><span class="se">\</span>
<span class="w">  </span>--flavor<span class="w"> </span>rnaseq<span class="w"> </span><span class="se">\</span>
<span class="w">  </span>--build-envs
</pre></div>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>If you want to run the tests then don’t deploy just yet – see
<a class="reference internal" href="tests.html#running-the-tests"><span class="std std-ref">Testing the installation</span></a> for details, and then come back here to deploy for
an actual project.</p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>See <a class="reference internal" href="conda.html#conda-envs"><span class="std std-ref">conda and conda envs in lcdb-wf</span></a> for more details on the conda environment building.</p>
</div>
</section>
</section>
<section id="configure">
<h2>2. Configure<a class="headerlink" href="#configure" title="Link to this heading">¶</a></h2>
<p>This step takes the most effort. The first time you set up a project it
will take some time to understand the configuration system.</p>
<ul class="simple">
<li><p>see <a class="reference internal" href="sampletable.html#sampletable"><span class="std std-ref">Sample tables</span></a> for how to write a sampletable, which includes where to find raw data and contains the associated metadata</p></li>
<li><p>see <a class="reference internal" href="config-yaml.html#config-yaml"><span class="std std-ref">Config YAML</span></a> for configuring each workflow</p></li>
<li><p>see <a class="reference internal" href="faqs.html#multiple-experiments"><span class="std std-ref">How do I handle multiple experiments in the same project?</span></a> for advice on how to handle multiple experiments that are intended to be analyzed together</p></li>
</ul>
</section>
<section id="run">
<h2>3. Run<a class="headerlink" href="#run" title="Link to this heading">¶</a></h2>
<p>Activate the main environment and go to the workflow you want to run. For
example if you have deployed and configured an RNA-seq run, then do:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>conda<span class="w"> </span>activate<span class="w"> </span>./env
<span class="nb">cd</span><span class="w"> </span>workflows/rnaseq
</pre></div>
</div>
<p>and run the following:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>snakemake<span class="w"> </span>--dryrun
</pre></div>
</div>
<p>If all goes well, this should print a list of jobs to be run.</p>
<p>You can run locally, but this is NOT recommended for a typicaly RNA-seq
project. To run locally, choose the number of CPUs you want to use with the
<code class="docutils literal notranslate"><span class="pre">-j</span></code> argument as is standard for Snakemake.</p>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>If you haven’t made any changes to the Snakefiles, be aware that the
default configuration needs a lot of RAM. For example, the MarkDuplicates
runs set 20 GB RAM for Java, and that’s for each job. Adjust the Snakefiles
accordingly if you don’t have enough RAM available (search for “Xmx” to
find the Java args that set memory).</p>
</div>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="c1"># run locally (not recommended)</span>
snakemake<span class="w"> </span>--use-conda<span class="w"> </span>-j<span class="w"> </span><span class="m">8</span>
</pre></div>
</div>
<p>The recommended way is to run on a cluster.</p>
<p>To run on a cluster, you will need a <a class="reference external" href="https://snakemake.readthedocs.io/en/stable/executing/cli.html#profiles">Snakemake profile</a> for
your cluster that translates generic resource requirements into arguments for
your cluster’s batch system.</p>
<p>On NIH’s Biowulf cluster, the profile can be found at
<a class="reference external" href="https://github.com/NIH-HPC/snakemake_profile">https://github.com/NIH-HPC/snakemake_profile</a>. If you are not already using this for other Snakemake workflows, you can set it up the first time like this:</p>
<ol class="arabic simple">
<li><p>Clone the profile to a location of your choosing, maybe
<code class="docutils literal notranslate"><span class="pre">~/snakemake_profile</span></code></p></li>
<li><p>Set the environment variable <code class="docutils literal notranslate"><span class="pre">LCDBWF_SNAKEMAKE_PROFILE</span></code>, perhaps in your
<code class="docutils literal notranslate"><span class="pre">~/.bashrc</span></code> file.</p></li>
</ol>
<p>Then back in your deployed and configured project, submit the wrapper script as
a batch job:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>sbatch<span class="w"> </span>../../include/WRAPPER_SLURM
</pre></div>
</div>
<p>This will submit Snakemake as a batch job, use the profile to translate
resources to cluster arguments and set default command-line arguments, and
submit the various jobs created by Snakemake to the cluster on your behalf. See
<a class="reference internal" href="config.html#cluster"><span class="std std-ref">Running on a cluster</span></a> for more details on this.</p>
<p>Other clusters will need different configuration, but everything in <cite>lcdb-wf</cite>
is standard Snakemake. The Snakemake documentation on <a class="reference external" href="https://snakemake.readthedocs.io/en/stable/executing/cluster.html">cluster execution</a> and
<a class="reference external" href="https://snakemake.readthedocs.io/en/stable/executing/cloud.html">cloud execution</a> can be
consulted for running on your particular system.</p>
<p>You can typically run simultaneous workflows when they are in different
directories; see <a class="reference internal" href="workflows.html#workflows"><span class="std std-ref">Overview of workflows</span></a> for details.</p>
<section id="next-steps">
<h3>Next steps<a class="headerlink" href="#next-steps" title="Link to this heading">¶</a></h3>
<p>Next, we give a brief overview of the file hierarchy of <code class="docutils literal notranslate"><span class="pre">lcdb-wf</span></code> in the
<a class="reference internal" href="guide.html#guide"><span class="std std-ref">Guide to file hierarchy</span></a> page.</p>
</section>
</section>
</section>


          </div>
          
        </div>
      </div>
    <div class="clearer"></div>
  </div>
    <div class="footer">
      &#169;2017, Ryan Dale, Justin Fear.
      
      |
      Powered by <a href="https://www.sphinx-doc.org/">Sphinx 8.1.0</a>
      &amp; <a href="https://alabaster.readthedocs.io">Alabaster 1.0.0</a>
      
      |
      <a href="_sources/getting-started.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>