<!DOCTYPE html>

<html lang="en" data-content_root="./">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>Introduction &#8212; lcdb-wf 1.9 documentation</title>
    <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=03e43079" />
    <link rel="stylesheet" type="text/css" href="_static/basic.css?v=b08954a9" />
    <link rel="stylesheet" type="text/css" href="_static/alabaster.css?v=9735eac1" />
    <script src="_static/documentation_options.js?v=754f84b7"></script>
    <script src="_static/doctools.js?v=9bcbadda"></script>
    <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Getting started" href="getting-started.html" />
    <link rel="prev" title="Table of Contents" href="toc.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  

  
  

  </head><body>
  <div class="document">
    
      <div class="sphinxsidebar" role="navigation" aria-label="Main">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="toc.html">lcdb-wf</a></h1>



<p class="blurb">Customizable workflows for high-throughput sequencing analysis</p>






<h3>Navigation</h3>
<ul class="current">
<li class="toctree-l1 current"><a class="current reference internal" href="#">Introduction</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#designed-with-customization-in-mind">Designed with customization in mind</a></li>
<li class="toctree-l2"><a class="reference internal" href="#extensive-downstream-rna-seq">Extensive downstream RNA-seq</a></li>
<li class="toctree-l2"><a class="reference internal" href="#extenstive-exploration-of-chip-seq-peaks">Extenstive exploration of ChIP-seq peaks</a></li>
<li class="toctree-l2"><a class="reference internal" href="#track-hubs">Track hubs</a></li>
<li class="toctree-l2"><a class="reference internal" href="#unified-approach-to-reference-genomes">Unified approach to reference genomes</a></li>
<li class="toctree-l2"><a class="reference internal" href="#integration-with-external-data-and-figure-making">Integration with external data and figure-making</a></li>
<li class="toctree-l2"><a class="reference internal" href="#tested-automatically">Tested automatically</a></li>
<li class="toctree-l2"><a class="reference internal" href="#all-the-advantages-of-snakemake">All the advantages of Snakemake</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#run-locally-or-on-a-cluster-with-the-same-code">Run locally or on a cluster with the same code</a></li>
<li class="toctree-l3"><a class="reference internal" href="#only-run-the-required-jobs">Only run the required jobs</a></li>
<li class="toctree-l3"><a class="reference internal" href="#parallelization">Parallelization</a></li>
<li class="toctree-l3"><a class="reference internal" href="#software-installation">Software installation</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="getting-started.html">Getting started</a></li>
<li class="toctree-l1"><a class="reference internal" href="guide.html">Guide to file hierarchy</a></li>
<li class="toctree-l1"><a class="reference internal" href="workflows.html">Overview of workflows</a></li>
<li class="toctree-l1"><a class="reference internal" href="config.html">Configuration</a></li>
<li class="toctree-l1"><a class="reference internal" href="references.html">References workflow</a></li>
<li class="toctree-l1"><a class="reference internal" href="rnaseq.html">RNA-seq workflow</a></li>
<li class="toctree-l1"><a class="reference internal" href="downstream-rnaseq.html">RNA-Seq downstream analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="chipseq.html">ChIP-seq workflow</a></li>
<li class="toctree-l1"><a class="reference internal" href="integrative.html">Integrative workflows</a></li>
<li class="toctree-l1"><a class="reference internal" href="conda.html">conda and conda envs in <cite>lcdb-wf</cite></a></li>
<li class="toctree-l1"><a class="reference internal" href="tests.html">Testing the installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="faqs.html">FAQs</a></li>
<li class="toctree-l1"><a class="reference internal" href="changelog.html">Changelog</a></li>
<li class="toctree-l1"><a class="reference internal" href="developers.html">For Developers</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="toc.html">Documentation overview</a><ul>
      <li>Previous: <a href="toc.html" title="previous chapter">Table of Contents</a></li>
      <li>Next: <a href="getting-started.html" title="next chapter">Getting started</a></li>
  </ul></li>
</ul>
</div>
<search id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" autocomplete="off" autocorrect="off" autocapitalize="off" spellcheck="false"/>
      <input type="submit" value="Go" />
    </form>
    </div>
</search>
<script>document.getElementById('searchbox').style.display = "block"</script>
        </div>
      </div>
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <section id="introduction">
<h1>Introduction<a class="headerlink" href="#introduction" title="Link to this heading">¶</a></h1>
<p><cite>lcdb-wf</cite> is a collection of Snakemake workflows for common high-throughput
sequencing analysis.</p>
<p>There are a multitude of workflows out there for high-throughput sequencing analysis.
What makes <cite>lcdb-wf</cite> different?</p>
<section id="designed-with-customization-in-mind">
<h2>Designed with customization in mind<a class="headerlink" href="#designed-with-customization-in-mind" title="Link to this heading">¶</a></h2>
<p>We recognize that every experiment has its own idiosyncracies. Rather
than provide a one-size-fits-all solution, we aim to provide a reasonable
starting point that users can modify for their own use.</p>
<p>Non-model organism? Custom gene annotations? Complicated regression models?
Unconventional command-line arguments to tools? New tools to add to the
workflow? No problem.</p>
</section>
<section id="extensive-downstream-rna-seq">
<h2>Extensive downstream RNA-seq<a class="headerlink" href="#extensive-downstream-rna-seq" title="Link to this heading">¶</a></h2>
<p>A comprehensive RMarkdown template, along with a custom R package, enables
sophisticated RNA-seq analysis that supports complex experimental designs and
many contrasts.</p>
</section>
<section id="extenstive-exploration-of-chip-seq-peaks">
<h2>Extenstive exploration of ChIP-seq peaks<a class="headerlink" href="#extenstive-exploration-of-chip-seq-peaks" title="Link to this heading">¶</a></h2>
<p>The ChIP-seq configuration supports multiple peak-callers as well as calling
peaks with many different parameter sets for each caller. Combined with
visualizaiton in track hubs (see below), this can identify the optimal
parameters for a given experiment.</p>
</section>
<section id="track-hubs">
<h2>Track hubs<a class="headerlink" href="#track-hubs" title="Link to this heading">¶</a></h2>
<p>The ChIP-seq and RNA-seq workflows generate track hubs that can be viewed in
the UCSC Genome Browser. ChIP-seq shows signal and called peaks (many
peak-calling runs, with different peak-callers and different parameters for
each, can be configured to automatically run and show up in the track hub);
RNA-seq shows strand-specific signal tracks. Both support the addition of
arbitrary additional tracks (primers, loci of interest, external data, etc) to
view alongside your data.</p>
</section>
<section id="unified-approach-to-reference-genomes">
<h2>Unified approach to reference genomes<a class="headerlink" href="#unified-approach-to-reference-genomes" title="Link to this heading">¶</a></h2>
<p>The references workflow defines the genome files for the organisms used in
the experiment. It it is shared by RNA-seq and ChIP-seq and is driven by
a config file that specifies URLs for FASTA and GTF files. Set it up once for
a site to get lots of genomes you can use for running <cite>fastq_screen</cite>, and
easily include arbitrary other genomes. They can then be automatically included
in RNA-seq and ChIP-seq workflows.</p>
<p>Arbitrary genomes can be used, whether local (e.g., customized with additional
genetic constructs) or on the web. The <cite>references</cite> workflow need only be run
once for all these genomes to be created, with the <cite>references_dir</cite> being used
as a centralized repository that can be then used with all other workflows.</p>
</section>
<section id="integration-with-external-data-and-figure-making">
<h2>Integration with external data and figure-making<a class="headerlink" href="#integration-with-external-data-and-figure-making" title="Link to this heading">¶</a></h2>
<p>It is designed to tie together your entire analysis, from downloading references
through producing final figures. The included examples tie together workflows that download external data, perform downstream analysis, and make figures.</p>
<p>If an upstream file changes (e.g., gene annotation), all dependent downstream
jobs – including figures – will be updated so you can ensure that even
complex analyses stay correct and up-to-date.</p>
</section>
<section id="tested-automatically">
<h2>Tested automatically<a class="headerlink" href="#tested-automatically" title="Link to this heading">¶</a></h2>
<p>Every change to the code on GitHub triggers an automated test, the results of
which you can find at <a class="reference external" href="https://circleci.com/gh/lcdb/lcdb-wf">https://circleci.com/gh/lcdb/lcdb-wf</a>. Each test sets the
system up from scratch, including installing all software, downloading example
data, and running everything up through the final results. This guarantees that
you can set up and test the code yourself.</p>
</section>
<section id="all-the-advantages-of-snakemake">
<h2>All the advantages of Snakemake<a class="headerlink" href="#all-the-advantages-of-snakemake" title="Link to this heading">¶</a></h2>
<p>The workflows are built using <a class="reference external" href="https://snakemake.readthedocs.io/en/stable/">Snakemake</a>, so we also get the following
for free:</p>
<section id="run-locally-or-on-a-cluster-with-the-same-code">
<h3>Run locally or on a cluster with the same code<a class="headerlink" href="#run-locally-or-on-a-cluster-with-the-same-code" title="Link to this heading">¶</a></h3>
<p>Run the same workflow locally or on a cluster with a single command-line flag.
Use <a class="reference external" href="https://snakemake.readthedocs.io/en/stable/executing/cli.html#profiles">Snakemake profiles</a> to
translate from general resources to cluster-specific
arguments.</p>
</section>
<section id="only-run-the-required-jobs">
<h3>Only run the required jobs<a class="headerlink" href="#only-run-the-required-jobs" title="Link to this heading">¶</a></h3>
<p>New gene annotation? Snakemake tracks dependencies, so it will detect that the
annotations changed. Only jobs that depend on that file at some point in their
dependency chain will be re-run and the independent files are untouched. Adding
a new sample will leave untouched any output from samples that have already
run.</p>
</section>
<section id="parallelization">
<h3>Parallelization<a class="headerlink" href="#parallelization" title="Link to this heading">¶</a></h3>
<p>It’s trivial to run jobs in parallel, for as many CPUs or cluster nodes as you
have available by using the <code class="docutils literal notranslate"><span class="pre">-j</span></code> argument to Snakemake.</p>
</section>
<section id="software-installation">
<h3>Software installation<a class="headerlink" href="#software-installation" title="Link to this heading">¶</a></h3>
<p>Installation of all dependencies is handled by conda, ensuring reproducibility,
streamlined setup, and no need for root administrator privileges.</p>
<p>See <a class="reference internal" href="getting-started.html#getting-started"><span class="std std-ref">Getting started</span></a> to get started.</p>
</section>
</section>
</section>


          </div>
          
        </div>
      </div>
    <div class="clearer"></div>
  </div>
    <div class="footer">
      &#169;2017, Ryan Dale, Justin Fear.
      
      |
      Powered by <a href="https://www.sphinx-doc.org/">Sphinx 8.2.3</a>
      &amp; <a href="https://alabaster.readthedocs.io">Alabaster 1.0.0</a>
      
      |
      <a href="_sources/index.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>