<!DOCTYPE html>

<html lang="en" data-content_root="./">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>Module lib.chipseq &#8212; lcdb-wf 1.9 documentation</title>
    <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=03e43079" />
    <link rel="stylesheet" type="text/css" href="_static/basic.css?v=b08954a9" />
    <link rel="stylesheet" type="text/css" href="_static/alabaster.css?v=9735eac1" />
    <script src="_static/documentation_options.js?v=754f84b7"></script>
    <script src="_static/doctools.js?v=9bcbadda"></script>
    <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Module lib.patterns_targets" href="lib.patterns_targets.html" />
    <link rel="prev" title="Module lib.common" href="lib.common.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  

  
  

  </head><body>
  <div class="document">
    
      <div class="sphinxsidebar" role="navigation" aria-label="Main">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="toc.html">lcdb-wf</a></h1>



<p class="blurb">Customizable workflows for high-throughput sequencing analysis</p>






<h3>Navigation</h3>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="index.html">Introduction</a></li>
<li class="toctree-l1"><a class="reference internal" href="getting-started.html">Getting started</a></li>
<li class="toctree-l1"><a class="reference internal" href="guide.html">Guide to file hierarchy</a></li>
<li class="toctree-l1"><a class="reference internal" href="workflows.html">Overview of workflows</a></li>
<li class="toctree-l1"><a class="reference internal" href="config.html">Configuration</a></li>
<li class="toctree-l1"><a class="reference internal" href="references.html">References workflow</a></li>
<li class="toctree-l1"><a class="reference internal" href="rnaseq.html">RNA-seq workflow</a></li>
<li class="toctree-l1"><a class="reference internal" href="downstream-rnaseq.html">RNA-Seq downstream analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="chipseq.html">ChIP-seq workflow</a></li>
<li class="toctree-l1"><a class="reference internal" href="integrative.html">Integrative workflows</a></li>
<li class="toctree-l1"><a class="reference internal" href="conda.html">conda and conda envs in <cite>lcdb-wf</cite></a></li>
<li class="toctree-l1"><a class="reference internal" href="tests.html">Testing the installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="faqs.html">FAQs</a></li>
<li class="toctree-l1"><a class="reference internal" href="changelog.html">Changelog</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="developers.html">For Developers</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="developers.html#creating-and-updating-conda-envs">Creating and updating conda envs</a></li>
<li class="toctree-l2"><a class="reference internal" href="developers.html#running-the-full-complex-datasets">Running the full complex datasets</a></li>
<li class="toctree-l2 current"><a class="reference internal" href="developers.html#module-documentation">Module documentation</a><ul class="current">
<li class="toctree-l3"><a class="reference internal" href="lib.common.html">Module <code class="docutils literal notranslate"><span class="pre">lib.common</span></code></a></li>
<li class="toctree-l3 current"><a class="current reference internal" href="#">Module <code class="docutils literal notranslate"><span class="pre">lib.chipseq</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="lib.patterns_targets.html">Module <code class="docutils literal notranslate"><span class="pre">lib.patterns_targets</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="developers.html#adding-a-new-aligner">Adding a new aligner</a></li>
<li class="toctree-l2"><a class="reference internal" href="developers.html#adding-a-new-peak-caller">Adding a new peak-caller</a></li>
</ul>
</li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="toc.html">Documentation overview</a><ul>
  <li><a href="developers.html">For Developers</a><ul>
      <li>Previous: <a href="lib.common.html" title="previous chapter">Module <code class="docutils literal notranslate"><span class="pre">lib.common</span></code></a></li>
      <li>Next: <a href="lib.patterns_targets.html" title="next chapter">Module <code class="docutils literal notranslate"><span class="pre">lib.patterns_targets</span></code></a></li>
  </ul></li>
  </ul></li>
</ul>
</div>
<search id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" autocomplete="off" autocorrect="off" autocapitalize="off" spellcheck="false"/>
      <input type="submit" value="Go" />
    </form>
    </div>
</search>
<script>document.getElementById('searchbox').style.display = "block"</script>
        </div>
      </div>
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <section id="module-lib-chipseq">
<h1>Module <code class="docutils literal notranslate"><span class="pre">lib.chipseq</span></code><a class="headerlink" href="#module-lib-chipseq" title="Link to this heading">¶</a></h1>
<p>Handling ChIP-seq peak-calling configuration correctly is complex. The
functions in this module help manipulate the config information so we can use
it more easily in the ChIP-seq workflow without cluttering the Snakefile.</p>
<table class="autosummary longtable docutils align-default">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#lib.chipseq.peak_calling_dict" title="lib.chipseq.peak_calling_dict"><code class="xref py py-obj docutils literal notranslate"><span class="pre">peak_calling_dict</span></code></a>(config[, algorithm])</p></td>
<td><p>Returns a dictionary of peak-calling runs from the config.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#lib.chipseq.block_for_run" title="lib.chipseq.block_for_run"><code class="xref py py-obj docutils literal notranslate"><span class="pre">block_for_run</span></code></a>(config, label, algorithm)</p></td>
<td><p>Returns the block for the (label, algorithm) run.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#lib.chipseq.samples_for_run" title="lib.chipseq.samples_for_run"><code class="xref py py-obj docutils literal notranslate"><span class="pre">samples_for_run</span></code></a>(config, label, algorithm, ...)</p></td>
<td><p>Returns the sample names configured for a particular peak-calling run</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#lib.chipseq.merged_input_for_ip" title="lib.chipseq.merged_input_for_ip"><code class="xref py py-obj docutils literal notranslate"><span class="pre">merged_input_for_ip</span></code></a>(sampletable, merged_ip)</p></td>
<td><p>Returns the merged input label for a merged IP label.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#lib.chipseq.detect_peak_format" title="lib.chipseq.detect_peak_format"><code class="xref py py-obj docutils literal notranslate"><span class="pre">detect_peak_format</span></code></a>(fn)</p></td>
<td><p>Figure out if a BED file is narrowPeak or broadPeak.</p></td>
</tr>
</tbody>
</table>
<section id="module-lib.chipseq">
<span id="details"></span><h2>Details<a class="headerlink" href="#module-lib.chipseq" title="Link to this heading">¶</a></h2>
<p>Helpers for ChIP-seq.</p>
<dl class="py function">
<dt class="sig sig-object py" id="lib.chipseq.block_for_run">
<span class="sig-prename descclassname"><span class="pre">lib.chipseq.</span></span><span class="sig-name descname"><span class="pre">block_for_run</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">config</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">label</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">algorithm</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/lib/chipseq.html#block_for_run"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#lib.chipseq.block_for_run" title="Link to this definition">¶</a></dt>
<dd><p>Returns the block for the (label, algorithm) run.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>config</strong> (<em>dict</em>)</p></li>
<li><p><strong>label</strong> (<em>str</em>)</p></li>
<li><p><strong>algorithm</strong> (<em>str</em>)</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="lib.chipseq.detect_peak_format">
<span class="sig-prename descclassname"><span class="pre">lib.chipseq.</span></span><span class="sig-name descname"><span class="pre">detect_peak_format</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fn</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/lib/chipseq.html#detect_peak_format"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#lib.chipseq.detect_peak_format" title="Link to this definition">¶</a></dt>
<dd><p>Figure out if a BED file is narrowPeak or broadPeak.</p>
<p>Returns None if undetermined.</p>
<p>This is useful for figuring out which autoSql file we should use or which
bigBed 6, 6+4, or 6+3 format to use.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="lib.chipseq.merged_input_for_ip">
<span class="sig-prename descclassname"><span class="pre">lib.chipseq.</span></span><span class="sig-name descname"><span class="pre">merged_input_for_ip</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sampletable</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">merged_ip</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/lib/chipseq.html#merged_input_for_ip"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#lib.chipseq.merged_input_for_ip" title="Link to this definition">¶</a></dt>
<dd><p>Returns the merged input label for a merged IP label.</p>
<p>This is primarily used for the <cite>fingerprint</cite> rule, where we collect all the
available input BAMs together.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>sampletable</strong> (<em>pandas.DataFrame</em>)</p></li>
<li><p><strong>merged_ip</strong> (<em>str</em>) – Label of IP to use, must be present in the <cite>label</cite> column of the
sampletable.</p></li>
</ul>
</dd>
</dl>
<p class="rubric">Examples</p>
<p>This should make more sense if we have an example to work with…..</p>
<p>Samples ip1 and ip2 are technical replicates. They are from a different
experiment than ip3 and input3, hence their different
biological_material.</p>
<p>The way we know that input1 should be paired with ip1 and ip2 is because
it shares the same biological material.</p>
<p>Compare input1 and input9. They are not technical replicates (since they
do not share the same <cite>label</cite>) but they are biological replicates because
they share the same biological material.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">io</span><span class="w"> </span><span class="kn">import</span> <span class="n">StringIO</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span><span class="w"> </span><span class="nn">pandas</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pd</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">StringIO</span><span class="p">(</span><span class="s1">&#39;&#39;&#39;</span>
<span class="gp">... </span><span class="s1">samplename  antibody   biological_material  label</span>
<span class="gp">... </span><span class="s1">ip1         gaf        s2cell-1             s2cell-gaf-1</span>
<span class="gp">... </span><span class="s1">ip2         gaf        s2cell-1             s2cell-gaf-1</span>
<span class="gp">... </span><span class="s1">ip3         ctcf       s2cell-2             s2cell-ctcf-1</span>
<span class="gp">... </span><span class="s1">input1      input      s2cell-1             s2cell-input-1</span>
<span class="gp">... </span><span class="s1">input3      input      s2cell-2             s2cell-input-3</span>
<span class="gp">... </span><span class="s1">input9      input      s2cell-1             s2cell-input-1&#39;&#39;&#39;</span><span class="p">),</span>
<span class="gp">... </span><span class="n">sep</span><span class="o">=</span><span class="s1">&#39;\s+&#39;</span><span class="p">)</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">merged_input_for_ip</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="s1">&#39;s2cell-gaf-1&#39;</span><span class="p">)</span>
<span class="go">[&#39;s2cell-input-1&#39;]</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">merged_input_for_ip</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="s1">&#39;s2cell-ctcf-1&#39;</span><span class="p">)</span>
<span class="go">[&#39;s2cell-input-3&#39;]</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="lib.chipseq.peak_calling_dict">
<span class="sig-prename descclassname"><span class="pre">lib.chipseq.</span></span><span class="sig-name descname"><span class="pre">peak_calling_dict</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">config</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">algorithm</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/lib/chipseq.html#peak_calling_dict"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#lib.chipseq.peak_calling_dict" title="Link to this definition">¶</a></dt>
<dd><p>Returns a dictionary of peak-calling runs from the config.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>config</strong> (<em>dict</em>)</p></li>
<li><p><strong>algorithm</strong> (<em>None</em>) – If <cite>algorithm</cite> is None, dictionary is keyed by (label, algorithm).
Otherwise, only the runs for <cite>algorithm</cite> are returned, keyed by
<cite>label</cite>.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="lib.chipseq.samples_for_run">
<span class="sig-prename descclassname"><span class="pre">lib.chipseq.</span></span><span class="sig-name descname"><span class="pre">samples_for_run</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">config</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">label</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">algorithm</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">treatment</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/lib/chipseq.html#samples_for_run"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#lib.chipseq.samples_for_run" title="Link to this definition">¶</a></dt>
<dd><p>Returns the sample names configured for a particular peak-calling run</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>config</strong> (<em>dict</em>)</p></li>
<li><p><strong>label</strong> – Used as keys into peak_calling_dict()</p></li>
<li><p><strong>algorithm</strong> – Used as keys into peak_calling_dict()</p></li>
<li><p><strong>treatment</strong> (<em>ip</em><em> | </em><em>input</em>)</p></li>
</ul>
</dd>
</dl>
</dd></dl>

</section>
</section>


          </div>
          
        </div>
      </div>
    <div class="clearer"></div>
  </div>
    <div class="footer">
      &#169;2017, Ryan Dale, Justin Fear.
      
      |
      Powered by <a href="https://www.sphinx-doc.org/">Sphinx 8.2.3</a>
      &amp; <a href="https://alabaster.readthedocs.io">Alabaster 1.0.0</a>
      
      |
      <a href="_sources/lib.chipseq.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>