
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

    <title>Patterns and targets &#8212; lcdb-wf 0.1 documentation</title>
    <link rel="stylesheet" type="text/css" href="_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="_static/alabaster.css" />
    <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
    <script src="_static/jquery.js"></script>
    <script src="_static/underscore.js"></script>
    <script src="_static/doctools.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="FAQs" href="faqs.html" />
    <link rel="prev" title="Sample tables" href="sampletable.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  <div class="document">
    
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="toc.html">lcdb-wf</a></h1>



<p class="blurb">Customizable workflows for high-throughput sequencing analysis</p>






<h3>Navigation</h3>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="index.html">Introduction</a></li>
<li class="toctree-l1"><a class="reference internal" href="getting-started.html">Getting started</a></li>
<li class="toctree-l1"><a class="reference internal" href="guide.html">Guide to file hierarchy</a></li>
<li class="toctree-l1"><a class="reference internal" href="tests.html">Testing the installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="workflows.html">Overview of workflows</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="config.html">Configuration details</a><ul class="current">
<li class="toctree-l2 current"><a class="reference internal" href="config.html#general-configuration">General configuration</a></li>
<li class="toctree-l2"><a class="reference internal" href="config.html#running-on-a-cluster">Running on a cluster</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="references.html">References workflow</a></li>
<li class="toctree-l1"><a class="reference internal" href="rnaseq.html">RNA-seq workflow</a></li>
<li class="toctree-l1"><a class="reference internal" href="downstream-rnaseq.html">RNA-Seq downstream analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="chipseq.html">ChIP-seq workflow</a></li>
<li class="toctree-l1"><a class="reference internal" href="integrative.html">Integrative workflows</a></li>
<li class="toctree-l1"><a class="reference internal" href="faqs.html">FAQs</a></li>
<li class="toctree-l1"><a class="reference internal" href="changelog.html">Changelog</a></li>
<li class="toctree-l1"><a class="reference internal" href="developers.html">For Developers</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="toc.html">Documentation overview</a><ul>
  <li><a href="config.html">Configuration details</a><ul>
      <li>Previous: <a href="sampletable.html" title="previous chapter">Sample tables</a></li>
      <li>Next: <a href="faqs.html" title="next chapter">FAQs</a></li>
  </ul></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" />
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <section id="patterns-and-targets">
<span id="id1"></span><h1>Patterns and targets<a class="headerlink" href="#patterns-and-targets" title="Permalink to this headline">¶</a></h1>
<p>We use a hybrid approach to specifying input and output patterns for Snakemake
rules. These are used in the RNA-seq and ChIP-seq workflows.</p>
<section id="patterns">
<h2>Patterns<a class="headerlink" href="#patterns" title="Permalink to this headline">¶</a></h2>
<p><strong>Patterns</strong> are filename-like strings with placeholders in them, like this:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">data</span><span class="o">/</span><span class="n">rnaseq_samples</span><span class="o">/</span><span class="p">{</span><span class="n">sample</span><span class="p">}</span><span class="o">/</span><span class="p">{</span><span class="n">sample</span><span class="p">}</span><span class="n">_R1</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span>
</pre></div>
</div>
<p>Generally you don’t need to modify them (though you can). The patterns file is
useful as a guide to the files that are created by the workflow.</p>
<ul class="simple">
<li><p>RNA-seq patterns are in <code class="docutils literal notranslate"><span class="pre">workflows/rnaseq/config/rnaseq_patterns.yaml</span></code></p></li>
<li><p>ChIP-seq patterns are in <code class="docutils literal notranslate"><span class="pre">workflows/chipseq/config/chipseq_patterns.yaml</span></code></p></li>
</ul>
</section>
<section id="targets">
<h2>Targets<a class="headerlink" href="#targets" title="Permalink to this headline">¶</a></h2>
<p>The metadata (sample names, peak-calling runs) configured in the sample table
and config file are used to fill in the patterns to create the targets. It’s the
equivalent of a complicated <cite>expand()</cite> call in a standard
Snakefile.</p>
<p>If we had 2 samples, A and B, then filling in the pattern:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">data</span><span class="o">/</span><span class="n">rnaseq_samples</span><span class="o">/</span><span class="p">{</span><span class="n">sample</span><span class="p">}</span><span class="o">/</span><span class="p">{</span><span class="n">sample</span><span class="p">}</span><span class="n">_R1</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span>
</pre></div>
</div>
<p>would result in:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">data</span><span class="o">/</span><span class="n">rnaseq_samples</span><span class="o">/</span><span class="n">A</span><span class="o">/</span><span class="n">A_R1</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span>
<span class="n">data</span><span class="o">/</span><span class="n">rnaseq_samples</span><span class="o">/</span><span class="n">B</span><span class="o">/</span><span class="n">B_R1</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span>
</pre></div>
</div>
</section>
<section id="how-patterns-and-targets-are-used-in-snakefiles">
<h2>How patterns and targets are used in Snakefiles<a class="headerlink" href="#how-patterns-and-targets-are-used-in-snakefiles" title="Permalink to this headline">¶</a></h2>
<p>Briefly:</p>
<ul class="simple">
<li><p>Each Snakefile has access to an object, <code class="docutils literal notranslate"><span class="pre">c</span></code>.</p></li>
<li><p><strong>Patterns</strong> are accessed via the <code class="docutils literal notranslate"><span class="pre">c.patterns</span></code> dictionary. The structure of
<code class="docutils literal notranslate"><span class="pre">c.patterns</span></code> matches that of the patterns file. Patterns still have the
<code class="docutils literal notranslate"><span class="pre">{}</span></code> placeholders as written in that file.</p></li>
<li><p><strong>Targets</strong> are accessed via the <code class="docutils literal notranslate"><span class="pre">c.targets</span></code> dictionary. The structure
matches that of <code class="docutils literal notranslate"><span class="pre">c.patterns</span></code>, but the placeholders are filled in based on
other configuration information such that each single string from patterns
may turn into a list after being filled in using the <cite>snakemake.expand()</cite>
function.</p></li>
<li><p>We can collapse arbitrary groups of patterns or targets together into
a flattened list with <code class="docutils literal notranslate"><span class="pre">lib.utils.flatten()</span></code>.</p></li>
</ul>
<p>Here is an example rule that uses patterns and targets from the <code class="docutils literal notranslate"><span class="pre">c</span></code> object:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">rule</span> <span class="nb">all</span><span class="p">:</span>
    <span class="nb">input</span><span class="p">:</span>
        <span class="n">c</span><span class="o">.</span><span class="n">targets</span><span class="p">[</span><span class="s1">&#39;cutadapt&#39;</span><span class="p">]</span>

<span class="n">rule</span> <span class="n">cutadapt</span><span class="p">:</span>
    <span class="nb">input</span><span class="p">:</span>
        <span class="n">c</span><span class="o">.</span><span class="n">patterns</span><span class="p">[</span><span class="s1">&#39;fastq&#39;</span><span class="p">]</span>
    <span class="n">output</span><span class="p">:</span>
        <span class="n">c</span><span class="o">.</span><span class="n">patterns</span><span class="p">[</span><span class="s1">&#39;cutadapt&#39;</span><span class="p">]</span>

    <span class="n">shell</span><span class="p">:</span>
        <span class="s1">&#39;cutadapt </span><span class="si">{input}</span><span class="s1"> -o </span><span class="si">{output}</span><span class="s1">&#39;</span>
</pre></div>
</div>
<p>In the example above, <code class="docutils literal notranslate"><span class="pre">c.patterns['fastq']</span></code> and <code class="docutils literal notranslate"><span class="pre">c.patterns['cutadapt']</span></code>
have the following values, configured in <code class="docutils literal notranslate"><span class="pre">config/rnaseq_patterns.yaml</span></code>:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">c</span><span class="o">.</span><span class="n">patterns</span><span class="p">[</span><span class="s1">&#39;fastq&#39;</span><span class="p">]</span>
<span class="c1"># data/rnaseq_samples/{sample}/{sample}_R1.fastq.gz</span>

<span class="n">c</span><span class="o">.</span><span class="n">patterns</span><span class="p">[</span><span class="s1">&#39;cutadapt&#39;</span><span class="p">]</span>
<span class="c1"># data/rnaseq_samples/{sample}/{sample}_R1.cutadapt.fastq.gz</span>
</pre></div>
</div>
<p>And <code class="docutils literal notranslate"><span class="pre">c.targets[['cutadapt']</span></code> might have the following values, after being
filled in with the sample table (see below for details):</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">c</span><span class="o">.</span><span class="n">targets</span><span class="p">[</span><span class="s1">&#39;cutadapt&#39;</span><span class="p">]</span>
<span class="c1"># data/rnaseq_samples/sample1/sample1_R1.cutadapt.fastq.gz</span>
<span class="c1"># data/rnaseq_samples/sample1/sample2_R1.cutadapt.fastq.gz</span>
<span class="c1"># data/rnaseq_samples/sample1/sample3_R1.cutadapt.fastq.gz</span>
<span class="c1"># data/rnaseq_samples/sample1/sample4_R1.cutadapt.fastq.gz</span>
</pre></div>
</div>
<p>This has several advantages:</p>
<ul class="simple">
<li><p>Patterns can be automatically filled in by the sample table and config file,
so the workflow is largely controlled by a TSV file and a YAML file.</p></li>
<li><p>Storing filenames outside individual Snakefiles allows us to access them much
more easily from other Snakefiles or downstream scripts.</p></li>
<li><p>Writing aggregation rules is much easier. For example, instead of lots of
<code class="docutils literal notranslate"><span class="pre">expand()</span></code> calls, we can get all the FastQC output across raw, trimmed, and
aligned runs with <code class="docutils literal notranslate"><span class="pre">flatten(c.targets[&quot;fastqc&quot;])</span></code>.</p></li>
<li><p>Toggling entire sections of the workflow can be performed by changing
a single line in the first <code class="docutils literal notranslate"><span class="pre">all</span></code> rule.</p></li>
<li><p>We can re-organize the output directories only by editing the patterns file
– no need to touch the Snakefile.</p></li>
<li><p>It’s easier to understand the output locations of files by looking at the
patterns file than it is to scroll through a Snakefile.</p></li>
<li><p>Letting the <code class="docutils literal notranslate"><span class="pre">c</span></code> objects do the work of filling in patterns allows complex
work to be abstracted away, resulting in simpler Snakefiles. For example,
filling in the output BED files across many arbitrary-named configured
peak-calling runs gets complicated, but since this is handled transparently
by the <code class="docutils literal notranslate"><span class="pre">c</span></code> object, we can do things like
<code class="docutils literal notranslate"><span class="pre">utils.flatten(c.targets['peaks'])</span></code> to get all the BED files for all
peak-callers and all peak-calling runs.</p></li>
</ul>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p>For more details, the code is the authoritative source.</p>
<p>In particular:</p>
<blockquote>
<div><ul class="simple">
<li><p><a class="reference internal" href="lib.patterns_targets.html#lib.patterns_targets.SeqConfig" title="lib.patterns_targets.SeqConfig"><code class="xref py py-class docutils literal notranslate"><span class="pre">lib.patterns_targets.SeqConfig</span></code></a></p></li>
<li><p><a class="reference internal" href="lib.patterns_targets.html#lib.patterns_targets.RNASeqConfig" title="lib.patterns_targets.RNASeqConfig"><code class="xref py py-class docutils literal notranslate"><span class="pre">lib.patterns_targets.RNASeqConfig</span></code></a></p></li>
<li><p><a class="reference internal" href="lib.patterns_targets.html#lib.patterns_targets.ChIPSeqConfig" title="lib.patterns_targets.ChIPSeqConfig"><code class="xref py py-class docutils literal notranslate"><span class="pre">lib.patterns_targets.ChIPSeqConfig</span></code></a></p></li>
</ul>
</div></blockquote>
<p>In addition, the <a class="reference external" href="https://github.com/lcdb/lcdb-wf/blob/master/workflows/figures/Snakefile">figures Snakefile</a>
demonstrates how the ChIP-seq and RNA-seq patterns and targets can be used
for downstream work.</p>
</div>
</section>
</section>


          </div>
          
        </div>
      </div>
    <div class="clearer"></div>
  </div>
    <div class="footer">
      &copy;2017, Ryan Dale, Justin Fear.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 4.0.2</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.12</a>
      
      |
      <a href="_sources/patterns-targets.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>