<!DOCTYPE html>

<html lang="en" data-content_root="./">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>Detailed documentation of RNA-Seq downstream &#8212; lcdb-wf 1.9 documentation</title>
    <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=fa44fd50" />
    <link rel="stylesheet" type="text/css" href="_static/basic.css?v=686e5160" />
    <link rel="stylesheet" type="text/css" href="_static/alabaster.css?v=9735eac1" />
    <script src="_static/documentation_options.js?v=754f84b7"></script>
    <script src="_static/doctools.js?v=9bcbadda"></script>
    <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="ChIP-seq workflow" href="chipseq.html" />
    <link rel="prev" title="RNA-Seq downstream analysis" href="downstream-rnaseq.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  

  
  

  </head><body>
  <div class="document">
    
      <div class="sphinxsidebar" role="navigation" aria-label="Main">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="toc.html">lcdb-wf</a></h1>



<p class="blurb">Customizable workflows for high-throughput sequencing analysis</p>






<h3>Navigation</h3>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="index.html">Introduction</a></li>
<li class="toctree-l1"><a class="reference internal" href="getting-started.html">Getting started</a></li>
<li class="toctree-l1"><a class="reference internal" href="guide.html">Guide to file hierarchy</a></li>
<li class="toctree-l1"><a class="reference internal" href="workflows.html">Overview of workflows</a></li>
<li class="toctree-l1"><a class="reference internal" href="config.html">Configuration</a></li>
<li class="toctree-l1"><a class="reference internal" href="references.html">References workflow</a></li>
<li class="toctree-l1"><a class="reference internal" href="rnaseq.html">RNA-seq workflow</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="downstream-rnaseq.html">RNA-Seq downstream analysis</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="downstream-rnaseq.html#how-to-use-this-code">How to use this code</a></li>
<li class="toctree-l2 current"><a class="reference internal" href="downstream-rnaseq.html#more-details">More details</a><ul class="current">
<li class="toctree-l3 current"><a class="current reference internal" href="#">Detailed documentation of RNA-Seq downstream</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="chipseq.html">ChIP-seq workflow</a></li>
<li class="toctree-l1"><a class="reference internal" href="integrative.html">Integrative workflows</a></li>
<li class="toctree-l1"><a class="reference internal" href="conda.html">conda and conda envs in <cite>lcdb-wf</cite></a></li>
<li class="toctree-l1"><a class="reference internal" href="tests.html">Testing the installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="faqs.html">FAQs</a></li>
<li class="toctree-l1"><a class="reference internal" href="changelog.html">Changelog</a></li>
<li class="toctree-l1"><a class="reference internal" href="developers.html">For Developers</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="toc.html">Documentation overview</a><ul>
  <li><a href="rnaseq.html">RNA-seq workflow</a><ul>
  <li><a href="downstream-rnaseq.html">RNA-Seq downstream analysis</a><ul>
      <li>Previous: <a href="downstream-rnaseq.html" title="previous chapter">RNA-Seq downstream analysis</a></li>
      <li>Next: <a href="chipseq.html" title="next chapter">ChIP-seq workflow</a></li>
  </ul></li>
  </ul></li>
  </ul></li>
</ul>
</div>
<search id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" autocomplete="off" autocorrect="off" autocapitalize="off" spellcheck="false"/>
      <input type="submit" value="Go" />
    </form>
    </div>
</search>
<script>document.getElementById('searchbox').style.display = "block"</script>
        </div>
      </div>
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <section id="detailed-documentation-of-rna-seq-downstream">
<span id="downstream-detailed"></span><h1>Detailed documentation of RNA-Seq downstream<a class="headerlink" href="#detailed-documentation-of-rna-seq-downstream" title="Link to this heading">¶</a></h1>
<p>Here we describe in detail the downstream analysis of RNA-Seq data performed in <code class="file docutils literal notranslate"><span class="pre">workflows/rnaseq/downstream/rnaseq.Rmd</span></code>.</p>
<p>This page has one section per named chunk. For example, if <code class="file docutils literal notranslate"><span class="pre">rnaseq.Rmd</span></code>
has the following code:</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">```{r load_libraries}</span>
<span class="n">library(DESeq2)</span>
<span class="n">library(dplyr)</span>
<span class="n">```</span>
</pre></div>
</div>
<p>Then you can find the corresponding documentation on this page under the
<code class="docutils literal notranslate"><span class="pre">load_libraries</span></code> heading.</p>
<p>The file <code class="file docutils literal notranslate"><span class="pre">ci/ensure_docs.py</span></code> double-checks to make sure all chunks are
documented and all documentation corresponds to a chunk, as part of the testing
framework.</p>
<section id="global-options">
<span id="rnaseqrmd"></span><h2><code class="docutils literal notranslate"><span class="pre">global_options</span></code><a class="headerlink" href="#global-options" title="Link to this heading">¶</a></h2>
<p>This chunk sets global rmarkdown options. Some of the lines provide
a mechanism for all cached chunks to have their cache invalidated if either of
the filenames’ modification times have changed. For example, the following
argument to <code class="docutils literal notranslate"><span class="pre">knitr::opts_chunk$set</span></code> will inject the option
<code class="docutils literal notranslate"><span class="pre">cache.extra_file_dep_1</span></code> into all chunks:</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">cache.extra_file_dep_1</span><span class="o">=</span><span class="nf">file.info</span><span class="p">(</span><span class="s">&#39;../config/sampletable.tsv&#39;</span><span class="p">)</span><span class="o">$</span><span class="n">mtime</span><span class="p">,</span>
</pre></div>
</div>
<p>Therefore if the sample table of the RNA-Seq workflow has been updated since
the last time the analysis script was run, the modification time (mtime) will
be changed, so the <code class="docutils literal notranslate"><span class="pre">cache.extra_file_dep_1</span></code> value will be different than it
was before for every chunk, and so every chunk will be re-run.</p>
</section>
<section id="lcdbwf">
<h2><code class="docutils literal notranslate"><span class="pre">lcdbwf</span></code><a class="headerlink" href="#lcdbwf" title="Link to this heading">¶</a></h2>
<p>Loads the <code class="docutils literal notranslate"><span class="pre">lcdbwf</span></code> R package, stored in <code class="docutils literal notranslate"><span class="pre">../../../lib/lcdbwf/R</span></code>. This chunk
is not cached and fully reloads the package each time using
<code class="docutils literal notranslate"><span class="pre">devtools::load_all</span></code>, so any changes to the code in that package will show up
when this file is rendered. Documentation is also automatically re-generated.</p>
<p>Note that throughout this RMarkdown file, functions from this package will use
the <code class="docutils literal notranslate"><span class="pre">lcdbwf::</span></code> namespace prefix to be explicit about where that function is
coming from.</p>
</section>
<section id="config">
<h2><code class="docutils literal notranslate"><span class="pre">config</span></code><a class="headerlink" href="#config" title="Link to this heading">¶</a></h2>
<p>This chunk loads the config files <code class="file docutils literal notranslate"><span class="pre">config.yaml</span></code> and <code class="file docutils literal notranslate"><span class="pre">text.yaml</span></code>.
This chunk is not cached, so any changes in the config files automatically show
up here.</p>
<p>See the <code class="file docutils literal notranslate"><span class="pre">config.yaml</span></code> file for configuring the code.</p>
<p>See the <code class="file docutils literal notranslate"><span class="pre">text.yaml</span></code> file for editing the explanatory text.</p>
<p>This chunk also configures the parallelization options in a chunk that is not
cached.</p>
</section>
<section id="libraries">
<h2><code class="docutils literal notranslate"><span class="pre">libraries</span></code><a class="headerlink" href="#libraries" title="Link to this heading">¶</a></h2>
<p>Standard loading of the library dependencies used throughout the code.</p>
</section>
<section id="coldata-setup">
<h2><code class="docutils literal notranslate"><span class="pre">coldata_setup</span></code><a class="headerlink" href="#coldata-setup" title="Link to this heading">¶</a></h2>
<p>This chunk loads the sample table.</p>
<p>Use this chunk to add additional columns to your sampletable.</p>
<p>The only requirement is that the rownames correspond to sample names.</p>
</section>
<section id="dds-initial">
<h2><code class="docutils literal notranslate"><span class="pre">dds_initial</span></code><a class="headerlink" href="#dds-initial" title="Link to this heading">¶</a></h2>
<p>Constructs the initial dds object and the variance stabilized counts.</p>
<p>Note that the design used is <code class="docutils literal notranslate"><span class="pre">~1</span></code> and the call to
<code class="docutils literal notranslate"><span class="pre">varianceStabilizingTransformation</span></code> uses <code class="docutils literal notranslate"><span class="pre">blind=TRUE</span></code>, so you don’t need to
change anything here.</p>
<p>Also note that the entire <code class="docutils literal notranslate"><span class="pre">config</span></code> object is passed to the <code class="docutils literal notranslate"><span class="pre">make_dds</span></code>
function, which reads options like whether to strip version numbers off of gene
IDs or whether (and how) to collapse technical replicates.</p>
</section>
<section id="print-coldata">
<h2><code class="docutils literal notranslate"><span class="pre">print_coldata</span></code><a class="headerlink" href="#print-coldata" title="Link to this heading">¶</a></h2>
<p>Simply prints the colData for reference – excluding columns that might be in
there that would clutter the output.</p>
</section>
<section id="sample-heatmap">
<h2><code class="docutils literal notranslate"><span class="pre">sample_heatmap</span></code><a class="headerlink" href="#sample-heatmap" title="Link to this heading">¶</a></h2>
<p>This chunk creates a clustered heatmap of sample distances. The columns
specified in the config file’s “covariates_for_plots” item will show up as
colors along the right side.</p>
</section>
<section id="pca">
<h2><code class="docutils literal notranslate"><span class="pre">pca</span></code><a class="headerlink" href="#pca" title="Link to this heading">¶</a></h2>
<p>Creates PCA plots, one tab per entry in the config file’s
“covariates_for_plots” item. Each plot has the same points, but the colors
differ. This can help assess the experimental design and set expectations on
how many differentially expressed genes one may find.</p>
<p>These are interactive plots, and hovering over a point indicates the sample.</p>
</section>
<section id="sizefactors">
<h2><code class="docutils literal notranslate"><span class="pre">sizefactors</span></code><a class="headerlink" href="#sizefactors" title="Link to this heading">¶</a></h2>
<p>This chunk makes diagnostic plots. In general, we expect sizeFactors to
correlate with total read count. When it doesn’t, it can indicate that a small
number of genes are very highly expressed.</p>
</section>
<section id="dds-list">
<span id="id1"></span><h2><code class="docutils literal notranslate"><span class="pre">dds_list</span></code><a class="headerlink" href="#dds-list" title="Link to this heading">¶</a></h2>
<p>This chunk sets up the <a class="reference internal" href="#term-dds"><span class="xref std std-term">dds</span></a> objects to be used in the <cite>results</cite> section
below for differential expression detection.</p>
<p>You may need different <code class="docutils literal notranslate"><span class="pre">dds</span></code> objects for testing different models, or perhaps
removing outlier samples. If you have technical replicates you might need to
combine them, and you might need to remove gene version identifiers. You might
want to use salmon instead of featureCounts. These would need to be done for
each <code class="docutils literal notranslate"><span class="pre">dds</span></code>, requiring code duplication.</p>
<p>After working on many complex and/or messy experimental designs, we have
settled on the approach of a named list of <code class="docutils literal notranslate"><span class="pre">dds</span></code> objects, where later code
refers to these objects by their name in the list.</p>
<p><strong>The</strong> <code class="docutils literal notranslate"><span class="pre">results</span></code> <strong>chunk below expects such a list.</strong></p>
<p>The simplest example is the following where we create a single <code class="docutils literal notranslate"><span class="pre">dds</span></code> and put
it into a list.</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">dds</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">DESeqFromCombinedFeatureCounts</span><span class="p">(</span>
<span class="w">   </span><span class="s">&#39;../data/rnaseq_aggregation/featurecounts.txt&#39;</span><span class="p">,</span>
<span class="w">   </span><span class="n">sampletable</span><span class="o">=</span><span class="n">colData</span><span class="p">,</span>
<span class="w">   </span><span class="n">design</span><span class="o">=~</span><span class="n">group</span><span class="p">)</span>
<span class="n">dds</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">DESeq</span><span class="p">(</span><span class="n">dds</span><span class="p">,</span><span class="w"> </span><span class="n">parallel</span><span class="o">=</span><span class="n">parallel</span><span class="p">)</span>

<span class="n">dds.list</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">list</span><span class="p">(</span><span class="n">main</span><span class="o">=</span><span class="n">dds</span><span class="p">)</span>
</pre></div>
</div>
<p>Now imagine a case where we want to remove a replicate that we think is an
outlier, but we still want to compare it to the results when it is included.
Let’s say we also need to collapse the technical replicates. Such code would
look like this:</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># The long way...</span>
<span class="c1">#</span>
<span class="c1"># First object with all replicates</span>
<span class="n">dds1</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">DESeqFromCombinedFeatureCounts</span><span class="p">(</span>
<span class="w">   </span><span class="s">&#39;../data/rnaseq_aggregation/featurecounts.txt&#39;</span><span class="p">,</span>
<span class="w">   </span><span class="n">sampletable</span><span class="o">=</span><span class="n">colData</span><span class="p">,</span>
<span class="w">   </span><span class="n">design</span><span class="o">=~</span><span class="n">group</span><span class="p">)</span>
<span class="n">dds1</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">collapseReplicates</span><span class="p">(</span><span class="n">dds1</span><span class="p">,</span><span class="w"> </span><span class="s">&#39;biorep&#39;</span><span class="p">)</span>
<span class="n">dds1</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">DESeq</span><span class="p">(</span><span class="n">dds1</span><span class="p">,</span><span class="w"> </span><span class="n">parallel</span><span class="o">=</span><span class="n">parallel</span><span class="p">)</span>

<span class="c1"># Similar to above, but remove replicate 4</span>
<span class="n">dds2</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">DESeqFromCombinedFeatureCounts</span><span class="p">(</span>
<span class="w">   </span><span class="s">&#39;../data/rnaseq_aggregation/featurecounts.txt&#39;</span><span class="p">,</span>
<span class="w">   </span><span class="n">sampletable</span><span class="o">=</span><span class="n">colData</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span><span class="nf">filter</span><span class="p">(</span><span class="n">replicate</span><span class="o">!=</span><span class="s">&#39;rep4&#39;</span><span class="p">),</span>
<span class="w">   </span><span class="n">design</span><span class="o">=~</span><span class="n">group</span><span class="p">,</span>
<span class="w">   </span><span class="c1"># need subset_counts=TRUE if we want to automatically</span>
<span class="w">   </span><span class="c1"># subset the featureCounts to match the filtered colData</span>
<span class="w">   </span><span class="c1"># we provided.</span>
<span class="w">   </span><span class="n">subset_counts</span><span class="o">=</span><span class="kc">TRUE</span>
<span class="w">   </span><span class="p">)</span>
<span class="n">dds2</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">collapseReplicates</span><span class="p">(</span><span class="n">dds</span><span class="p">,</span><span class="w"> </span><span class="s">&#39;biorep&#39;</span><span class="p">)</span>
<span class="n">dds2</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">DESeq</span><span class="p">(</span><span class="n">dds2</span><span class="p">,</span><span class="w"> </span><span class="n">parallel</span><span class="o">=</span><span class="n">parallel</span><span class="p">)</span>
</pre></div>
</div>
<p>Based on our experience, as we add more <code class="docutils literal notranslate"><span class="pre">dds</span></code> objects the code gets more
error-prone. So for more complex use-cases, we have a function
<code class="docutils literal notranslate"><span class="pre">lcdbwf::make_dds</span></code>.</p>
<p>Here is how the code above would look using this method:</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">lst</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">list</span><span class="p">(</span>

<span class="w">   </span><span class="n">main</span><span class="o">=</span><span class="nf">list</span><span class="p">(</span><span class="n">sampletable</span><span class="o">=</span><span class="n">colData</span><span class="p">,</span><span class="w"> </span><span class="n">design</span><span class="o">=~</span><span class="n">group</span><span class="p">),</span>

<span class="w">   </span><span class="n">no.rep.4</span><span class="o">=</span><span class="nf">list</span><span class="p">(</span>
<span class="w">      </span><span class="n">sampletable</span><span class="o">=</span><span class="n">colData</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span><span class="nf">filter</span><span class="p">(</span><span class="n">replicate</span><span class="o">!=</span><span class="s">&#39;rep4&#39;</span><span class="p">),</span>
<span class="w">      </span><span class="n">design</span><span class="o">=~</span><span class="n">group</span><span class="p">,</span>
<span class="w">      </span><span class="n">subset.counts</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">))</span>
<span class="p">)</span>

<span class="n">dds_list</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">map</span><span class="p">(</span><span class="n">lst</span><span class="p">,</span><span class="w"> </span><span class="n">lcdbwf</span><span class="o">::</span><span class="n">make_dds</span><span class="p">,</span><span class="w"> </span><span class="n">config</span><span class="o">=</span><span class="n">config</span><span class="p">,</span><span class="w"> </span><span class="n">parallel</span><span class="o">=</span><span class="n">config</span><span class="o">$</span><span class="n">parallel</span><span class="o">$</span><span class="n">parallel</span><span class="p">)</span>
</pre></div>
</div>
<p>That is, first we create a list of lists (<code class="docutils literal notranslate"><span class="pre">lst</span></code>), and then we used <code class="docutils literal notranslate"><span class="pre">map()</span></code> to apply
the <code class="docutils literal notranslate"><span class="pre">make_dds</span></code> function to all items in the list. The collapsing of
replicates and other dds-creation configuration like stripping dotted version
names is determined by the config object which is passed along.</p>
<p>See the help for <code class="docutils literal notranslate"><span class="pre">lcdbwf::make_dds</span></code> for more details.</p>
<p>This chunk becomes a dependency of all of the <code class="docutils literal notranslate"><span class="pre">results</span></code> chunks below.</p>
</section>
<section id="dds-diagnostics">
<h2><code class="docutils literal notranslate"><span class="pre">dds_diagnostics</span></code><a class="headerlink" href="#dds-diagnostics" title="Link to this heading">¶</a></h2>
<p>If configured, this chunk will run the diagnostics on the dds objects and show
tabbed reports on each dds.</p>
</section>
<section id="results">
<h2><code class="docutils literal notranslate"><span class="pre">results_*</span></code><a class="headerlink" href="#results" title="Link to this heading">¶</a></h2>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This is where most of the customization needs to happen for each project.</p>
</div>
<p>This is actually a series of chunks where the bulk of the differential
expression analysis takes place.</p>
<p>For simple cases, you probably just need one of these. But for complex
experimental designs where you end up doing lots of contrasts, it can get time
consuming to run them every time you change the RMarkdown file.</p>
<p>The end result of these chunks is a single list containing DESeq2 results
objects and associated metadata in (sub)lists. Each of these sublists has:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">res</span></code>, the results object</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">dds</span></code>, the string name in <code class="docutils literal notranslate"><span class="pre">names(dds.list)</span></code></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">label</span></code>, a “nice” label which is used for headings and other output</p></li>
<li><p>additional optional arguments that are passed along to <code class="docutils literal notranslate"><span class="pre">DESeq2::results()</span></code>
and/or <code class="docutils literal notranslate"><span class="pre">DESeq2::lfcshrink()</span></code></p></li>
</ul>
<p>To continue our example from above, we might want to run the same contrast on
all samples (the “main” dds) and after removing replicate 4 (the “no.rep.4”
dds). To illustrate how additional arguments are used, let’s imagine we also
want to use <cite>ashr</cite> as the shrinkage method for the second contrast.</p>
<p>Use the <code class="docutils literal notranslate"><span class="pre">lcdbwf::make_results()</span></code> function for this. This function is
a loose wrapper around <code class="docutils literal notranslate"><span class="pre">DESeq2::results()</span></code> and <code class="docutils literal notranslate"><span class="pre">DESeq2::lfcshrink()</span></code> that
adds some extra convenience when working with lists of dds objects, including
the detection of parallelization as set up in the config object. See the help
for <code class="docutils literal notranslate"><span class="pre">lcdbwf::make_results()</span></code> for more details.</p>
<p>By default, if no test argument is specified in the parameters for
<code class="docutils literal notranslate"><span class="pre">lcdbwf::make_dds</span></code> found here in <a class="reference external" href="https://github.com/lcdb/lcdb-wf/blob/LRT/workflows/rnaseq/downstream/rnaseq.Rmd#L164-L187">examples 1-4,</a>
the Wald test is performed. When <code class="docutils literal notranslate"><span class="pre">lcdbwf::make_results</span></code> processes a Wald test dds object, it
detects the Wald test and expects a <code class="docutils literal notranslate"><span class="pre">contrast</span></code> or <code class="docutils literal notranslate"><span class="pre">coef</span></code> argument to specify which
p-values and log2FoldChange values to report.</p>
<p>DESeq2 also supports the nBinomLRT (LRT). <a class="reference external" href="https://github.com/lcdb/lcdb-wf/blob/LRT/workflows/rnaseq/downstream/rnaseq.Rmd#L189-L194">Example 5</a>
demonstrates how to create a dds object with LRT data. Since the LRT tests
the removal of one or more terms from the design formula, a single
log2FoldChange column doesn’t reflect the test’s complexity. DESeq2’s results
object is optimized for the Wald test, and when storing LRT results, it
maintains consistency in datastructure by choosing a single pair-wise comparison for
log2FoldChange values. To avoid confusion, <strong>*we set all log2FoldChange values to
0 for LRT results*</strong>.</p>
<p>For more details, see the DESeq2 documentation:
<a class="reference external" href="https://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#i-ran-a-likelihood-ratio-test-but-results-only-gives-me-one-comparison">DESeq2 Likelihood Ratio Test</a>.</p>
<p id="rules">To take advantage of this infrastructure, we put each of those contrasts into
its own chunk <strong>according to the following rules</strong>:</p>
<ul class="simple">
<li><p>the chunk name must start with <code class="docutils literal notranslate"><span class="pre">results_</span></code></p></li>
<li><p>the chunk is cached</p></li>
<li><p>the chunk depends on the <code class="docutils literal notranslate"><span class="pre">'dds_list'</span></code> chunk</p></li>
<li><p>the variable name starts with <code class="docutils literal notranslate"><span class="pre">contr_[index]_</span></code>, and the rest of the variable name
will be used as the name in the list. index is a string, containing 1 or more alphanumeric
characters, and will be used as a sorting index for contrasts when generating output files.
Note that the index string cannot contain “_”.</p></li>
</ul>
<p>Our example would look like the following. Note that we’re showing the chunks
here because that will be come meaningful in a moment. They are shown as
comments here just to get the syntax highlighting to look OK.</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># ```{r results_01, dependson=&#39;dds_list&#39;, cache=TRUE}</span>
<span class="n">contr_1_ko.vs.wt</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">lcdbwf</span><span class="o">::</span><span class="nf">make_results</span><span class="p">(</span>
<span class="w">  </span><span class="n">dds_name</span><span class="o">=</span><span class="s">&#39;main&#39;</span><span class="p">,</span>
<span class="w">  </span><span class="n">label</span><span class="o">=</span><span class="s">&#39;Using all samples&#39;</span><span class="p">,</span>
<span class="w">  </span><span class="n">contrast</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="s">&#39;genotype&#39;</span><span class="p">,</span><span class="w"> </span><span class="s">&#39;KO&#39;</span><span class="p">,</span><span class="w"> </span><span class="s">&#39;WT&#39;</span><span class="p">)</span>
<span class="p">)</span>
<span class="c1"># ```</span>

<span class="c1"># ```{r results_02, dependson=&#39;dds_list&#39;, cache=TRUE}</span>
<span class="n">contr_2a_no.rep.4</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">lcdbwf</span><span class="o">::</span><span class="nf">make_results</span><span class="p">(</span>
<span class="w">  </span><span class="n">dds_name</span><span class="o">=</span><span class="s">&#39;no.rep.4&#39;</span><span class="p">,</span>
<span class="w">  </span><span class="n">label</span><span class="o">=</span><span class="s">&#39;Removing replicate 4 and using ashr for shrinkage&#39;</span><span class="p">,</span>
<span class="w">  </span><span class="n">contrast</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="s">&#39;genotype&#39;</span><span class="p">,</span><span class="w"> </span><span class="s">&#39;KO&#39;</span><span class="p">,</span><span class="w"> </span><span class="s">&#39;WT&#39;</span><span class="p">),</span>
<span class="w">  </span><span class="n">type</span><span class="o">=</span><span class="s">&#39;ashr&#39;</span>
<span class="p">)</span>
<span class="c1"># ```</span>
</pre></div>
</div>
<p>When combined with the <code class="docutils literal notranslate"><span class="pre">assemble_variables</span></code> chunk described below, this
allows us to:</p>
<ul class="simple">
<li><p>retain caching at the level of individual contrasts</p></li>
<li><p>combine all results into a single list used in later code while still respecting dependencies</p></li>
<li><p>reduce the bookkeeping overhead</p></li>
</ul>
<p>In our experience this scales well with very complex experimental designs with
lots of contrasts. For more information on creating complex contrasts, see
<a class="reference internal" href="#contrast"><span class="std std-ref">Specifying contrasts</span></a>. For more information on how these results are collected, see <a class="reference internal" href="#assemble-variables"><span class="std std-ref">assemble_variables</span></a>.</p>
</section>
<section id="assemble-variables">
<span id="id2"></span><h2><code class="docutils literal notranslate"><span class="pre">assemble_variables</span></code><a class="headerlink" href="#assemble-variables" title="Link to this heading">¶</a></h2>
<p>If we had put all <code class="docutils literal notranslate"><span class="pre">results()</span></code> calls into the same chunk and cached that, then
a change anywhere in that chunk would invalidate the cache which would cause
all results to be regenerated. With many contrasts, this can get quite
time-consuming. An alternative would be to put each <code class="docutils literal notranslate"><span class="pre">results()</span></code> call into its
own chunk. But then we would need to keep track of dependencies and ensure
those dependencies were specified in downstream chunks.</p>
<p>For example, if you add a new results chunk and cache it, but forget to add
that chunk as a dependency in a later chunk, that later chunk will be
inconsistent and may even be missing the new results. Keeping track of this can
be error-prone.</p>
<p>Our solution is to set up the contrasts according to the <a class="reference internal" href="#rules"><span class="std std-ref">rules described
above</span></a>. By following those rules, the following becomes possible:</p>
<ul class="simple">
<li><p>we can detect all chunks creating results by looking for <code class="docutils literal notranslate"><span class="pre">results_</span></code> in the
chunk name and automatically inject these into dependencies of future chunks.</p></li>
<li><p>we can detect all results objects created by looking for variables starting
with <code class="docutils literal notranslate"><span class="pre">contr_[index]_</span></code></p></li>
<li><p>we can assemble all results objects into a list, and name each item in the
list according to its variable name (minus the <code class="docutils literal notranslate"><span class="pre">contr_[index]_</span></code>).</p></li>
<li><p>we can alter the order of contrasts by simply modifying the index string in a single chunk.
For example, if we have three contrasts contr_1_ko.vs.wt, contr_2_no.rep.4, and contr_3_no.rep.3,
we can change the order of contrasts simply by modifying one index string (ex: change contr_3_no.rep.3 to
contr_1a_no.rep.3).</p></li>
</ul>
<p>The <code class="docutils literal notranslate"><span class="pre">assemble_variables</span></code> chunk does all of this. The end result of this chunk
is a list of lists that is used by functions in the <cite>lcdbwf</cite> R package for
downstream work. For more details, see <a class="reference internal" href="#term-res_list"><span class="xref std std-term">res_list</span></a>.</p>
<p>For each contrast (that is, each entry in <cite>res_list</cite>) the below chunks will
automatically create a DE results section including:</p>
<ul class="simple">
<li><p>a tabbed section using the label as a header</p></li>
<li><p>summary table</p></li>
<li><p>MA plot</p></li>
<li><p>counts plots of top 3 up- and down-regulated genes</p></li>
<li><p>p-value distribution</p></li>
<li><p>exported results tables with links</p></li>
</ul>
<section id="specifying-contrasts">
<span id="contrast"></span><h3>Specifying contrasts<a class="headerlink" href="#specifying-contrasts" title="Link to this heading">¶</a></h3>
<p>Contrasts can be specified in three different ways.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>In these examples, “control” and “treatment” are factor levels in the
“group” factor (which was in the <a class="reference internal" href="#term-colData"><span class="xref std std-term">colData</span></a>), and the <a class="reference internal" href="#term-dds"><span class="xref std std-term">dds</span></a>
object was created with the design <code class="docutils literal notranslate"><span class="pre">~group</span></code>:</p>
</div>
<ol class="arabic">
<li><p>A character vector to the <cite>contrast</cite> parameter.</p>
<p>This should be a three element vector:</p>
<ul class="simple">
<li><p>the name of a factor in the design formula</p></li>
<li><p>name of the numerator for the fold change</p></li>
<li><p>the name of the denominator for the fold change. E.g.,</p></li>
</ul>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">res</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">results</span><span class="p">(</span><span class="n">dds</span><span class="p">,</span><span class="w"> </span><span class="n">contrast</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="s">&#39;group&#39;</span><span class="p">,</span><span class="w"> </span><span class="s">&#39;treatment&#39;</span><span class="p">,</span><span class="w"> </span><span class="s">&#39;control&#39;</span><span class="p">)</span>
</pre></div>
</div>
<p>That is, <strong>the control must be last</strong>.</p>
</li>
<li><p><cite>name</cite> parameter for <code class="docutils literal notranslate"><span class="pre">results()</span></code> function call or <cite>coef</cite> parameter for
<code class="docutils literal notranslate"><span class="pre">lfcShrink()</span></code> call</p>
<p><cite>name</cite> or <cite>coef</cite> should be one of the values returned by
<code class="docutils literal notranslate"><span class="pre">resultsNames(dds)</span></code> that corresponds to the precomputed results. E.g.</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">resultsNames</span><span class="p">(</span><span class="n">dds</span><span class="p">)</span>
<span class="c1"># [1] &quot;Intercept&quot;  &quot;group_treatment_vs_control&quot;</span>

<span class="n">res</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">results</span><span class="p">(</span><span class="n">dds</span><span class="p">,</span><span class="w"> </span><span class="n">name</span><span class="o">=</span><span class="s">&#39;group_treatment_vs_control&#39;</span><span class="p">)</span>
</pre></div>
</div>
</li>
<li><p>A numeric contrast vector with one element for each element in the
<code class="docutils literal notranslate"><span class="pre">resultsNames()</span></code> function call. This is useful for arbitrary comparisons
in multi-factor designs with a grouping variable.</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">resultsNames</span><span class="p">(</span><span class="n">dds</span><span class="p">)</span>
<span class="c1"># [1] &quot;Intercept&quot;  &quot;group_treatment_vs_control&quot;</span>

<span class="n">res</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">results</span><span class="p">(</span><span class="n">dds</span><span class="p">,</span><span class="w"> </span><span class="n">contrast</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="m">0</span><span class="p">,</span><span class="w"> </span><span class="m">1</span><span class="p">))</span>
</pre></div>
</div>
</li>
</ol>
</section>
<section id="the-most-general-way-to-specify-contrasts">
<h3>The most general way to specify contrasts<a class="headerlink" href="#the-most-general-way-to-specify-contrasts" title="Link to this heading">¶</a></h3>
<p>The most general way to specify contrasts is with a numeric vector (third
option above).</p>
<p>Here is a worked example, using a two-factor experiment.</p>
<p><cite>group</cite> encodes all combinations of a two-factor experiment, so we construct
a sampletable that looks like the following (here, showing 2 replicates per
group):</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">sample</span>   <span class="n">genotype</span>   <span class="n">condition</span>   <span class="n">group</span>
<span class="mi">1</span>        <span class="n">A</span>          <span class="n">I</span>           <span class="n">IA</span>
<span class="mi">2</span>        <span class="n">A</span>          <span class="n">I</span>           <span class="n">IA</span>
<span class="mi">3</span>        <span class="n">B</span>          <span class="n">I</span>           <span class="n">IB</span>
<span class="mi">4</span>        <span class="n">B</span>          <span class="n">I</span>           <span class="n">IB</span>
<span class="mi">5</span>        <span class="n">A</span>          <span class="n">II</span>          <span class="n">IIA</span>
<span class="mi">6</span>        <span class="n">A</span>          <span class="n">II</span>          <span class="n">IIA</span>
<span class="mi">7</span>        <span class="n">B</span>          <span class="n">II</span>          <span class="n">IIB</span>
<span class="mi">8</span>        <span class="n">B</span>          <span class="n">II</span>          <span class="n">IIB</span>
</pre></div>
</div>
<p>We can make arbitrary comparisons by fitting an ‘intercept-less’ model, e.g.
<code class="docutils literal notranslate"><span class="pre">design=~group</span> <span class="pre">+</span> <span class="pre">0</span></code>, and numeric contrast vectors:</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">dds</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">DESeqDataSetFromCombinedFeatureCounts</span><span class="p">(</span>
<span class="w">    </span><span class="s">&#39;../data/rnaseq_aggregation/featurecounts.txt&#39;</span><span class="p">,</span>
<span class="w">    </span><span class="n">sampletable</span><span class="o">=</span><span class="n">colData</span><span class="p">,</span>
<span class="w">    </span><span class="c1"># NOTE: the design is now different</span>
<span class="w">    </span><span class="n">design</span><span class="o">=~</span><span class="n">group</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="m">0</span>
<span class="p">)</span>
<span class="n">dds</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">DESeq</span><span class="p">(</span><span class="n">dds</span><span class="p">)</span>
</pre></div>
</div>
<p>Check <code class="docutils literal notranslate"><span class="pre">resultsNames</span></code>:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">resultsNames</span><span class="p">(</span><span class="n">dds</span><span class="p">)</span>
<span class="c1"># [1] &quot;groupIA&quot;  &quot;groupIB&quot;  &quot;groupIIA&quot;  &quot;groupIIB&quot;</span>
</pre></div>
</div>
<p>So any numeric vectors we provide must be 4 items long. Here is how we can make
various contrasts with this experimental design. In each example, the
coefficients are indicated above the resultsNames to make it easier to see.</p>
<p>To compare IA and IB (that is, the genotype effect only in condition I):</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1">#     1          -1         0           0</span>
<span class="c1"># &quot;groupIA&quot;  &quot;groupIB&quot;  &quot;groupIIA&quot;  &quot;groupIIB&quot;</span>

<span class="n">res</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">results</span><span class="p">(</span><span class="n">dds</span><span class="p">,</span><span class="w"> </span><span class="n">contrast</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="m">-1</span><span class="p">,</span><span class="w"> </span><span class="m">0</span><span class="p">,</span><span class="w"> </span><span class="m">0</span><span class="p">)</span>
</pre></div>
</div>
<p>Effect of genotype B (that is, disregard information about condition):</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1">#     1          -1         1           -1</span>
<span class="c1"># &quot;groupIA&quot;  &quot;groupIB&quot;  &quot;groupIIA&quot;  &quot;groupIIB&quot;</span>

<span class="n">res</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">results</span><span class="p">(</span><span class="n">dds</span><span class="p">,</span><span class="w"> </span><span class="n">contrast</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="m">-1</span><span class="p">,</span><span class="w"> </span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="m">-1</span><span class="p">)</span>
</pre></div>
</div>
<p>Effect of condition II (that is, disregard information about genotype):</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1">#     1           1         -1          -1</span>
<span class="c1"># &quot;groupIA&quot;  &quot;groupIB&quot;  &quot;groupIIA&quot;  &quot;groupIIB&quot;</span>

<span class="n">res</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">results</span><span class="p">(</span><span class="n">dds</span><span class="p">,</span><span class="w"> </span><span class="n">contrast</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="m">-1</span><span class="p">,</span><span class="w"> </span><span class="m">-1</span><span class="p">)</span>
</pre></div>
</div>
<p>Interaction term, that is, (IA vs IB) vs (IIA vs IIB). This is effectively <code class="docutils literal notranslate"><span class="pre">(IA</span>
<span class="pre">-</span> <span class="pre">IB)</span> <span class="pre">-</span> <span class="pre">(IIA</span> <span class="pre">-</span> <span class="pre">IIB)</span></code>, which in turn becomes <code class="docutils literal notranslate"><span class="pre">IA</span> <span class="pre">-</span> <span class="pre">IB</span> <span class="pre">-</span> <span class="pre">IIA</span> <span class="pre">+</span> <span class="pre">IIB</span></code>:</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1">#     1          -1         -1          1</span>
<span class="c1"># &quot;groupIA&quot;  &quot;groupIB&quot;  &quot;groupIIA&quot;  &quot;groupIIB&quot;</span>

<span class="n">res</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">results</span><span class="p">(</span><span class="n">dds</span><span class="p">,</span><span class="w"> </span><span class="n">contrast</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="m">-1</span><span class="p">,</span><span class="w"> </span><span class="m">-1</span><span class="p">,</span><span class="w"> </span><span class="m">1</span><span class="p">)</span>
</pre></div>
</div>
</section>
</section>
<section id="summary">
<h2><code class="docutils literal notranslate"><span class="pre">summary</span></code><a class="headerlink" href="#summary" title="Link to this heading">¶</a></h2>
<p>This chunk prints a high-level overview of all the contrasts.</p>
</section>
<section id="reportresults">
<h2><code class="docutils literal notranslate"><span class="pre">reportresults</span></code><a class="headerlink" href="#reportresults" title="Link to this heading">¶</a></h2>
<p>This is the section that creates multiple, tabbed outputs for each of the
contrasts in the <a class="reference internal" href="#term-res_list"><span class="xref std std-term">res_list</span></a>.</p>
<p>If the config specifies results diagnostics
(<code class="docutils literal notranslate"><span class="pre">config$toggle$results_diagnostics</span></code> is TRUE), then this chunk will also run
the diagnostics. You can select just the ones you want diagnostics on using the
<code class="docutils literal notranslate"><span class="pre">config$plotting$diagnostics_results_names</span></code> config option.</p>
</section>
<section id="upsetplots">
<h2><code class="docutils literal notranslate"><span class="pre">upsetplots</span></code><a class="headerlink" href="#upsetplots" title="Link to this heading">¶</a></h2>
<p>This chunk produces UpSet plots comparing the contrasts.</p>
</section>
<section id="excel">
<h2><code class="docutils literal notranslate"><span class="pre">excel</span></code><a class="headerlink" href="#excel" title="Link to this heading">¶</a></h2>
<p>This chunk outputs an Excel spreadsheet with one contrast per sheet. Normalized
counts for each sample, from the respective dds object used for the contrast,
are also included on each sheet.</p>
</section>
<section id="write-output">
<h2><code class="docutils literal notranslate"><span class="pre">write_output</span></code><a class="headerlink" href="#write-output" title="Link to this heading">¶</a></h2>
<p>TSVs for each contrast’s results are written to disk.</p>
</section>
<section id="combined-rds">
<h2><code class="docutils literal notranslate"><span class="pre">combined_rds</span></code><a class="headerlink" href="#combined-rds" title="Link to this heading">¶</a></h2>
<p>A single object is written as an .Rds file. This can then be used for
downstream visualization or it can be used as input to the functional
enrichment RMarkdown document.</p>
</section>
<section id="sessioninfo">
<h2><code class="docutils literal notranslate"><span class="pre">sessioninfo</span></code><a class="headerlink" href="#sessioninfo" title="Link to this heading">¶</a></h2>
<p>The output of sessionInfo records the versions of packages used in the analysis.</p>
</section>
<section id="glossary">
<h2>Glossary<a class="headerlink" href="#glossary" title="Link to this heading">¶</a></h2>
<dl class="simple glossary">
<dt id="term-colData">colData<a class="headerlink" href="#term-colData" title="Link to this term">¶</a></dt><dd><p>The metadata describing the samples. This is originally defined in the
sampletable for the entire lcdb-wf run, is imported into rnaseq.Rmd, and
may be subsequently modified.</p>
</dd>
<dt id="term-dds">dds<a class="headerlink" href="#term-dds" title="Link to this term">¶</a></dt><dd><p>DESeq data set object. Typically this is incrementally added to, as in
the DESeq2 vignette.</p>
</dd>
<dt id="term-vsd">vsd<a class="headerlink" href="#term-vsd" title="Link to this term">¶</a></dt><dd><p>The variance-stabilized transformed version of the counts. Used for PCA,
clustered heatmaps, and gene patterns.</p>
</dd>
<dt id="term-res_list">res_list<a class="headerlink" href="#term-res_list" title="Link to this term">¶</a></dt><dd><p>A list, with one item per contrast. Each of those items in turn is a list
of objects that together compose the contrast (dds name, results object, and
label). This list-of-lists, which we call <cite>res_list</cite> for short, is used
by functions in the <cite>lcdbwf</cite> R package for more downstream work, like
gene patterns, functional enrichment, and Shiny apps.</p>
</dd>
</dl>
</section>
</section>


          </div>
          
        </div>
      </div>
    <div class="clearer"></div>
  </div>
    <div class="footer">
      &#169;2017, Ryan Dale, Justin Fear.
      
      |
      Powered by <a href="https://www.sphinx-doc.org/">Sphinx 8.1.0</a>
      &amp; <a href="https://alabaster.readthedocs.io">Alabaster 1.0.0</a>
      
      |
      <a href="_sources/rnaseq-rmd.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>