<!DOCTYPE html>

<html lang="en" data-content_root="./">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>Sample tables &#8212; lcdb-wf 1.9 documentation</title>
    <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=fa44fd50" />
    <link rel="stylesheet" type="text/css" href="_static/basic.css?v=b08954a9" />
    <link rel="stylesheet" type="text/css" href="_static/alabaster.css?v=9735eac1" />
    <script src="_static/documentation_options.js?v=754f84b7"></script>
    <script src="_static/doctools.js?v=9bcbadda"></script>
    <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Patterns and targets" href="patterns-targets.html" />
    <link rel="prev" title="References config" href="references-config.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  

  
  

  </head><body>
  <div class="document">
    
      <div class="sphinxsidebar" role="navigation" aria-label="Main">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="toc.html">lcdb-wf</a></h1>



<p class="blurb">Customizable workflows for high-throughput sequencing analysis</p>






<h3>Navigation</h3>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="index.html">Introduction</a></li>
<li class="toctree-l1"><a class="reference internal" href="getting-started.html">Getting started</a></li>
<li class="toctree-l1"><a class="reference internal" href="guide.html">Guide to file hierarchy</a></li>
<li class="toctree-l1"><a class="reference internal" href="workflows.html">Overview of workflows</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="config.html">Configuration</a><ul class="current">
<li class="toctree-l2 current"><a class="reference internal" href="config.html#general-configuration">General configuration</a><ul class="current">
<li class="toctree-l3"><a class="reference internal" href="config-yaml.html">Config YAML</a></li>
<li class="toctree-l3"><a class="reference internal" href="references-config.html">References config</a></li>
<li class="toctree-l3 current"><a class="current reference internal" href="#">Sample tables</a></li>
<li class="toctree-l3"><a class="reference internal" href="patterns-targets.html">Patterns and targets</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="config.html#running-on-a-cluster">Running on a cluster</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="references.html">References workflow</a></li>
<li class="toctree-l1"><a class="reference internal" href="rnaseq.html">RNA-seq workflow</a></li>
<li class="toctree-l1"><a class="reference internal" href="downstream-rnaseq.html">RNA-Seq downstream analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="chipseq.html">ChIP-seq workflow</a></li>
<li class="toctree-l1"><a class="reference internal" href="integrative.html">Integrative workflows</a></li>
<li class="toctree-l1"><a class="reference internal" href="conda.html">conda and conda envs in <cite>lcdb-wf</cite></a></li>
<li class="toctree-l1"><a class="reference internal" href="tests.html">Testing the installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="faqs.html">FAQs</a></li>
<li class="toctree-l1"><a class="reference internal" href="changelog.html">Changelog</a></li>
<li class="toctree-l1"><a class="reference internal" href="developers.html">For Developers</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="toc.html">Documentation overview</a><ul>
  <li><a href="config.html">Configuration</a><ul>
      <li>Previous: <a href="references-config.html" title="previous chapter">References config</a></li>
      <li>Next: <a href="patterns-targets.html" title="next chapter">Patterns and targets</a></li>
  </ul></li>
  </ul></li>
</ul>
</div>
<search id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" autocomplete="off" autocorrect="off" autocapitalize="off" spellcheck="false"/>
      <input type="submit" value="Go" />
    </form>
    </div>
</search>
<script>document.getElementById('searchbox').style.display = "block"</script>
        </div>
      </div>
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <section id="sample-tables">
<span id="sampletable"></span><h1>Sample tables<a class="headerlink" href="#sample-tables" title="Link to this heading">¶</a></h1>
<p>Sample tables map sample names to files on disk and provide additional
metadata. It is expected to have a header and be tab-delimited. Empty lines and
lines that start with a comment (<code class="docutils literal notranslate"><span class="pre">#</span></code>) are skipped.</p>
<p>For running new experiments, you will need to write your own sample table. For
running experiments uploaded to SRA (Sequence Read Archive), <strong>you can use the SRA sample table
as-is</strong>, with the addition of a new column indicating what you would like to
name each sample. This makes it almost trivial to run arbitrary SRA RNA-seq
data sets! For ChIP-seq data from SRA, the additional columns <cite>antibody</cite>,
<cite>label</cite>, and <cite>biological_material</cite> as described below will need to be added,
but often that information is already in the SRA sampletable so the columns
just need to be renamed.</p>
<section id="rna-seq-sample-table">
<span id="rnaseq-sampletable"></span><h2>RNA-seq sample table<a class="headerlink" href="#rna-seq-sample-table" title="Link to this heading">¶</a></h2>
<p>Here is an example minimal sample table for RNA-seq. It only contains sample
IDs for four samples:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># Minimal RNA-seq sample table</span>
<span class="n">sample</span>
<span class="n">c1</span>
<span class="n">c2</span>
<span class="n">t1</span>
<span class="n">t2</span>
</pre></div>
</div>
<p>In this minimal example, the original FASTQ files are expected to be at the
locations <code class="docutils literal notranslate"><span class="pre">data/rnaseq_samples/{sample}/{sample}_R1.fastq.gz</span></code>. That pattern
is configured in the <code class="docutils literal notranslate"><span class="pre">config/rnaseq_patterns.yaml</span></code> file if you would like to
change it (see <a class="reference internal" href="patterns-targets.html#patterns-and-targets"><span class="std std-ref">Patterns and targets</span></a>). Specifically, the workflow will
expect the following files to already exist (paths relative to the Snakefile):</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># The above sample table expects these files to exist:</span>
<span class="n">data</span><span class="o">/</span><span class="n">rnaseq_samples</span><span class="o">/</span><span class="n">c1</span><span class="o">/</span><span class="n">c1_R1</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span>
<span class="n">data</span><span class="o">/</span><span class="n">rnaseq_samples</span><span class="o">/</span><span class="n">c2</span><span class="o">/</span><span class="n">c2_R1</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span>
<span class="n">data</span><span class="o">/</span><span class="n">rnaseq_samples</span><span class="o">/</span><span class="n">t1</span><span class="o">/</span><span class="n">t1_R1</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span>
<span class="n">data</span><span class="o">/</span><span class="n">rnaseq_samples</span><span class="o">/</span><span class="n">t2</span><span class="o">/</span><span class="n">t2_R1</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span>
</pre></div>
</div>
<section id="symlinking-fastqs">
<span id="symlinks"></span><h3>Symlinking FASTQs<a class="headerlink" href="#symlinking-fastqs" title="Link to this heading">¶</a></h3>
<p>To avoid having to copy or symlink files over into the expected directory
structure, we can instead list the original filenames in a column called
<code class="docutils literal notranslate"><span class="pre">orig_filename</span></code> and they will be automatically symlinked into
<code class="docutils literal notranslate"><span class="pre">data/rnaseq_samples/{sample}/{sample}_R1.fastq.gz</span></code>. That is, the following
sampletable:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># Example RNA-seq sample table with original filenames are specified</span>
<span class="n">sample</span>   <span class="n">orig_filename</span>
<span class="n">c1</span>       <span class="o">/</span><span class="n">data</span><span class="o">/</span><span class="n">c1</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span>
<span class="n">c2</span>       <span class="o">/</span><span class="n">data</span><span class="o">/</span><span class="n">c2</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span>
<span class="n">t1</span>       <span class="o">/</span><span class="n">data</span><span class="o">/</span><span class="n">other</span><span class="o">/</span><span class="n">t1</span><span class="o">.</span><span class="n">fq</span><span class="o">.</span><span class="n">gz</span>
<span class="n">t2</span>       <span class="o">../</span><span class="n">raw</span><span class="o">-</span><span class="n">data</span><span class="o">/</span><span class="n">t2</span><span class="o">.</span><span class="n">fq</span><span class="o">.</span><span class="n">gz</span>
</pre></div>
</div>
<p>Will result in the following symlinks:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">data</span><span class="o">/</span><span class="n">rnaseq_samples</span><span class="o">/</span><span class="n">c1</span><span class="o">/</span><span class="n">c1_R1</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span> <span class="o">-&gt;</span> <span class="o">/</span><span class="n">data</span><span class="o">/</span><span class="n">c1</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span>
<span class="n">data</span><span class="o">/</span><span class="n">rnaseq_samples</span><span class="o">/</span><span class="n">c2</span><span class="o">/</span><span class="n">c2_R1</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span> <span class="o">-&gt;</span> <span class="o">/</span><span class="n">data</span><span class="o">/</span><span class="n">c2</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span>
<span class="n">data</span><span class="o">/</span><span class="n">rnaseq_samples</span><span class="o">/</span><span class="n">t1</span><span class="o">/</span><span class="n">t1_R1</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span> <span class="o">-&gt;</span> <span class="o">/</span><span class="n">data</span><span class="o">/</span><span class="n">other</span><span class="o">/</span><span class="n">t1</span><span class="o">.</span><span class="n">fq</span><span class="o">.</span><span class="n">gz</span>
<span class="n">data</span><span class="o">/</span><span class="n">rnaseq_samples</span><span class="o">/</span><span class="n">t2</span><span class="o">/</span><span class="n">t2_R1</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span> <span class="o">-&gt;</span> <span class="o">../</span><span class="n">raw</span><span class="o">-</span><span class="n">data</span><span class="o">/</span><span class="n">t2</span><span class="o">.</span><span class="n">fq</span><span class="o">.</span><span class="n">gz</span>
</pre></div>
</div>
<p>Note that <cite>orig_filename</cite> paths in the sampletable are considered <em>relative to
the Snakefile</em>.</p>
</section>
<section id="additional-metadata">
<h3>Additional metadata<a class="headerlink" href="#additional-metadata" title="Link to this heading">¶</a></h3>
<p>For RNA-seq, only the first column and optionally the <cite>orig_filename</cite> column
are used directly by the RNA-seq workflow.</p>
<p>However, the sampletable is imported into the <code class="docutils literal notranslate"><span class="pre">downstream/rnaseq.Rmd</span></code> file
(see <a class="reference internal" href="downstream-rnaseq.html#downstream"><span class="std std-ref">RNA-Seq downstream analysis</span></a> for more info). It’s often useful to include
any metadata in the sampletable so it’s all in one place, and you’ll get all
that information imported into R.</p>
<p>For example, with this sample table we would be easily able to use a DESeq
model of <code class="docutils literal notranslate"><span class="pre">~condition</span></code> since the condition column will be imported into R.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># Example RNA-seq sampletable with &quot;condition&quot; metadata included</span>
<span class="n">sample</span>   <span class="n">orig_filename</span>          <span class="n">condition</span>
<span class="n">c1</span>       <span class="o">/</span><span class="n">data</span><span class="o">/</span><span class="n">c1</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span>      <span class="n">control</span>
<span class="n">c2</span>       <span class="o">/</span><span class="n">data</span><span class="o">/</span><span class="n">c2</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span>      <span class="n">control</span>
<span class="n">t1</span>       <span class="o">/</span><span class="n">data</span><span class="o">/</span><span class="n">other</span><span class="o">/</span><span class="n">t1</span><span class="o">.</span><span class="n">fq</span><span class="o">.</span><span class="n">gz</span>   <span class="n">treatment</span>
<span class="n">t2</span>       <span class="o">/</span><span class="n">data</span><span class="o">/</span><span class="n">other</span><span class="o">/</span><span class="n">t2</span><span class="o">.</span><span class="n">fq</span><span class="o">.</span><span class="n">gz</span>   <span class="n">treatment</span>
</pre></div>
</div>
</section>
<section id="paired-end-data">
<h3>Paired-end data<a class="headerlink" href="#paired-end-data" title="Link to this heading">¶</a></h3>
<p><strong>Paired-end and single-end data may be mixed in the same sampletable.</strong>
A sample is specified as paired-end using a separate column in the sampletable.
That column can either be named <cite>layout</cite> (easiest if you’re writing your own
sample table) or <cite>LibraryLayout</cite> (if you’re using an SRA sampletable, in which
case you can leave it as-is). An error will be raised if both columns are
provided.</p>
<p>If one of these columns exists, the values of the column are converted to
lowercase. For each sample, if the value is either <cite>pe</cite> or <cite>paired</cite>, the sample
will be considered paired-end. In all other cases the sample will be considered
single-end.</p>
<p>For paired-end samples that will be symlinked, both <cite>orig_filename</cite> and
<cite>orig_filename_R2</cite> must be specified as paths relative to the Snakefile (see
<a class="reference internal" href="#symlinks"><span class="std std-ref">Symlinking FASTQs</span></a> above). If there is a mix of SE and PE samples, the SE sample
must have an empty entry for <cite>orig_filename_R2</cite> (in the context of the
tab-delimited sampletable, this means two tab characters next to
each other with nothing in between).</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>If the sample table contains both single- and paired-end samples, the
<cite>fastq_dump</cite> and <cite>cutadapt</cite> rules will create empty R2 files.</p>
<p>Once the BAM files are created (after alignment in a single- or paired-end
fashion as appropriate for the sample), we operate mostly on the BAM.</p>
<p>After the alignment stage, remaining rules <strong>do not</strong> differentiate between
single- and paired-end reads. In particular, featureCounts and bamCoverage
may need different parameters depending on the library layout.</p>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># Example RNA-seq sample table with original filenames are specified,</span>
<span class="c1"># and c1 is a paired-end sample</span>
<span class="n">sample</span>   <span class="n">orig_filename</span>         <span class="n">orig_filename_R2</span>      <span class="n">layout</span>
<span class="n">c1</span>       <span class="o">/</span><span class="n">data</span><span class="o">/</span><span class="n">c1_R1</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span>  <span class="o">/</span><span class="n">data</span><span class="o">/</span><span class="n">c1_R2</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span>  <span class="n">PE</span>
<span class="n">c2</span>       <span class="o">/</span><span class="n">data</span><span class="o">/</span><span class="n">c2</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span>                           <span class="n">SE</span>
<span class="n">t1</span>       <span class="o">/</span><span class="n">data</span><span class="o">/</span><span class="n">other</span><span class="o">/</span><span class="n">t1</span><span class="o">.</span><span class="n">fq</span><span class="o">.</span><span class="n">gz</span>                        <span class="n">SE</span>
<span class="n">t2</span>       <span class="o">/</span><span class="n">data</span><span class="o">/</span><span class="n">other</span><span class="o">/</span><span class="n">t2</span><span class="o">.</span><span class="n">fq</span><span class="o">.</span><span class="n">gz</span>                        <span class="n">SE</span>
</pre></div>
</div>
</section>
</section>
<section id="chip-seq-sample-table">
<span id="chipseq-sampletable"></span><h2>ChIP-seq sample table<a class="headerlink" href="#chip-seq-sample-table" title="Link to this heading">¶</a></h2>
<p><strong>Three additional columns are required</strong> for ChIP-seq: <code class="docutils literal notranslate"><span class="pre">antibody</span></code>,
<code class="docutils literal notranslate"><span class="pre">biological_material</span></code> and <code class="docutils literal notranslate"><span class="pre">label</span></code>.</p>
<dl>
<dt><code class="docutils literal notranslate"><span class="pre">antibody</span></code></dt><dd><p>Used for differentiating between input and IP samples. Input samples should
be listed with an antibody of exactly <code class="docutils literal notranslate"><span class="pre">input</span></code>.</p>
</dd>
<dt><code class="docutils literal notranslate"><span class="pre">biological_material</span></code></dt><dd><p>Ties together which samples came from the same chromatin. This is how we
know a particular input sample is the matched control for a particular IP
sample. This is primarily used in the <cite>fingerprint</cite> rule, where we collect
all the input BAMs together for performing QC. See the
<cite>lib.chipseq.merged_input_for_ip</cite> function for the technical details of how
this is handled.</p>
</dd>
<dt><code class="docutils literal notranslate"><span class="pre">label</span></code></dt><dd><p>Used to tie together technical replicates, and <strong>used to configure the
ChIP-seq peak-calling runs</strong> (see <a class="reference internal" href="config-yaml.html#cfg-chipseq"><span class="std std-ref">chipseq config section</span></a>).</p>
<p>Technical replicates share the same label. If you don’t have technical
replicates, then this column can be a copy of the first column containing
sample names. Technical replicates will have their BAMs merged together
and duplicates removed from the merged BAM.</p>
</dd>
</dl>
<p>The reason that the ChIP-seq sample table is more complicated than RNA-seq is
because RNA-seq is often analyzed in R, and complicated sample handling (like
summing technical replicates) can be performed very flexibly in R. In contrast,
ChIP-seq peak-callers are command-line tools and frequently only take a single
biological replicate, and so are run as Snakemake rules. As a result, more
complex configuration is required to ensure complex experimental designs are
handled correctly.</p>
<section id="minimal-chip-seq-sample-table-no-replicates">
<h3>Minimal ChIP-seq sample table, no replicates<a class="headerlink" href="#minimal-chip-seq-sample-table-no-replicates" title="Link to this heading">¶</a></h3>
<p>A minimal ChIP-seq sample table, with no biological replicates, looks like
this:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># Example minimal ChIP-seq sample table</span>
<span class="n">sampleid</span>    <span class="n">antibody</span>   <span class="n">biological_material</span>  <span class="n">label</span>          <span class="n">orig_filename</span>
<span class="n">ip1</span>         <span class="n">gaf</span>        <span class="n">s2cell</span><span class="o">-</span><span class="mi">1</span>             <span class="n">s2cell</span><span class="o">-</span><span class="n">gaf</span><span class="o">-</span><span class="mi">1</span>   <span class="o">/</span><span class="n">data</span><span class="o">/</span><span class="n">ip1</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span>
<span class="n">input1</span>      <span class="nb">input</span>      <span class="n">s2cell</span><span class="o">-</span><span class="mi">1</span>             <span class="n">s2cell</span><span class="o">-</span><span class="nb">input</span><span class="o">-</span><span class="mi">1</span> <span class="o">/</span><span class="n">data</span><span class="o">/</span><span class="nb">input</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span>
</pre></div>
</div>
<ul class="simple">
<li><p>The input sample is required to have the antibody as “input”</p></li>
<li><p>For an IP, its matched input is the sample with <code class="docutils literal notranslate"><span class="pre">antibody</span> <span class="pre">==</span> <span class="pre">input</span></code> that
also has the same biological material as the IP. Here, we know <cite>input1</cite> goes
with <cite>ip1</cite> because they both have the same biological material.</p></li>
</ul>
</section>
<section id="chip-seq-sample-table-biological-replicates">
<h3>ChIP-seq sample table, biological replicates<a class="headerlink" href="#chip-seq-sample-table-biological-replicates" title="Link to this heading">¶</a></h3>
<p>Here is another example, this time with biological replicates:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># Example ChIP-seq sampletable with biological replicates</span>
<span class="n">sampleid</span>    <span class="n">antibody</span>   <span class="n">biological_material</span>  <span class="n">label</span>          <span class="n">orig_filename</span>
<span class="n">ip1</span>         <span class="n">gaf</span>        <span class="n">s2cell</span><span class="o">-</span><span class="mi">1</span>             <span class="n">s2cell</span><span class="o">-</span><span class="n">gaf</span><span class="o">-</span><span class="mi">1</span>   <span class="o">/</span><span class="n">data</span><span class="o">/</span><span class="n">ip1</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span>
<span class="n">ip2</span>         <span class="n">gaf</span>        <span class="n">s2cell</span><span class="o">-</span><span class="mi">2</span>             <span class="n">s2cell</span><span class="o">-</span><span class="n">gaf</span><span class="o">-</span><span class="mi">2</span>   <span class="o">/</span><span class="n">data</span><span class="o">/</span><span class="n">run2</span><span class="o">/</span><span class="n">ip3</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span>
<span class="n">input1</span>      <span class="nb">input</span>      <span class="n">s2cell</span><span class="o">-</span><span class="mi">1</span>             <span class="n">s2cell</span><span class="o">-</span><span class="nb">input</span><span class="o">-</span><span class="mi">1</span> <span class="o">/</span><span class="n">data</span><span class="o">/</span><span class="nb">input</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span>
<span class="n">input2</span>      <span class="nb">input</span>      <span class="n">s2cell</span><span class="o">-</span><span class="mi">2</span>             <span class="n">s2cell</span><span class="o">-</span><span class="nb">input</span><span class="o">-</span><span class="mi">2</span> <span class="o">/</span><span class="n">data</span><span class="o">/</span><span class="n">run2</span><span class="o">/</span><span class="n">input2</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span>
</pre></div>
</div>
<ul class="simple">
<li><p>As before, <cite>ip1</cite> and <cite>input1</cite> share the same biological material, indicating
that <cite>input1</cite> is the matched input for <cite>ip1</cite>.</p></li>
<li><p>The matched input for <cite>ip2</cite> is <cite>input2</cite> because they share the same
biological material (<cite>s2cell-2</cite>) and <cite>input2</cite> has <code class="docutils literal notranslate"><span class="pre">antibody</span> <span class="pre">==</span> <span class="pre">input</span></code>.</p></li>
<li><p>Each sample has a unique label because there are no technical replicates
here.</p></li>
</ul>
</section>
<section id="chip-seq-sample-table-biological-and-technical-replicates">
<h3>ChIP-seq sample table, biological and technical replicates<a class="headerlink" href="#chip-seq-sample-table-biological-and-technical-replicates" title="Link to this heading">¶</a></h3>
<p>Another example, this time with biological and technical replicates:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># Example ChIP-seq sampletable with bio and tech reps</span>
<span class="n">sampleid</span>    <span class="n">antibody</span>   <span class="n">biological_material</span>  <span class="n">label</span>          <span class="n">orig_filename</span>
<span class="n">ip1</span>         <span class="n">gaf</span>        <span class="n">s2cell</span><span class="o">-</span><span class="mi">1</span>             <span class="n">s2cell</span><span class="o">-</span><span class="n">gaf</span><span class="o">-</span><span class="mi">1</span>   <span class="o">/</span><span class="n">data</span><span class="o">/</span><span class="n">ip1</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span>
<span class="n">ip1a</span>        <span class="n">gaf</span>        <span class="n">s2cell</span><span class="o">-</span><span class="mi">1</span>             <span class="n">s2cell</span><span class="o">-</span><span class="n">gaf</span><span class="o">-</span><span class="mi">1</span>   <span class="o">/</span><span class="n">data</span><span class="o">/</span><span class="n">ip2</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span>
<span class="n">ip2</span>         <span class="n">gaf</span>        <span class="n">s2cell</span><span class="o">-</span><span class="mi">2</span>             <span class="n">s2cell</span><span class="o">-</span><span class="n">gaf</span><span class="o">-</span><span class="mi">2</span>   <span class="o">/</span><span class="n">data</span><span class="o">/</span><span class="n">run2</span><span class="o">/</span><span class="n">ip3</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span>
<span class="n">input1</span>      <span class="nb">input</span>      <span class="n">s2cell</span><span class="o">-</span><span class="mi">1</span>             <span class="n">s2cell</span><span class="o">-</span><span class="nb">input</span><span class="o">-</span><span class="mi">1</span> <span class="o">/</span><span class="n">data</span><span class="o">/</span><span class="nb">input</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span>
<span class="n">input2</span>      <span class="nb">input</span>      <span class="n">s2cell</span><span class="o">-</span><span class="mi">2</span>             <span class="n">s2cell</span><span class="o">-</span><span class="nb">input</span><span class="o">-</span><span class="mi">2</span> <span class="o">/</span><span class="n">data</span><span class="o">/</span><span class="n">run2</span><span class="o">/</span><span class="n">input2</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span>
</pre></div>
</div>
<ul class="simple">
<li><p><cite>ip1</cite> and <cite>ip1a</cite> are technical replicates because they share the label
<cite>s2cell-gaf-1</cite>. This is often the case when we need to sequence the same
sample again for higher depth.</p></li>
<li><p><cite>ip1</cite> and <cite>ip1a</cite> will be merged into one BAM file named after their common
label, <cite>s2cell-gaf-1</cite> (described further below). The remaining <cite>ip2</cite>,
<cite>input1</cite>, and <cite>input2</cite> do not have to be merged with anything, so they will
be symlinked.</p></li>
</ul>
</section>
<section id="merging-technical-replicates-for-chip-seq">
<h3>Merging technical replicates for ChIP-seq<a class="headerlink" href="#merging-technical-replicates-for-chip-seq" title="Link to this heading">¶</a></h3>
<p>In contrast to technical replicates in RNA-seq, where counts can be summed in
R, ChIP-seq is a bit more complicated. The ChIP-seq workflow uses <code class="docutils literal notranslate"><span class="pre">samtools</span>
<span class="pre">merge</span></code> to merge together the unique, duplicates-removed BAM files from
technical replicates into a single BAM, and then removes the duplicates again
from that merged file.</p>
<p>There is a “merged_techreps” key in <code class="docutils literal notranslate"><span class="pre">config/chipseq_patterns.yaml</span></code> which
defines the filenames to which technical replicates will be merged. By default
this pattern is
<code class="docutils literal notranslate"><span class="pre">data/chipseq_merged/{label}/{label}.cutadapt.unique.nodups.merged.bam</span></code>.
After trimming, aligning, removing multimappers, and removing duplicates, tech
reps are merged together. Specifically, these files:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">data</span><span class="o">/</span><span class="n">chipseq_samples</span><span class="o">/</span><span class="n">ip1</span><span class="o">/</span><span class="n">ip1</span><span class="o">.</span><span class="n">cutadapt</span><span class="o">.</span><span class="n">unique</span><span class="o">.</span><span class="n">nodups</span><span class="o">.</span><span class="n">bam</span>
<span class="n">data</span><span class="o">/</span><span class="n">chipseq_samples</span><span class="o">/</span><span class="n">ip1a</span><span class="o">/</span><span class="n">ip1a</span><span class="o">.</span><span class="n">cutadapt</span><span class="o">.</span><span class="n">unique</span><span class="o">.</span><span class="n">nodups</span><span class="o">.</span><span class="n">bam</span>
</pre></div>
</div>
<p>get merged and then duplicates removed again from that merged file, resulting
in this file:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">data</span><span class="o">/</span><span class="n">chipseq_merged</span><span class="o">/</span><span class="n">s2cell</span><span class="o">-</span><span class="n">gaf</span><span class="o">-</span><span class="mi">1</span><span class="o">/</span><span class="n">s2cell</span><span class="o">-</span><span class="n">gaf</span><span class="o">-</span><span class="mf">1.</span><span class="n">cutadapt</span><span class="o">.</span><span class="n">unique</span><span class="o">.</span><span class="n">nodups</span><span class="o">.</span><span class="n">merged</span><span class="o">.</span><span class="n">bam</span>
</pre></div>
</div>
<p>For samples with no technical replicates, only symlinks are performed, so for
example this file:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">data</span><span class="o">/</span><span class="n">chipseq_samples</span><span class="o">/</span><span class="n">ip2</span><span class="o">/</span><span class="n">ip2</span><span class="o">.</span><span class="n">cutadapt</span><span class="o">.</span><span class="n">unique</span><span class="o">.</span><span class="n">nodups</span><span class="o">.</span><span class="n">bam</span>
</pre></div>
</div>
<p>will get symlinked to this file:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">data</span><span class="o">/</span><span class="n">chipseq_merged</span><span class="o">/</span><span class="n">s2cell</span><span class="o">-</span><span class="n">gaf</span><span class="o">-</span><span class="mi">2</span><span class="o">/</span><span class="n">s2cell</span><span class="o">-</span><span class="n">gaf</span><span class="o">-</span><span class="mf">2.</span><span class="n">cutadapt</span><span class="o">.</span><span class="n">unique</span><span class="o">.</span><span class="n">nodups</span><span class="o">.</span><span class="n">merged</span><span class="o">.</span><span class="n">bam</span>
</pre></div>
</div>
<p>For peak-calling (see <a class="reference internal" href="config-yaml.html#cfg-chipseq"><span class="std std-ref">chipseq config section</span></a>) and any other downstream analysis,
<strong>the files to use are these merged (or symlinked) BAM files.</strong></p>
</section>
</section>
</section>


          </div>
          
        </div>
      </div>
    <div class="clearer"></div>
  </div>
    <div class="footer">
      &#169;2017, Ryan Dale, Justin Fear.
      
      |
      Powered by <a href="https://www.sphinx-doc.org/">Sphinx 8.2.3</a>
      &amp; <a href="https://alabaster.readthedocs.io">Alabaster 1.0.0</a>
      
      |
      <a href="_sources/sampletable.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>