
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Overview of workflows &#8212; lcdb-wf 0.1 documentation</title>
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <link rel="stylesheet" href="_static/alabaster.css" type="text/css" />
    <script id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
    <script src="_static/jquery.js"></script>
    <script src="_static/underscore.js"></script>
    <script src="_static/doctools.js"></script>
    <script src="_static/language_data.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="References workflow" href="references.html" />
    <link rel="prev" title="Using conda and conda envs" href="conda.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  <div class="document">
    
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="toc.html">lcdb-wf</a></h1>



<p class="blurb">Customizable workflows for high-throughput sequencing analysis</p>






<h3>Navigation</h3>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="index.html">Introduction</a></li>
<li class="toctree-l1"><a class="reference internal" href="getting-started.html">Getting started</a></li>
<li class="toctree-l1"><a class="reference internal" href="guide.html">Guide to file hierarchy</a></li>
<li class="toctree-l1"><a class="reference internal" href="tests.html">Testing the installation</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Overview of workflows</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#primary-analysis-workflows">Primary analysis workflows</a></li>
<li class="toctree-l2"><a class="reference internal" href="#integrative-analysis-workflows">Integrative analysis workflows</a></li>
<li class="toctree-l2"><a class="reference internal" href="#features-common-to-workflows">Features common to workflows</a></li>
<li class="toctree-l2"><a class="reference internal" href="#next-steps">Next Steps</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="config.html">Configuration details</a></li>
<li class="toctree-l1"><a class="reference internal" href="references.html">References workflow</a></li>
<li class="toctree-l1"><a class="reference internal" href="rnaseq.html">RNA-seq workflow</a></li>
<li class="toctree-l1"><a class="reference internal" href="downstream-rnaseq.html">RNA-Seq downstream analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="chipseq.html">ChIP-seq workflow</a></li>
<li class="toctree-l1"><a class="reference internal" href="integrative.html">Integrative workflows</a></li>
<li class="toctree-l1"><a class="reference internal" href="faqs.html">FAQs</a></li>
<li class="toctree-l1"><a class="reference internal" href="changelog.html">Changelog</a></li>
<li class="toctree-l1"><a class="reference internal" href="developers.html">For Developers</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="toc.html">Documentation overview</a><ul>
      <li>Previous: <a href="conda.html" title="previous chapter">Using conda and conda envs</a></li>
      <li>Next: <a href="references.html" title="next chapter">References workflow</a></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" />
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <div class="section" id="overview-of-workflows">
<span id="workflows"></span><h1>Overview of workflows<a class="headerlink" href="#overview-of-workflows" title="Permalink to this headline">¶</a></h1>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>These workflows <strong>are intended to be edited and customized by the user</strong>.</p>
<p>See <a class="reference internal" href="getting-started.html#getting-started"><span class="std std-ref">Getting started</span></a> for recommendations on setting up these workflows in
your project directory.</p>
</div>
<p>Each workflow lives in its own directory:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>├── references/
│   ├── Snakefile
│   └── ...
├── rnaseq/
│   ├── Snakefile
│   └── ...
├── chipseq/
│   ├── Snakefile
│   └── ...
├── colocalization/
│   ├── Snakefile
│   └── ...
├── external/
│   ├── Snakefile
│   └── ...
└── figures/
    ├── Snakefile
    └── ...
</pre></div>
</div>
<p>There are two general classes of workflows, <strong>primary analysis</strong> and the
<strong>integrative analysis</strong>.</p>
<p>Each workflow is driven by a <code class="docutils literal notranslate"><span class="pre">Snakefile</span></code> and is configured by plain text
<a class="reference external" href="https://en.wikipedia.org/wiki/YAML">YAML</a> and <a class="reference external" href="https://en.wikipedia.org/wiki/Tab-separated_values">TSV</a> format files (see
<a class="reference internal" href="config.html#config"><span class="std std-ref">Configuration details</span></a> for much more on this).</p>
<div class="section" id="primary-analysis-workflows">
<h2>Primary analysis workflows<a class="headerlink" href="#primary-analysis-workflows" title="Permalink to this headline">¶</a></h2>
<p>The primary analysis workflows are generally used for transforming raw data
(fastq files) into usable results. For RNA-seq, that’s differentially-expressed
genes (along with comprehensive QC and analysis). For ChIP-seq, that’s called
peaks or differentially bound chromatin regions.</p>
<p>The primary analysis workflows are:</p>
<div class="toctree-wrapper compound">
<ul>
<li class="toctree-l1"><a class="reference internal" href="references.html">References workflow</a></li>
<li class="toctree-l1"><a class="reference internal" href="rnaseq.html">RNA-seq workflow</a></li>
<li class="toctree-l1"><a class="reference internal" href="chipseq.html">ChIP-seq workflow</a></li>
</ul>
</div>
<p>While the references workflow can be stand-alone, usually it is run as
a by-product of running the RNA-seq or ChIP-seq workflows. Here we will
focus on RNA-seq and ChIP-seq which share some common properties.</p>
<p>Where possible, we prefer to have rules use the normal command-line syntax for
tools (examples include rules calling samtools, deepTools bamCoverage, picard,
salmon).  However in some cases we use wrapper scripts. Situtations where we
use wrappers:</p>
<ul class="simple">
<li><p>Ensuring various aligners (HISAT2, Bowtie2, STAR, bwa) behave uniformly.
These wrappers call the aligner, followed by samtools sort and view. The end
result is that FASTQs go in, and a sorted BAM comes out.</p></li>
<li><p>Tools with legacy dependencies like Python 2.7 that must be run in an
independent environment (macs2, sicer, rseqc)</p></li>
<li><p>R analyses (particularly spp and dupradar, which build up an R script
incrementally before calling it).</p></li>
<li><p>Tools that need complicated setup, or handling output files hard-coded by the
tool (fastqc, fastq_screen).</p></li>
</ul>
<p>In all cases, search for the string <strong>NOTE:</strong> in the Snakefile to read notes on
how to configure each rule, and make adjustments as necessary. You may see some
comments that say <cite># [TEST SETTINGS]</cite>; you can ignore these, and see
<a class="reference internal" href="tests.html#test-settings"><span class="std std-ref">A note about test settings</span></a> for more info.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>If you have two different RNA-seq experiments, from different species, they
have to be run separately. However, if downstream analyses will use them both
then you would like to keep them in the same project. In this case, you can copy
the <code class="docutils literal notranslate"><span class="pre">workflows/rnaseq</span></code> directory to two other directories:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>cp -r workflows/rnaseq workflows/genome1-rnaseq
cp -r workflows/rnaseq workflows/genome2-rnaseq
</pre></div>
</div>
<p>Now, downstream analyses can link to and utilize results from these individual
folders, while the whole project can remain self-contained.</p>
</div>
</div>
<div class="section" id="integrative-analysis-workflows">
<h2>Integrative analysis workflows<a class="headerlink" href="#integrative-analysis-workflows" title="Permalink to this headline">¶</a></h2>
<p>The integrative analysis workflows take input from the primary workflows and
tie them together.</p>
<p>The integrative analysis workflows are described in <a class="reference internal" href="integrative.html#integrative"><span class="std std-ref">Integrative workflows</span></a>:</p>
<div class="toctree-wrapper compound">
<ul>
<li class="toctree-l1"><a class="reference internal" href="integrative.html">Integrative workflows</a><ul>
<li class="toctree-l2"><a class="reference internal" href="integrative.html#colocalization-workflow">Colocalization workflow</a></li>
<li class="toctree-l2"><a class="reference internal" href="integrative.html#external-workflow">“External” workflow</a></li>
<li class="toctree-l2"><a class="reference internal" href="integrative.html#figures-workflow">“Figures” workflow</a></li>
</ul>
</li>
</ul>
</div>
</div>
<div class="section" id="features-common-to-workflows">
<h2>Features common to workflows<a class="headerlink" href="#features-common-to-workflows" title="Permalink to this headline">¶</a></h2>
<p>In this section, we will take a higher-level look at the features common to
the primary analysis workflows.</p>
<ul class="simple">
<li><p>There is some shared code across the multiple Snakefiles. For instance,
The directory <code class="docutils literal notranslate"><span class="pre">../..</span></code> is added to Python’s path. This way, the <code class="docutils literal notranslate"><span class="pre">../../lib</span></code>
module can be found, and we can use the various helper functions there. This is
also simpler than providing a <cite>setup.py</cite> to install the helper functions.</p></li>
<li><p>The config file is hard-coded to be <cite>config/config.yaml</cite>. This allows the config file to be
in the <cite>config</cite> dir with other config files without having to be specified on
the command line, while also affording the user flexibility. For instance, a custom
config can be specified at the command-line, using  <code class="docutils literal notranslate"><span class="pre">snakemake</span>
<span class="pre">--configfile</span> <span class="pre">&lt;path</span> <span class="pre">to</span> <span class="pre">other</span> <span class="pre">config</span> <span class="pre">file&gt;</span></code>.</p></li>
<li><p>The config file is loaded using <code class="docutils literal notranslate"><span class="pre">common.load_config</span></code>. This function resolves
various paths (especially the references config section) and checks to see
if the config is well-formatted.</p></li>
<li><p>To make it easier to work with the config, a <cite>SeqConfig</cite> object is created. It
needs that parsed config file as well as the patterns file (see
<a class="reference internal" href="patterns-targets.html#patterns-and-targets"><span class="std std-ref">Patterns and targets</span></a> for more on this). The act of creating this object
reads the sample table, fills in the patterns with sample names, creates
a reference dictionary (see <code class="docutils literal notranslate"><span class="pre">common.references_dict</span></code>) for easy access to
reference files, and for ChIP-seq, also fills in the filenames for the
configured peak-calling runs. This object, called <code class="docutils literal notranslate"><span class="pre">c</span></code> for convenience, can be
accessed to get all sort of information – <code class="docutils literal notranslate"><span class="pre">c.sampletable</span></code>, <code class="docutils literal notranslate"><span class="pre">c.config</span></code>,
<code class="docutils literal notranslate"><span class="pre">c.patterns</span></code>, <code class="docutils literal notranslate"><span class="pre">c.targets</span></code>, and <code class="docutils literal notranslate"><span class="pre">c.refdict</span></code> are frequently used in rules
throughout the Snakefiles.</p></li>
<li><p>Various files can be used to specify cluster-specific parameters if the workflows
are being run in a high-performance cluster environment. For example, a config file
<code class="docutils literal notranslate"><span class="pre">config/clusterconfig.yaml</span></code> can be used to specify global and rule-specific
memory and disk-space requirements for the Snakefile to use at run-time. For more
details, see <a class="reference internal" href="config.html#cluster"><span class="std std-ref">Running on a cluster</span></a>.</p></li>
</ul>
</div>
<div class="section" id="next-steps">
<h2>Next Steps<a class="headerlink" href="#next-steps" title="Permalink to this headline">¶</a></h2>
<p>Next we look at <a class="reference internal" href="config.html#config"><span class="std std-ref">Configuration details</span></a> for details on how to configure specific workflows,
before going into the implemented workflows:</p>
<ul class="simple">
<li><dl class="simple">
<dt>Primary analysis workflows</dt><dd><ul>
<li><p><a class="reference internal" href="references.html#references"><span class="std std-ref">References workflow</span></a></p></li>
<li><p><a class="reference internal" href="rnaseq.html#rnaseq"><span class="std std-ref">RNA-seq workflow</span></a></p></li>
<li><p><a class="reference internal" href="chipseq.html#chipseq"><span class="std std-ref">ChIP-seq workflow</span></a></p></li>
</ul>
</dd>
</dl>
</li>
<li><p><a class="reference internal" href="integrative.html#integrative"><span class="std std-ref">Integrative workflows</span></a></p></li>
</ul>
</div>
</div>


          </div>
          
        </div>
      </div>
    <div class="clearer"></div>
  </div>
    <div class="footer">
      &copy;2017, Ryan Dale, Justin Fear.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 3.3.1</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.12</a>
      
      |
      <a href="_sources/workflows.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>