---
title: Differential expression analysis
output:
    html_document:
        code_folding: hide
        toc: true
        toc_float: true
        toc_depth: 3
---

```{r global_options, include=FALSE}
# Sets up global options for rendering RMarkdown into HTML.
knitr::opts_chunk$set(
    warning=FALSE,
    message=FALSE,
    cache.extra_file_dep_1 = file.info('../config/sampletable.tsv')$mtime,
    cache.extra_file_dep_2 = file.info('../data/rnaseq_aggregation/featurecounts.txt')$mtime,
    cache.extra_file_dep_3 = file.info('../data/rnaseq_samples/*/*.kallisto/abundance.h5')$mtime,
    cache.extra_file_dep_4 = file.info('../data/rnaseq_samples/*/*.salmon/quant.sf')$mtime
)
```

```{r lcdbwf, results='hide'}
# Load the lcdbwf R package, which is stored locally.
# This package has many custom functions used throughout this document.
devtools::document('../../../lib/lcdbwf')
devtools::load_all('../../../lib/lcdbwf')
```

```{r libraries}
library(AnnotationHub)
library(BiocParallel)
library(clusterProfiler)
library(cowplot)
library(DESeq2)
library(dplyr)
library(DT)
library(genefilter)
library(ggplot2)
library(gridExtra)
library(plotly)
library(purrr)
library(readr)
library(reshape)
library(tibble)
library(tximport)
library(UpSetR)
```

```{r config}

# HOW TO CONFIGURE ------------------------------------------------------
# Any chunks below that depend on config options should be cached and use one
# or more sections of the config object as arbitrary additional chunk options.
# By convention, we use the argument name "config" although there is nothing
# special about this name e.g.,
#
#      {r, cache=TRUE, config=config$annotation}
#
# Thereafter, any changes to config options under the "annotation" section will then
# invalidate that chunk's cache -- along with any other chunks that also have
# config$annotation included as a chunk option.
#
config <- lcdbwf:::load_config('config.yaml')

# To keep this Rmd streamlined, much of the text is stored in a different file,
# text.yaml, and accessed via the `text` list after it is loaded here.
text <- yaml::yaml.load_file('text.yaml')

parallel <- config$parallel$parallel
if (config$parallel$parallel){
    register(MulticoreParam(config$parallel$cores))
}
```


# Changelog

**Initial results**

Last run: `r date()`

```{r coldata_setup}
# Set up all of the metadata for the samples and experimental design. Use this
# chunk to modify if needed.
colData <- read.table(config$main$sampletable, sep='\t', header=TRUE, stringsAsFactors=FALSE)

# lcdb-wf requires that the first column of the sampletable contains sample
# names. Use that to set the rownames.
rownames(colData) <- colData[,1]
```

```{r dds_initial, cache=TRUE, config=c(config$main, config$toggle$salmon, config$toggle$kallisto)}
# Convert featureCounts gene-level counts into DESeq2 object, and run
# variance-stabiliizing transform.
dds_initial <- lcdbwf:::make_dds(
  list(sampletable=colData, design=~1),
  config=config,
  parallel=config$parallel$parallel
)
vsd <- varianceStabilizingTransformation(dds_initial, blind=TRUE)
```

# Experiment overview

Here is the sample table with metadata used for this analysis:

```{r print_coldata}
exclude.for.printing <- c('featurecounts.path', 'salmon.path', 'kallisto.path',
                          'orig_filename', 'orig_filename_R2', 'layout',
                          'sizeFactor')
colData(dds_initial) %>%
    as.data.frame() %>%
    dplyr::select(-contains(exclude.for.printing)) %>%
    datatable()
```


# Sample similarity and QC

## Clustered heatmap

```{r sample_heatmap, results='asis', cache=TRUE, config=c(text$qc$clustered_heatmap, config$plotting$covariates_for_plots), dependson='dds_initial'}
# Plot a clustered heatmap of sample distances, for QC.
lcdbwf:::mdcat(text$qc$clustered_heatmap)
lcdbwf:::plot_heatmap(vsd, colData, cols_for_grouping=config$plotting$covariates_for_plots)
```

## PCA {.tabset}

```{r pca, results='asis', cache=TRUE, config=c(text$qc$pca, config$plotting$covariates_for_plots), dependson='dds_initial'}
# Plot PCA plots, optionally tabbed with different factors colored, for QC.
lcdbwf:::mdcat(text$qc$pca)
for(group in config$plotting$covariates_for_plots){
    lcdbwf:::mdcat('### ', paste(group, sep=", "))
    p <- lcdbwf:::plotPCA.ly(vsd, intgroup=group)
    print(htmltools::tagList(ggplotly(p)))
}
```

```{r sizefactors, results='asis', eval=!(config$toggle$salmon | config$toggle$kallisto)}
# Note that when loading Salmon or Kallisto, DESeq2 does not calculate size
# factors.

lcdbwf:::mdcat(text$sizefactors)
lcdbwf:::mdcat("### Size factors")
p <- lcdbwf:::sizefactors_barplot(dds_initial)
ggplotly(p)

lcdbwf:::mdcat("### Size factors vs total read count")
p <- lcdbwf:::sizefactors_vs_total(dds_initial)
ggplotly(p)
```

```{r dds_list, cache=TRUE, config=config$main}
# Create a list of dds objects used as the basis for differential expression.
#
# May need substantial editing.
lst <- list(
            # ------------------------------------------------------------------
            # NOTE: Delete these for real data!
            # They are just examples to demonstrate functionality.
            #
            # Example 1: simplest case
            main=list(
                sampletable=colData,
                design=~group),

            # Example 2: demonstrates how to filter samples.
            # Note that subset.counts=TRUE is required when filtering.
            no_rep4=list(
                sampletable=colData %>% filter(samplename != 'sample4'),
                design=~group,
                subset_counts=TRUE
            ),

            # Example 3: use salmon
            salmon=list(
                sampletable=colData,
                design=~group,
                salmon=TRUE),

            # Example 4: use kallisto
            kallisto=list(
                sampletable=colData,
                design=~group,
                kallisto=TRUE)
            # ------------------------------------------------------------------
          )
dds_list <- map(lst, lcdbwf:::make_dds, config=config, parallel=config$parallel$parallel)
```


```{r dds_diagnostics, results='asis', cache=TRUE, dependson='dds_list', config=config$toggle$dds_diagnostics}
if (config$toggle$dds_diagnostics){ lcdbwf:::dds_diagnostics(dds_list, text) }
```

# Differential expression {.tabset}

Now we perform differential expression analysis for the contrasts listed in the
table below.

Under the table are tabs, one for each contrast. Under each contrast's tab,
there are more tabs for the various output.

```{r results_01, dependson='dds_list', cache=TRUE}
# Perform differential expression testing.
#
# ------------------------------------------------------------------------------
# To take advantage of caching:
#   - create each list in a different chunk
#   - ensure chunk name starts with "results_"
#   - ensure list name starts with "contr_[index]_". The rest of the name will be used
#     as a readable label for each constrast. [index] is an alphanumeric string 
#     (ex: contr_01_* or contr_2b_*) that will be used to sort contrasts for output files.
#     The index string must contain at least 1 character and cannot contain "_"
#
# This will need substantial editing.
#
# See ?make_results for details.
# ------------------------------------------------------------------------------

# Delete these for real data!
#
# Example 1:
#  - demonstrates creating and manipulating coefficients
control <- lcdbwf:::dds_coefs('main', group=='control')
treatment <- lcdbwf:::dds_coefs('main', group=='treatment')
contr_01_main <- lcdbwf:::make_results(
  dds_name='main',
  label='LFC>0',
  contrast=treatment - control,
  type='ashr'
)
```

```{r results_02, dependson='dds_list', cache=TRUE}
# Example 2:
#  - demonstrates using a log2FoldChange threshold
contr_02_lfc1 <- lcdbwf:::make_results(
  dds_name="no_rep4",
  contrast=c("group", "treatment", "control"),
  lfcThreshold=1,
  type="ashr",
  label="No rep 4 and using log2FoldChange of >1"
)
```

```{r results_03, dependson='dds_list', cache=TRUE}
# Example 3:
#   - demonstrates using normal shrinkage
#   - demonstrates how to use salmon
contr_03a_salmon <- lcdbwf:::make_results(
  dds_name="salmon",
  contrast=c("group", "treatment", "control"),
  type="normal",
  label="Using Salmon"
)
```

```{r results_04, dependson='dds_list', cache=TRUE}
# Example 4:
#   - like example 3, but kallisto instead of salmon
contr_03_kallisto <- lcdbwf:::make_results(
  dds_name="kallisto",
  contrast=c('group', 'treatment', 'control'),
  type='normal',
  label='Using Kallisto'
)
```


```{r assemble_variables, cache=TRUE, config=config$annotation, dependson=knitr::all_labels()[grepl("^results_", knitr::all_labels())]}
res_list <- lcdbwf:::collect_objects("^contr_[^_]+_")
res_list <- lcdbwf:::attach_extra(res_list, config)
```

```{r summary, results='asis'}
lcdbwf:::folded_markdown(text$results_table, "Help for table")
lcdbwf:::summarize_res_list(res_list) %>% datatable
```

```{r reportresults, results='asis', config=c(text$results_plots, config$toggle$results_diagnostics, config$plotting$diagnostics_results_names), dependson='assemble_variables', cache=TRUE}
# Create nested tabs showing the results.
# First level of tabs will be labels of results; second level will be different
# output (MA plot, volcano plot, etc)
lcdbwf:::build_results_tabs(res_list, dds_list, config, text)
```

```{r upsetplots, results='asis', eval=length(res_list)>1, config=config$annotation}
# Compare overlap across multiple differential expression contrasts.
lcdbwf:::mdcat(text$upset_plots)
lcdbwf:::plot_upsets(res_list, label_column=config$annotation$label_column)
```

# Exported results

```{r excel, results='asis'}
lcdbwf:::exported_excel(res_list, dds_list, file='results/consolidated_results.xlsx')
```

Here is a single Excel file with one worksheet for each contrast: [results/consolidated_results.xlsx](results/consolidated_results.xlsx)

Alternatively, the files below are TSVs that can be opened in Excel or used
progammatically with downstream tools:

```{r write_output, results='asis'}
# Write out files for full and each selection, and create a link to them in the
# HTML generated by this RMarkdown.
tbl <- lcdbwf:::exported_tsvs(res_list)
knitr::kable(tbl, row.names=FALSE)
```

```{r combined_rds, cache=TRUE, dependson='assemble_variables'}
obj <- lcdbwf:::compose_results(res_list, dds_list)
saveRDS(obj, file='combined.Rds', compress=FALSE)
```


# Session info
For reproducibility purposes, here is the output of `sessionInfo()` showing the
versions of all packages used here.

```{r sessioninfo}
sessionInfo()
```
