---
output:
    html_document:
        code_folding: hide
        toc: true
        toc_float: true
        toc_depth: 3
---

```{r, include=FALSE}
# Each named chunk has a corresponding entry at:
#
#   https://lcdb.github.io/lcdb-wf/rnaseq-rmarkdown-docs.html
```

```{r global_options, include=FALSE}
# Docs: https://lcdb.github.io/lcdb-wf/rnaseq-rmd.html#global_options
knitr::opts_chunk$set(
    warning=FALSE,
    message=FALSE,
    cache.extra_file_dep_1=file.info('../config/sampletable.tsv')$mtime,
    cache.extra_file_dep_2 = file.info('../data/rnaseq_aggregation/featurecounts.txt')$mtime
)
```

# Changelog

**Initial results**

Last run: `r date()`

# RNA-seq results

```{r}
library(AnnotationHub)
library(BiocParallel)
library(clusterProfiler)
library(cowplot)
library(DESeq2)
library(dplyr)
library(DT)
library(genefilter)
library(ggplot2)
library(gridExtra)
library(plotly)
library(ReactomePA)
library(readr)
library(reshape)
library(tibble)
library(tximport)
library(UpSetR)
```


```{r load_helpers}
# Docs: https://lcdb.github.io/lcdb-wf/rnaseq-rmd.html#load_helpers

devtools::document('../../../lib/lcdbwf')
devtools::load_all('../../../lib/lcdbwf')
```

```{r annotationhub_setup}
# Docs: https://lcdb.github.io/lcdb-wf/rnaseq-rmd.html#annotationhub_setup
annotation_genus_species <- 'Homo sapiens'
annotation_key_override <- NA
hub.cache <- '../../../include/AnnotationHubCache'
orgdb <- get.orgdb(
    annotation_genus_species,
    cache=hub.cache,
    annotation_key_override=annotation_key_override
)
```

```{r coldata_setup}
# Docs: https://lcdb.github.io/lcdb-wf/rnaseq-rmd.html#coldata_setup
sample.table.filename <- '../config/sampletable.tsv'
strip.dotted.version <- TRUE
exclude.for.printing <- c('featurecounts.path', 'salmon.path', 'orig_filename', 'orig_filename_R2')
factor.columns <- c('group')
salmon.path.func <- function (x) file.path('..', 'data', 'rnaseq_samples', x, paste0(x, '.salmon'), 'quant.sf')

colData <- read.table(sample.table.filename, sep='\t', header=TRUE, stringsAsFactors=FALSE)

colData$salmon.path <- sapply(colData$samplename, salmon.path.func)

for (col in factor.columns){
    colData[[col]] <- as.factor(colData[[col]])
}

colData$group <- relevel(colData$group, ref='DMSO_1h')
rownames(colData) <- colData[,1]

# remove JQ1 samples
colData <- colData %>%
    filter(treatment != 'JQ1')
```

## Experiment overview

Here is the sample table with metadata used for this analysis:

```{r}
knitr::kable(colData[, colnames(colData)[!colnames(colData) %in% exclude.for.printing]])
```

```{r salmon}
# Docs: https://lcdb.github.io/lcdb-wf/rnaseq-rmd.html#salmon
txi <- tximport(colData[, 'salmon.path'], type='salmon', txOut=TRUE)
transcript.tpm <- txi$abundance
colnames(transcript.tpm) <- colData$samplename
```

```{r ddstxi, cache=TRUE, eval=FALSE}
# Docs: https://lcdb.github.io/lcdb-wf/rnaseq-rmd.html#ddstxi

dds.txi <- DESeqDataSetFromTximport(
    txi, colData=colData[, -grepl('path', colnames(colData)), drop=FALSE],
    design=~group
)

vsd.txi <- varianceStabilizingTransformation(dds.txi, blind=TRUE)
```


```{r dds_initial, cache=TRUE}
# Docs: https://lcdb.github.io/lcdb-wf/rnaseq-rmd.html#dds_initial

dds <- lcdbwf::DESeqDataSetFromCombinedFeatureCounts(
    '../data/rnaseq_aggregation/featurecounts.txt',
    sampletable=colData,
    design=~group,
    subset.counts=TRUE)

# NOTE: collapse technical replicates ----------------------------------------
#   If collapsing technical replicates, do so here
# dds <- lcdbwf::collapseReplicates2(dds, dds$biorep)

if (strip.dotted.version){
    rownames(dds) <- sapply(strsplit(rownames(dds), '.', fixed=TRUE),
                            function (x) {ifelse(grepl('_', x[2]),
                                                 paste(x[1], x[2], sep='.'),
                                                 x[1])}
                            )
}

vsd <- varianceStabilizingTransformation(dds, blind=TRUE)

outliers <- c('HAP1_MTX_6h_2', 'outlier_HAP1_dBet6_1h_1')
dds.no.outlier <- lcdbwf::DESeqDataSetFromCombinedFeatureCounts(
    '../data/rnaseq_aggregation/featurecounts.txt',
    sampletable=colData %>% filter(!samplename %in% outliers),
    design=~group,
    subset.counts=TRUE)
dds.no.outlier <-collapseReplicates(dds.no.outlier, dds.no.outlier$biorep)
vsd.no.outlier <- varianceStabilizingTransformation(dds.no.outlier, blind=TRUE)

# remove batch effect of cell line
#vsd.batch <- vsd.no.outlier
#assay(vsd.batch) <- removebatchEffect(assay(vsd.batch),
#                                      batches=list(colData(vsd.batch)$cell_Line),
#                                      design=model.matrix(~group, data=colData(vsd.batch)))
```

## Sample clustering and QC

The following heatmap shows a hierarchical clustering of pairwise distances
between samples. Darker blue means less distant (i.e. more similar). In general
we expect to see replicates clustering together and separation of treatments.

## Clustered heatmap

```{r sample_heatmap, cache=TRUE, dependson='dds_initial'}
# Docs: https://lcdb.github.io/lcdb-wf/rnaseq-rmd.html#sample_heatmap
lcdbwf::plot.heatmap(vsd, colData, cols.for.grouping=c('group'))
cat('\n\n### Without outliers\n\n')
lcdbwf::plot.heatmap(vsd.no.outlier, colData, cols.for.grouping=c('group'))
cat('\n\n### After removing cell line batch effect\n\n')
#lcdbwf::plot.heatmap(vsd.batch, colData, cols.for.grouping=c('group'))
cat('\n\nDISABLED\n\n')
```

**Note: HAP1_MTX_6h_2 and outlier_HAP1_dBet6_1h_1 are outliers and will be removed from subsequent analyses.**

## PCA {.tabset}

Another way of looking at sample clustering is principal components analysis
(PCA). The x- and y-axes do not have units, rather, they represent the
dimensions along which the samples vary the most. The amount of variance
explained by each principal component is indicated in the axes label.


```{r pca, results='asis', cache=TRUE, dependson='dds_initial'}
# Docs: https://lcdb.github.io/lcdb-wf/rnaseq-rmd.html#pca

groups <- list(
               group=c('group'),
               cell_line_group=c('cell_line_group'),
               cell_Line=c('cell_Line'),
               treatment=c('treatment'),
               timepoint=c('timepoint')
               )

p <- list()
for(group in groups){
    p[[group]] <- lcdbwf::plotPCA.ly(vsd.no.outlier, intgroup=group)
}

i <- 1; lcdbwf::mdcat('### ', names(p)[i]); ggplotly(p[[i]])
i <- 2; lcdbwf::mdcat('### ', names(p)[i]); ggplotly(p[[i]])
i <- 3; lcdbwf::mdcat('### ', names(p)[i]); ggplotly(p[[i]])
i <- 4; lcdbwf::mdcat('### ', names(p)[i]); ggplotly(p[[i]])
i <- 5; lcdbwf::mdcat('### ', names(p)[i]); ggplotly(p[[i]])
```

## PCA after batch effect removal {.tabset}

DISABLED

```{r pca_batch, results='asis', cache=TRUE, dependson='dds_initial'}
#p <- list()
#for(group in groups){
#    p[[group]] <- lcdbwf::plotPCA.ly(vsd.batch, intgroup=group)
#}
#
#i <- 1; lcdbwf::mdcat('### ', names(p)[i]); ggplotly(p[[i]])
#i <- 2; lcdbwf::mdcat('### ', names(p)[i]); ggplotly(p[[i]])
#i <- 3; lcdbwf::mdcat('### ', names(p)[i]); ggplotly(p[[i]])
#i <- 4; lcdbwf::mdcat('### ', names(p)[i]); ggplotly(p[[i]])
#i <- 5; lcdbwf::mdcat('### ', names(p)[i]); ggplotly(p[[i]])
```


## Size factors {.tabset}
Ideally, all libraries were sequenced to identical depth, in which case all
size factors would be 1.0. In practice, this is almost never the case due to
the difficulties of accurately measuring low concentrations of cDNA. These size
factor estimates are DESeq2's way of showing how sequencing depth varies across
libraries. If some libraries are much higher or lower than
1 then those libraries had dramatically different coverage and we should be
careful about interpreting results.

Size factors are calculated according to the median-ratio method (equation 5 of
[Anders & Huber
2010](http://dx.doi.org/10.1186/gb-2010-11-10-r106)).

These diagnostic plots show the size factors (as a ranked bar plot) and the
relationship between the size factors and the total read count (as
a scatterplot).

```{r sizefactors, results='asis'}
# Docs: https://lcdb.github.io/lcdb-wf/rnaseq-rmd.html#sizefactors

cat('\n\n\n\n\n\n')
lcdbwf::mdcat('### Size factors')
dds <- estimateSizeFactors(dds)
sf <- sizeFactors(dds)
sf <- sf[order(sf)] %>%
        enframe(value = 'Size Factor')

p <- ggplot(sf) +
    aes(x=reorder(name, `Size Factor`), y=`Size Factor`) +
    xlab('sample name') +
    geom_col() +
    theme_bw() +
    theme(axis.text.x=element_text(angle=90, vjust=0.5, hjust=1))
ggplotly(p)
cat('\n\n\n\n\n\n')

lcdbwf::mdcat('### Size factors vs total read count')
trc <- colSums(counts(dds)) %>%
        enframe(value = 'Total Read Count')
trc_vs_sf <- full_join(sf, trc, by='name')
p <- ggplot(data=trc_vs_sf, aes_string(x="`Total Read Count`", y="`Size Factor`", label='name')) +
    geom_point(size=3) +
    theme_bw() +
    theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))
ggplotly(p)
cat('\n\n\n\n\n\n')
```

```{r parallel_config}
# Docs: https://lcdb.github.io/lcdb-wf/rnaseq-rmd.html#parallel_config

parallel <- TRUE
register(MulticoreParam(4))
```

```{r dds_list, cache=TRUE}
# Docs: https://lcdb.github.io/lcdb-wf/rnaseq-rmd.html#dds_list

lst <- list(
            main=list(
                sampletable=colData %>% filter(!samplename %in% outliers),
                design=~treatment,
                args=list(subset.counts=TRUE)
            ),
            #jq_cell_line=list(
            #    sampletable=colData %>% filter(treatment %in% c('DMSO', 'JQ1'), !samplename %in% outliers),
            #    design=~cell_line_group + 0,
            #    args=list(subset.counts=TRUE)
            #),
            dbet_cell_line=list(
                sampletable=colData %>% filter(treatment %in% c('DMSO', 'dBet6'), !samplename %in% outliers),
                design=~cell_line_group + 0,
                args=list(subset.counts=TRUE)
            )
)

dds.list <- map(lst, lcdbwf::make.dds, salmon.files=NULL, remove.version=TRUE,
                combine.by='biorep', parallel=parallel)
```


```{r results, dependson='dds_list', cache=TRUE}
# Docs: https://lcdb.github.io/lcdb-wf/rnaseq-rmd.html#results

res.list <- list()

# NOTE: Example contrast #1: JQ1 vs. DMSO ---------------------------------
#   Using the example data, this compares treatment group to control group.
#   Change to reflect your experiment.
# IMPORTANT: Control must always be last

# including only treatment in the contrast
#res.list[['main_jq']] <- list(
#    res=results(dds.list[['main']], contrast=c('treatment', 'JQ1', 'DMSO')),
#    dds='main',
#    label='JQ1 vs. DMSO (LFC > 0); dds main; cts treatment'
#)

# JQ1 vs. DMSO at 1h
#res.list[['jq1h-batch']] <- list(
#    res=results(dds.list[['jq_cell_line']], contrast=c(-1,0,1,0,-1,0,1,0,-1,0,1,0)),
#    dds='jq_cell_line',
#    label='JQ1 vs. DMSO 1h (LFC>0); dds jq_cell_line; cts cell_Line_group; batch effect'
#)

# JQ1 vs. DMSO at 6h
#res.list[['jq6h-batch']] <- list(
#    res=results(dds.list[['jq_cell_line']], contrast=c(0,-1,0,1,0,-1,0,1,0,-1,0,1)),
#    dds='jq_cell_line',
#    label='JQ1 vs. DMSO 6h (LFC>0); dds jq_cell_line; cts cell_Line_group; batch effect'
#)

# JQ1 vs. DMSO 6h vs. 1h
#res.list[['jq6h.vs.1h-batch']] <- list(
#    res=results(dds.list[['jq_cell_line']], contrast=c(1,-1,-1,1,1,-1,-1,1,1,-1,-1,1)),
#    dds='jq_cell_line',
#    label='JQ1 vs. DMSO 6h vs. 1h (LFC>0); dds jq_cell_line; cts cell_Line_group; batch effect'
#)

# dBet6 ---------------------------

res.list[['main_dbet']] <- list(
    res=results(dds.list[['main']], contrast=c('treatment', 'dBet6', 'DMSO')),
    dds='main',
    label='dBet6 vs. DMSO (LFC > 0); dds main; cts treatment'
)

# dBET6 vs. DMSO at 1h
res.list[['dbet1h-batch']] <- list(
    res=results(dds.list[['dbet_cell_line']], contrast=c(-1,0,1,0,-1,0,1,0,-1,0,1,0)),
    dds='dbet_cell_line',
    label='dBet6 vs. DMSO 1h (LFC>0); dds dbet_cell_line; cts cell_Line_group; batch effect'
)

# dBet6 vs. DMSO at 6h
res.list[['dbet6h-batch']] <- list(
    res=results(dds.list[['dbet_cell_line']], contrast=c(0,-1,0,1,0,-1,0,1,0,-1,0,1)),
    dds='dbet_cell_line',
    label='dBet6 vs. DMSO 6h (LFC>0); dds dbet_cell_line; cts cell_Line_group; batch effect'
)

# dBet6 vs. DMSO 6h vs. 1h
res.list[['dbet6h.vs.1h-batch']] <- list(
    res=results(dds.list[['dbet_cell_line']], contrast=c(1,-1,-1,1,1,-1,-1,1,1,-1,-1,1)),
    dds='dbet_cell_line',
    label='dBet6 vs. DMSO 6h vs. 1h (LFC>0); dds dbet_cell_line; cts cell_Line_group; batch effect'
)

```


```{r attach, cache=TRUE, dependson='results'}
# Docs: https://lcdb.github.io/lcdb-wf/rnaseq-rmd.html#attach
keytype <- 'ENSEMBL'
columns <- c('SYMBOL', 'UNIPROT', 'ALIAS')

for (name in names(res.list)){
    res.list[[name]][['res']] <- attach.info(
        res.list[[name]][['res']],
        keytype=keytype,
        columns=columns)
}
```

# Differential expression {.tabset}

```{r}
# NOTE: significance level---------------------------------------------------
alpha <- 0.1
lfc.thresh <- 0 # Must change if using an lfcThreshold other than 0
```

Here is a table summarizing the comparisons. See the [Background and
help](#Help) section for details.

```{r, results='asis'}
knitr::kable(summarize.res.list(res.list, alpha, lfc.thresh, dds.list))
```

For each comparison there is a tab and under each tab are the following:

- summary of results (the line from the summary table above for this comparison)
- Normalized counts plots for the top 3 upregulated genes
- Normalized counts plots for the top 3 down-regulated genes
- an M-A plot
- a p-value histogram


See the [Background and help](#Help) section for details on these.


```{r reportresults, results='asis'}
# Docs: https://lcdb.github.io/lcdb-wf/rnaseq-rmd.html#reportresults

# NOTE: Which columns to add to the top plots' titles?-------------------------
#    This will add nicer titles to the plots. These may have come from the
#    `attach.info` call above.
add_cols <- c('symbol', 'alias')

for (name in names(res.list)){
  dds.i <- dds.list[[ res.list[[name]][['dds']] ]]
  res.i <- res.list[[name]][['res']]
  label <- res.list[[name]][['label']]

  mdcat('## ', label, ' {.tabset}')
  mdcat('### Summary of results')
  print(knitr::kable(my.summary(res.i, dds.i)))
  cat(paste0('resultsNames: ',paste(resultsNames(dds.i), collapse=', '), '\n\n'))
  cat(paste(elementMetadata(res.i)$description, collapse='\n\n'))
  mdcat('### Normalized counts of top 3 upregulated genes')
  print(lcdbwf::counts.plot(lcdbwf::counts.df(dds.i, res.i, sel.genes=lfc.filter(res.i),
                              label='symbol', rank.col='padj'),
                    3, no.aes=TRUE) + aes(y=normalized_counts, x=group, color=cell_Line))
  mdcat('### Normalized counts of top 3 downregulated genes')
  print(lcdbwf::counts.plot(lcdbwf::counts.df(dds.i, res.i, sel.genes=lfc.filter(res.i, reverse=TRUE),
                              label='symbol', rank.col='padj'),
                    3, no.aes=TRUE) + aes(y=normalized_counts, x=group, color=cell_Line))

  # NOTE: gene labels --------------------------------------------------------
  #   By default we plot the symbols of the top 10 upregulated or downregulated
  #   genes, with the lowest padj. To change this behavior edit the following
  #   lines
  res.i <- res.i[order(res.i$padj),]

  genes.to.label <- res.i[1:10, 'symbol']

  mdcat('### M-A plot')
  print(plotMA.label(res.i, genes.to.label=genes.to.label, col='symbol'))

  mdcat('### P-value distribution')
  pval.hist(res.i)
}
```

```{r selections, cache=TRUE, dependson='attach'}
# Docs: https://lcdb.github.io/lcdb-wf/rnaseq-rmd.html#selections
sel.list <- list()
for (name in names(res.list)){
  res <- res.list[[name]][['res']]


  # NOTE: any other selections?------------------------------------------------
  #    Here we just get the up- and downregulated genes, but any arbitrary
  #    subsets of the results can be added.
  #
  #    For each selection:
  #      - TSV of the subset of genes will be written to file and a link
  #        created for it in the Markdown
  #      - GO analysis will be performed on each group separately below.
  sel.list[[name]] <- list(
    up=res[get.sig(res, alpha=alpha, lfc.thresh=lfc.thresh, direction='up'),],
    dn=res[get.sig(res, alpha=alpha, lfc.thresh=lfc.thresh, direction='dn'),]
    )
}

# Keep track of our sel.list entries, here just using the last one
sel.names <- names(sel.list[[name]])
```

```{r upsetplots, results='asis', cache=TRUE, dependson='selections'}
# Docs: https://lcdb.github.io/lcdb-wf/rnaseq-rmd.html#upsetplots
# UpSet plots only make sense for more than one set of genes. The Markdown
# explanatory text and the plots themselves are only created if res.list has
# multiple items in it.

plot.upset <- function(sel.list.key, label) {
    mdcat("## UpSet plot: ", label)
    ll <- lapply(sel.list, function (x) rownames(x[[sel.list.key]]))
    ll <- ll[lapply(ll, length) > 0]
    if (length(ll) > 1) {
        print(upset(fromList(ll), order.by='freq', nsets=length(ll)))
        outfile <- file.path('upset_plots', paste0(label, '.tsv'))
        lldf <- rownames.first.col(fromList.with.names(ll))
        write.table(lldf, file=outfile, sep='\t', row.names=FALSE)
        mdcat('- [', outfile, ']', '(', outfile, ')')
        pdf.file <- file.path('upset_plots', paste0(label, '.pdf'))
        dev.copy(pdf, file=pdf.file)
        dev.off()
        mdcat('- [', pdf.file, ']', '(', pdf.file, ')')

    } else {
        mdcat('not enough contrasts with upregulated genes')
    }
}


if (length(res.list) > 1){

    dir.create('upset_plots', showWarnings=FALSE)

    mdcat("# UpSet plots {.tabset}")
    mdcat("Here we gather together all the interesting gene sets into an ",
          "[UpSet plot](http://caleydo.org/tools/upset/). These plots show ",
          "the combinatorial overlaps of genes found to be up or down ",
          "across the different contrasts performed. It's much like a ",
          "Venn diagram, but easier to interpret and can scale to many comparisons. ",
          "A TSV file is linked under each plot. This file has rows for each ",
          "gene and columns for each contrast. A `1` indicates that gene was found to be ",
          "up/down/changed in that contrast. "
    )

    for (name in names(sel.list[[1]])){
        plot.upset(name, name)
    }
}
```

```{r genepatterns, child='gene-patterns.Rmd', eval=TRUE}
# Docs: https://lcdb.github.io/lcdb-wf/rnaseq-rmd.html#genepatterns
rmarkdown::render('gene-patterns.Rmd', run_pandoc=FALSE)
```

# Exported results

The files below are TSVs that can be opened in Excel:


```{r, results='asis'}
# Write out files for full and each selection, and create a link to them in the
# HTML generated by this RMarkdown.
for (name in names(res.list)){
  mdcat('## ', res.list[[name]][['label']])
  fn <- paste0(name, '.tsv')
  write.table(res.list[[name]][['res']], file=fn, row.names=FALSE, sep='\t', quote=FALSE)
  mdcat('- [', fn, '](', fn, '), results for ', res.list[[name]][['label']])
}
```

```{r functionalenrichment, child='functional-enrichment.Rmd', eval=TRUE}
# Docs: https://lcdb.github.io/lcdb-wf/rnaseq-rmd.html#functionalenrichment

rmarkdown::render('functional-enrichment.Rmd', run_pandoc=FALSE)
```

# Session info
For reproducibility purposes, here is the output of `sessionInfo()` showing the
versions of all packages used here.

```{r, collapse=FALSE}
sessionInfo()
```

# Help

```{r helpdocs, child="help_docs.Rmd", run_pandoc=FALSE}
# Docs: https://lcdb.github.io/lcdb-wf/rnaseq-rmd.html#helpdocs
# NOTE: optional help section--------------------------------------------------
#   Delete this chunk, or set to eval=FALSE, if you don't want to include the
#   help text from "help_docs.Rmd"
```
